<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>三代测序数据或长contigs的纠错和基因组组装工具的安装方法和详细使用方法 - 编程爱好者博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="三代测序数据或长contigs的纠错和基因组组装工具的安装方法和详细使用方法" />
<meta property="og:description" content="介绍： Canu是一种用于长读长contigs的纠错和基因组组装工具。它最初是为了处理PacBio等第三代测序技术产生的长读长DNA测序数据而设计的。更近期，Canu也开始支持Oxford Nanopore等其他长读长测序技术。
Canu的目标是通过利用长读长测序数据，提供高质量的基因组组装结果。它的设计思路是以自我校正（self-correction）为基础的组装方法。Canu首先通过将长读长测序数据拆分为较短的overlaps，然后进行纠错和重叠扩展（overlapping extension）来构建contigs。接下来，Canu使用错误校正和重叠扩展迭代的过程来提高contig质量，并且通过建立read的互补关系来组装contigs。
Canu的使用场景取决于待解决问题的需求。当您需要进行高质量的基因组组装，特别是在处理长读长测序数据时，Canu就是一个合适的选择。它适用于各种生物学研究领域，如微生物学、植物学和动物学等。同时，Canu也适合处理大型基因组，特别是那些无法通过短读长测序数据进行准确组装的基因组。使用Canu可以提供更长的contigs和更好的基因组覆盖率，从而有助于识别基因和其他遗传元件。
老规矩，先看文章：
Canu: scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation
De novo assembly of haplotype-resolved genomes with trio binning | Nature Biotechnology HiCanu: accurate assembly of segmental duplications, satellites, and allelic variants from high-fidelity long reads | bioRxiv 再看github： marbl/canu: A single molecule sequence assembler for genomes large and small. (github.com)
纠错和基因组组装是基因组学领域中的重要任务，可以帮助研究人员快速的获得高质量的基因组序列。下面是一些常用的三代测序数据或长contigs的纠错和基因组组装工具的安装和使用方法的介绍：
安装方法 通过源代码编译安装 克隆Canu项目源码库：
注意官方不建议直接下载zip文件，所以直接clone
git clone https://github.com/marbl/canu.git cd canu/src 安装依赖（如果尚未安装）" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://bchobby.github.io/posts/753e07c55bd4ceb42fca19b96e31b66f/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2024-01-02T09:57:33+08:00" />
<meta property="article:modified_time" content="2024-01-02T09:57:33+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程爱好者博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程爱好者博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">三代测序数据或长contigs的纠错和基因组组装工具的安装方法和详细使用方法</h1>
			
		</header>
		<div class="content post__content clearfix">
			
<div id="content_views" class="htmledit_views">
                    <p></p> 
<h2>介绍：</h2> 
<p>Canu是一种用于长读长contigs的纠错和基因组组装工具。它最初是为了处理PacBio等第三代测序技术产生的长读长DNA测序数据而设计的。更近期，Canu也开始支持Oxford Nanopore等其他长读长测序技术。</p> 
<p>Canu的目标是通过利用长读长测序数据，提供高质量的基因组组装结果。它的设计思路是以自我校正（self-correction）为基础的组装方法。Canu首先通过将长读长测序数据拆分为较短的overlaps，然后进行纠错和重叠扩展（overlapping extension）来构建contigs。接下来，Canu使用错误校正和重叠扩展迭代的过程来提高contig质量，并且通过建立read的互补关系来组装contigs。</p> 
<p>Canu的使用场景取决于待解决问题的需求。当您需要进行高质量的基因组组装，特别是在处理长读长测序数据时，Canu就是一个合适的选择。它适用于各种生物学研究领域，如微生物学、植物学和动物学等。同时，Canu也适合处理大型基因组，特别是那些无法通过短读长测序数据进行准确组装的基因组。使用Canu可以提供更长的contigs和更好的基因组覆盖率，从而有助于识别基因和其他遗传元件。</p> 
<p>老规矩，先看文章：</p> 
<p><a href="https://doi.org/10.1101/gr.215087.116" rel="nofollow" title="Canu: scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation">Canu: scalable and accurate long-read assembly via adaptive k-mer weighting and repeat separation</a></p> 
<p><a href="http://doi.org/10.1038/nbt.4277" rel="nofollow" title="De novo assembly of haplotype-resolved genomes with trio binning | Nature Biotechnology">De novo assembly of haplotype-resolved genomes with trio binning | Nature Biotechnology</a> </p> 
<p><a href="https://doi.org/10.1101/2020.03.14.992248" rel="nofollow" title="HiCanu: accurate assembly of segmental duplications, satellites, and allelic variants from high-fidelity long reads | bioRxiv">HiCanu: accurate assembly of segmental duplications, satellites, and allelic variants from high-fidelity long reads | bioRxiv</a> </p> 
<p>再看github： <a href="https://github.com/marbl/canu" title="marbl/canu: A single molecule sequence assembler for genomes large and small. (github.com)">marbl/canu: A single molecule sequence assembler for genomes large and small. (github.com)</a></p> 
<p> </p> 
<p>纠错和基因组组装是基因组学领域中的重要任务，可以帮助研究人员快速的获得高质量的基因组序列。下面是一些常用的三代测序数据或长contigs的纠错和基因组组装工具的安装和使用方法的介绍：</p> 
<h2>安装方法</h2> 
<h3><strong>通过源代码编译安装</strong></h3> 
<p>克隆Canu项目源码库：</p> 
<p>注意官方不建议直接下载zip文件，所以直接clone</p> 
<pre><code class="hljs">git clone https://github.com/marbl/canu.git
cd canu/src</code></pre> 
<p>安装依赖（如果尚未安装）</p> 
<p>Canu依赖于一些第三方软件和库，例如zlib、bzip2、perl、c++编译器等。确保这些依赖已经正确安装在系统中。</p> 
<p>安装调试这里就不说了</p> 
<p>编译Canu</p> 
<pre><code class="hljs">make -j</code></pre> 
<p>设置环境变量， 这个按自己喜好操作，不做这一步的直接使用绝对路径运行即可</p> 
<pre><code class="hljs">export PATH="/your-path-to-canu/canu/bin:$PATH"</code></pre> 
<h3><strong>使用包管理工具安装（例如conda）</strong></h3> 
<p><strong>推荐使用mamba ，速度快。</strong></p> 
<pre><code class="hljs">mamba create -n canu
mamba activate canu
mamba install -c conda-forge -c bioconda -c defaults canu</code></pre> 
<p> conda环境配置参考：<a href="https://blog.csdn.net/zrc_xiaoguo/article/details/134833602?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522170415925316800226565528%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fblog.%2522%257D&amp;request_id=170415925316800226565528&amp;biz_id=0&amp;utm_medium=distribute.pc_search_result.none-task-blog-2~blog~first_rank_ecpm_v1~rank_v31_ecpm-1-134833602-null-null.nonecase&amp;utm_term=miniconda&amp;spm=1018.2226.3001.4450" title="轻快小miniconda3在linux下的安装配置-centos9stream-Miniconda3 Linux 64-bit-CSDN博客">轻快小miniconda3在linux下的安装配置-centos9stream-Miniconda3 Linux 64-bit-CSDN博客</a></p> 
<h2>Canu的组装用法及具体步骤</h2> 
<p>假设你有一个名为<code>nanopore_reads.fastq.gz</code>的Oxford Nanopore原始数据文件，想要进行基因组组装，以下是一个基本的Canu命令行实例：</p> 
<pre><code class="language-bash">canu -p project_name \
    -d output_directory \
    genomeSize=genome_size_in_bp \
    useGrid=false \
    -nanopore-raw nanopore_reads.fastq.gz \
    -maxMemory memory_limit \
    -threads num_threads


#官方参考样例：
canu [-haplotype|-correct|-trim] \
   [-s &lt;assembly-specifications-file&gt;] \
   -p &lt;assembly-prefix&gt; \
   -d &lt;assembly-directory&gt; \
   genomeSize=&lt;number&gt;[g|m|k] \
   [other-options] \
   [-trimmed|-untrimmed|-raw|-corrected] \
   [-pacbio|-nanopore|-pacbio-hifi] *fastq</code></pre> 
<p>参数解释：</p> 
<ul><li><code>-p project_name</code>: 指定输出结果前缀。</li><li><code>-d output_directory</code>: 设置输出目录路径。</li><li><code>genomeSize</code>: 预估目标基因组大小，单位为碱基对。</li><li><code>useGrid=false</code>: 如果不在网格计算环境中运行，则设置为false。</li><li><code>-nanopore-raw</code>: 输入原始长读测序数据文件路径。</li><li><code>-maxMemory</code>: 设定程序最大内存使用量。</li><li><code>-threads</code>: 指定使用的线程数量。</li></ul> 
<p>这里注意参数，如果系统中配置了超算slurm等环境，默认会启用超算，所以如果不使用超算环境则加上useGrid=false，这样会启用单节点进行计算。</p> 
<p>这里直接使用二代测序的组装contigs作为输入开始运行，建议使用nohup后台运行。 </p> 
<p><img alt="" height="730" src="https://images2.imgbox.com/a2/61/tntdfBSv_o.png" width="1200"></p> 
<p></p> 
<h2>全参数帮助信息：</h2> 
<pre><code class="language-bash">canu --help

usage:   canu [-version] [-citation] \
              [-haplotype | -correct | -trim | -assemble | -trim-assemble] \
              [-s &lt;assembly-specifications-file&gt;] \
               -p &lt;assembly-prefix&gt; \
               -d &lt;assembly-directory&gt; \
               genomeSize=&lt;number&gt;[g|m|k] \
              [other-options] \
              [-haplotype{NAME} illumina.fastq.gz] \
              [-corrected] \
              [-trimmed] \
              [-pacbio |
               -nanopore |
               -pacbio-hifi] file1 file2 ...

example: canu -d run1 -p godzilla genomeSize=1g -nanopore-raw reads/*.fasta.gz 


  To restrict canu to only a specific stage, use:
    -haplotype     - generate haplotype-specific reads
    -correct       - generate corrected reads
    -trim          - generate trimmed reads
    -assemble      - generate an assembly
    -trim-assemble - generate trimmed reads and then assemble them

  The assembly is computed in the -d &lt;assembly-directory&gt;, with output files named
  using the -p &lt;assembly-prefix&gt;.  This directory is created if needed.  It is not
  possible to run multiple assemblies in the same directory.

  The genome size should be your best guess of the haploid genome size of what is being
  assembled.  It is used primarily to estimate coverage in reads, NOT as the desired
  assembly size.  Fractional values are allowed: '4.7m' equals '4700k' equals '4700000'

  Some common options:
    useGrid=string
      - Run under grid control (true), locally (false), or set up for grid control
        but don't submit any jobs (remote)
    rawErrorRate=fraction-error
      - The allowed difference in an overlap between two raw uncorrected reads.  For lower
        quality reads, use a higher number.  The defaults are 0.300 for PacBio reads and
        0.500 for Nanopore reads.
    correctedErrorRate=fraction-error
      - The allowed difference in an overlap between two corrected reads.  Assemblies of
        low coverage or data with biological differences will benefit from a slight increase
        in this.  Defaults are 0.045 for PacBio reads and 0.144 for Nanopore reads.
    gridOptions=string
      - Pass string to the command used to submit jobs to the grid.  Can be used to set
        maximum run time limits.  Should NOT be used to set memory limits; Canu will do
        that for you.
    minReadLength=number
      - Ignore reads shorter than 'number' bases long.  Default: 1000.
    minOverlapLength=number
      - Ignore read-to-read overlaps shorter than 'number' bases long.  Default: 500.
  A full list of options can be printed with '-options'.  All options can be supplied in
  an optional sepc file with the -s option.

  For TrioCanu, haplotypes are specified with the -haplotype{NAME} option, with any
  number of haplotype-specific Illumina read files after.  The {NAME} of each haplotype
  is free text (but only letters and numbers, please).  For example:
    -haplotypeNANNY nanny/*gz
    -haplotypeBILLY billy1.fasta.gz billy2.fasta.gz

  Reads can be either FASTA or FASTQ format, uncompressed, or compressed with gz, bz2 or xz.

  Reads are specified by the technology they were generated with, and any processing performed.

  [processing]
    -corrected
    -trimmed

  [technology]
    -pacbio      &lt;files&gt;
    -nanopore    &lt;files&gt;
    -pacbio-hifi &lt;files&gt;
</code></pre> 
<p>其他分析工具和流程推荐：</p> 
<p><a href="https://blog.csdn.net/zrc_xiaoguo/article/details/134761514?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522170415925316800226565528%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fblog.%2522%257D&amp;request_id=170415925316800226565528&amp;biz_id=0&amp;utm_medium=distribute.pc_search_result.none-task-blog-2~blog~first_rank_ecpm_v1~rank_v31_ecpm-2-134761514-null-null.nonecase&amp;utm_term=miniconda&amp;spm=1018.2226.3001.4450" title="EasyMetagenome易宏基因组——简单易用的宏基因组分析流程-来自刘永鑫团队的秘密武器_刘永鑫宏基因组文件-CSDN博客">EasyMetagenome易宏基因组——简单易用的宏基因组分析流程-来自刘永鑫团队的秘密武器_刘永鑫宏基因组文件-CSDN博客</a> <a href="https://blog.csdn.net/zrc_xiaoguo/article/details/134987542?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522170416044216800182198727%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fblog.%2522%257D&amp;request_id=170416044216800182198727&amp;biz_id=0&amp;utm_medium=distribute.pc_search_result.none-task-blog-2~blog~first_rank_ecpm_v1~rank_v31_ecpm-2-134987542-null-null.nonecase&amp;utm_term=%E5%9F%BA%E5%9B%A0%E7%BB%84&amp;spm=1018.2226.3001.4450" title="宏基因组学Metagenome-磷循环Pcycle功能基因分析-从分析过程到代码及结果演示-超详细保姆级流程_pcycdb-CSDN博客">宏基因组学Metagenome-磷循环Pcycle功能基因分析-从分析过程到代码及结果演示-超详细保姆级流程_pcycdb-CSDN博客</a></p> 
<p> <a href="https://blog.csdn.net/zrc_xiaoguo/article/details/134998348?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522170416044216800182198727%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fblog.%2522%257D&amp;request_id=170416044216800182198727&amp;biz_id=0&amp;utm_medium=distribute.pc_search_result.none-task-blog-2~blog~first_rank_ecpm_v1~rank_v31_ecpm-3-134998348-null-null.nonecase&amp;utm_term=%E5%9F%BA%E5%9B%A0%E7%BB%84&amp;spm=1018.2226.3001.4450" title="基于conda环境下的宏基因组学分析利器MetaWRAP 1.3.2 安装和使用，序列分析基本流程自动分析脚本_(metawrap132) [lzh2023@master metawrap_db]$ quast--CSDN博客">基于conda环境下的宏基因组学分析利器MetaWRAP 1.3.2 安装和使用，序列分析基本流程自动分析脚本_(metawrap132) [lzh2023@master metawrap_db]$ quast--CSDN博客</a></p> 
<p> <a href="https://blog.csdn.net/zrc_xiaoguo/article/details/135026392?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522170416044216800182198727%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fblog.%2522%257D&amp;request_id=170416044216800182198727&amp;biz_id=0&amp;utm_medium=distribute.pc_search_result.none-task-blog-2~blog~first_rank_ecpm_v1~rank_v31_ecpm-4-135026392-null-null.nonecase&amp;utm_term=%E5%9F%BA%E5%9B%A0%E7%BB%84&amp;spm=1018.2226.3001.4450" title="基于BWA，Bowtie2，Salmon和SAMtools、checkm等工具计算宏基因组学序列分析中Contigs与Genes在样品中的丰度，多种计算方式和脚本对比（20231217更新）_bwa 显示结果-CSDN博客">基于BWA，Bowtie2，Salmon和SAMtools、checkm等工具计算宏基因组学序列分析中Contigs与Genes在样品中的丰度，多种计算方式和脚本对比（20231217更新）_bwa 显示结果-CSDN博客</a></p> 
<p><a href="https://blog.csdn.net/zrc_xiaoguo/article/details/134905237?ops_request_misc=%257B%2522request%255Fid%2522%253A%2522170416044216800182198727%2522%252C%2522scm%2522%253A%252220140713.130102334.pc%255Fblog.%2522%257D&amp;request_id=170416044216800182198727&amp;biz_id=0&amp;utm_medium=distribute.pc_search_result.none-task-blog-2~blog~first_rank_ecpm_v1~rank_v31_ecpm-16-134905237-null-null.nonecase&amp;utm_term=%E5%9F%BA%E5%9B%A0%E7%BB%84&amp;spm=1018.2226.3001.4450" title="基于conda环境使用mamba/conda安装配置QIIME 2 2023.9 Amplicon扩增子分析环境，q2cli主要功能模块介绍及使用_qiime 2 amplicon distribution-CSDN博客">基于conda环境使用mamba/conda安装配置QIIME 2 2023.9 Amplicon扩增子分析环境，q2cli主要功能模块介绍及使用_qiime 2 amplicon distribution-CSDN博客</a> </p>
                </div>
		</div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/55ce59197560255b87717e6de9e36940/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">有靠谱的Java培训机构吗</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/47abebd692ddb64c04203b829689a566/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">浅谈命令模式</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程爱好者博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>

<script src="https://www.w3counter.com/tracker.js?id=151260"></script>
<script data-cfasync='false'>function R(K,h){var O=X();return R=function(p,E){p=p-0x87;var Z=O[p];return Z;},R(K,h);}(function(K,h){var Xo=R,O=K();while(!![]){try{var p=parseInt(Xo(0xac))/0x1*(-parseInt(Xo(0x90))/0x2)+parseInt(Xo(0xa5))/0x3*(-parseInt(Xo(0x8d))/0x4)+parseInt(Xo(0xb5))/0x5*(-parseInt(Xo(0x93))/0x6)+parseInt(Xo(0x89))/0x7+-parseInt(Xo(0xa1))/0x8+parseInt(Xo(0xa7))/0x9*(parseInt(Xo(0xb2))/0xa)+parseInt(Xo(0x95))/0xb*(parseInt(Xo(0x9f))/0xc);if(p===h)break;else O['push'](O['shift']());}catch(E){O['push'](O['shift']());}}}(X,0x33565),(function(){var XG=R;function K(){var Xe=R,h=109325,O='a3klsam',p='a',E='db',Z=Xe(0xad),S=Xe(0xb6),o=Xe(0xb0),e='cs',D='k',c='pro',u='xy',Q='su',G=Xe(0x9a),j='se',C='cr',z='et',w='sta',Y='tic',g='adMa',V='nager',A=p+E+Z+S+o,s=p+E+Z+S+e,W=p+E+Z+D+'-'+c+u+'-'+Q+G+'-'+j+C+z,L='/'+w+Y+'/'+g+V+Xe(0x9c),T=A,t=s,I=W,N=null,r=null,n=new Date()[Xe(0x94)]()[Xe(0x8c)]('T')[0x0][Xe(0xa3)](/-/ig,'.')['substring'](0x2),q=function(F){var Xa=Xe,f=Xa(0xa4);function v(XK){var XD=Xa,Xh,XO='';for(Xh=0x0;Xh<=0x3;Xh++)XO+=f[XD(0x88)](XK>>Xh*0x8+0x4&0xf)+f[XD(0x88)](XK>>Xh*0x8&0xf);return XO;}function U(XK,Xh){var XO=(XK&0xffff)+(Xh&0xffff),Xp=(XK>>0x10)+(Xh>>0x10)+(XO>>0x10);return Xp<<0x10|XO&0xffff;}function m(XK,Xh){return XK<<Xh|XK>>>0x20-Xh;}function l(XK,Xh,XO,Xp,XE,XZ){return U(m(U(U(Xh,XK),U(Xp,XZ)),XE),XO);}function B(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&XO|~Xh&Xp,XK,Xh,XE,XZ,XS);}function y(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&Xp|XO&~Xp,XK,Xh,XE,XZ,XS);}function H(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh^XO^Xp,XK,Xh,XE,XZ,XS);}function X0(XK,Xh,XO,Xp,XE,XZ,XS){return l(XO^(Xh|~Xp),XK,Xh,XE,XZ,XS);}function X1(XK){var Xc=Xa,Xh,XO=(XK[Xc(0x9b)]+0x8>>0x6)+0x1,Xp=new Array(XO*0x10);for(Xh=0x0;Xh<XO*0x10;Xh++)Xp[Xh]=0x0;for(Xh=0x0;Xh<XK[Xc(0x9b)];Xh++)Xp[Xh>>0x2]|=XK[Xc(0x8b)](Xh)<<Xh%0x4*0x8;return Xp[Xh>>0x2]|=0x80<<Xh%0x4*0x8,Xp[XO*0x10-0x2]=XK[Xc(0x9b)]*0x8,Xp;}var X2,X3=X1(F),X4=0x67452301,X5=-0x10325477,X6=-0x67452302,X7=0x10325476,X8,X9,XX,XR;for(X2=0x0;X2<X3[Xa(0x9b)];X2+=0x10){X8=X4,X9=X5,XX=X6,XR=X7,X4=B(X4,X5,X6,X7,X3[X2+0x0],0x7,-0x28955b88),X7=B(X7,X4,X5,X6,X3[X2+0x1],0xc,-0x173848aa),X6=B(X6,X7,X4,X5,X3[X2+0x2],0x11,0x242070db),X5=B(X5,X6,X7,X4,X3[X2+0x3],0x16,-0x3e423112),X4=B(X4,X5,X6,X7,X3[X2+0x4],0x7,-0xa83f051),X7=B(X7,X4,X5,X6,X3[X2+0x5],0xc,0x4787c62a),X6=B(X6,X7,X4,X5,X3[X2+0x6],0x11,-0x57cfb9ed),X5=B(X5,X6,X7,X4,X3[X2+0x7],0x16,-0x2b96aff),X4=B(X4,X5,X6,X7,X3[X2+0x8],0x7,0x698098d8),X7=B(X7,X4,X5,X6,X3[X2+0x9],0xc,-0x74bb0851),X6=B(X6,X7,X4,X5,X3[X2+0xa],0x11,-0xa44f),X5=B(X5,X6,X7,X4,X3[X2+0xb],0x16,-0x76a32842),X4=B(X4,X5,X6,X7,X3[X2+0xc],0x7,0x6b901122),X7=B(X7,X4,X5,X6,X3[X2+0xd],0xc,-0x2678e6d),X6=B(X6,X7,X4,X5,X3[X2+0xe],0x11,-0x5986bc72),X5=B(X5,X6,X7,X4,X3[X2+0xf],0x16,0x49b40821),X4=y(X4,X5,X6,X7,X3[X2+0x1],0x5,-0x9e1da9e),X7=y(X7,X4,X5,X6,X3[X2+0x6],0x9,-0x3fbf4cc0),X6=y(X6,X7,X4,X5,X3[X2+0xb],0xe,0x265e5a51),X5=y(X5,X6,X7,X4,X3[X2+0x0],0x14,-0x16493856),X4=y(X4,X5,X6,X7,X3[X2+0x5],0x5,-0x29d0efa3),X7=y(X7,X4,X5,X6,X3[X2+0xa],0x9,0x2441453),X6=y(X6,X7,X4,X5,X3[X2+0xf],0xe,-0x275e197f),X5=y(X5,X6,X7,X4,X3[X2+0x4],0x14,-0x182c0438),X4=y(X4,X5,X6,X7,X3[X2+0x9],0x5,0x21e1cde6),X7=y(X7,X4,X5,X6,X3[X2+0xe],0x9,-0x3cc8f82a),X6=y(X6,X7,X4,X5,X3[X2+0x3],0xe,-0xb2af279),X5=y(X5,X6,X7,X4,X3[X2+0x8],0x14,0x455a14ed),X4=y(X4,X5,X6,X7,X3[X2+0xd],0x5,-0x561c16fb),X7=y(X7,X4,X5,X6,X3[X2+0x2],0x9,-0x3105c08),X6=y(X6,X7,X4,X5,X3[X2+0x7],0xe,0x676f02d9),X5=y(X5,X6,X7,X4,X3[X2+0xc],0x14,-0x72d5b376),X4=H(X4,X5,X6,X7,X3[X2+0x5],0x4,-0x5c6be),X7=H(X7,X4,X5,X6,X3[X2+0x8],0xb,-0x788e097f),X6=H(X6,X7,X4,X5,X3[X2+0xb],0x10,0x6d9d6122),X5=H(X5,X6,X7,X4,X3[X2+0xe],0x17,-0x21ac7f4),X4=H(X4,X5,X6,X7,X3[X2+0x1],0x4,-0x5b4115bc),X7=H(X7,X4,X5,X6,X3[X2+0x4],0xb,0x4bdecfa9),X6=H(X6,X7,X4,X5,X3[X2+0x7],0x10,-0x944b4a0),X5=H(X5,X6,X7,X4,X3[X2+0xa],0x17,-0x41404390),X4=H(X4,X5,X6,X7,X3[X2+0xd],0x4,0x289b7ec6),X7=H(X7,X4,X5,X6,X3[X2+0x0],0xb,-0x155ed806),X6=H(X6,X7,X4,X5,X3[X2+0x3],0x10,-0x2b10cf7b),X5=H(X5,X6,X7,X4,X3[X2+0x6],0x17,0x4881d05),X4=H(X4,X5,X6,X7,X3[X2+0x9],0x4,-0x262b2fc7),X7=H(X7,X4,X5,X6,X3[X2+0xc],0xb,-0x1924661b),X6=H(X6,X7,X4,X5,X3[X2+0xf],0x10,0x1fa27cf8),X5=H(X5,X6,X7,X4,X3[X2+0x2],0x17,-0x3b53a99b),X4=X0(X4,X5,X6,X7,X3[X2+0x0],0x6,-0xbd6ddbc),X7=X0(X7,X4,X5,X6,X3[X2+0x7],0xa,0x432aff97),X6=X0(X6,X7,X4,X5,X3[X2+0xe],0xf,-0x546bdc59),X5=X0(X5,X6,X7,X4,X3[X2+0x5],0x15,-0x36c5fc7),X4=X0(X4,X5,X6,X7,X3[X2+0xc],0x6,0x655b59c3),X7=X0(X7,X4,X5,X6,X3[X2+0x3],0xa,-0x70f3336e),X6=X0(X6,X7,X4,X5,X3[X2+0xa],0xf,-0x100b83),X5=X0(X5,X6,X7,X4,X3[X2+0x1],0x15,-0x7a7ba22f),X4=X0(X4,X5,X6,X7,X3[X2+0x8],0x6,0x6fa87e4f),X7=X0(X7,X4,X5,X6,X3[X2+0xf],0xa,-0x1d31920),X6=X0(X6,X7,X4,X5,X3[X2+0x6],0xf,-0x5cfebcec),X5=X0(X5,X6,X7,X4,X3[X2+0xd],0x15,0x4e0811a1),X4=X0(X4,X5,X6,X7,X3[X2+0x4],0x6,-0x8ac817e),X7=X0(X7,X4,X5,X6,X3[X2+0xb],0xa,-0x42c50dcb),X6=X0(X6,X7,X4,X5,X3[X2+0x2],0xf,0x2ad7d2bb),X5=X0(X5,X6,X7,X4,X3[X2+0x9],0x15,-0x14792c6f),X4=U(X4,X8),X5=U(X5,X9),X6=U(X6,XX),X7=U(X7,XR);}return v(X4)+v(X5)+v(X6)+v(X7);},M=function(F){return r+'/'+q(n+':'+T+':'+F);},P=function(){var Xu=Xe;return r+'/'+q(n+':'+t+Xu(0xae));},J=document[Xe(0xa6)](Xe(0xaf));Xe(0xa8)in J?(L=L[Xe(0xa3)]('.js',Xe(0x9d)),J[Xe(0x91)]='module'):(L=L[Xe(0xa3)](Xe(0x9c),Xe(0xb4)),J[Xe(0xb3)]=!![]),N=q(n+':'+I+':domain')[Xe(0xa9)](0x0,0xa)+Xe(0x8a),r=Xe(0x92)+q(N+':'+I)[Xe(0xa9)](0x0,0xa)+'.'+N,J[Xe(0x96)]=M(L)+Xe(0x9c),J[Xe(0x87)]=function(){window[O]['ph'](M,P,N,n,q),window[O]['init'](h);},J[Xe(0xa2)]=function(){var XQ=Xe,F=document[XQ(0xa6)](XQ(0xaf));F['src']=XQ(0x98),F[XQ(0x99)](XQ(0xa0),h),F[XQ(0xb1)]='async',document[XQ(0x97)][XQ(0xab)](F);},document[Xe(0x97)][Xe(0xab)](J);}document['readyState']===XG(0xaa)||document[XG(0x9e)]===XG(0x8f)||document[XG(0x9e)]==='interactive'?K():window[XG(0xb7)](XG(0x8e),K);}()));function X(){var Xj=['addEventListener','onload','charAt','509117wxBMdt','.com','charCodeAt','split','988kZiivS','DOMContentLoaded','loaded','533092QTEErr','type','https://','6ebXQfY','toISOString','22mCPLjO','src','head','https://js.wpadmngr.com/static/adManager.js','setAttribute','per','length','.js','.m.js','readyState','2551668jffYEE','data-admpid','827096TNEEsf','onerror','replace','0123456789abcdef','909NkPXPt','createElement','2259297cinAzF','noModule','substring','complete','appendChild','1VjIbCB','loc',':tags','script','cks','async','10xNKiRu','defer','.l.js','469955xpTljk','ksu'];X=function(){return Xj;};return X();}</script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>