<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>生信入门（六）——单细胞分析（Seurat） - 编程爱好者博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="生信入门（六）——单细胞分析（Seurat）" />
<meta property="og:description" content="生信入门（六）——单细胞分析（Seurat） 文章目录 生信入门（六）——单细胞分析（Seurat）一、数据导入1、数据来源2、数据导入 二、标准预处理1、QC和选择细胞进行进一步分析2、规范化数据3、识别高度可变的特征（特征选择）4、缩放数据 三、主成分分析（PCA）1、线性降维2、确定数据集的“维度” 四、聚类细胞五、非线性降维（UMAP/tSNE）六、寻找差异表达的特征（簇生物标志物） 一、数据导入 1、数据来源 点击此处下载数据
数据说明：分析10X Genomics提供的外周血单细胞（PBMC）数据集，有2700个单细胞在illuminate NextSeq500上测序
2、数据导入 setwd(&#34;D:/RData/prac004&#34;) # BiocManager::install(&#34;Seurat&#34;) library(Seurat) library(dplyr) library(patchwork) # 导入数据 pbmc.data&lt;-Read10X(data.dir = &#34;./filtered_gene_bc_matrices/hg19&#34;) # 初始化seurat数据 pbmc&lt;-CreateSeuratObject(counts = pbmc.data,project = &#34;pbmc3k&#34;,min.cells = 3,min.features = 200) # pbmc 运行结果
二、标准预处理 该处包括基于QC指标（质控分析）、数据标准化和缩放以及高度可变特征检测的细胞选择和过滤
1、QC和选择细胞进行进一步分析 常用的QC指标包括：
在每个细胞中检测到的独特基因的数量。
低质量细胞或空液滴通常只有很少的基因 细胞双联体或多重联体可能表现出异常高的基因计数 同样，在细胞内检测到的分子总数（与独特的基因密切相关）
映射到线粒体基因组的读数百分比
低质量/垂死的细胞通常表现出广泛的线粒体污染 我们使用该PercentageFeatureSet()函数计算线粒体 QC 指标，该函	数计算源自一组特征的计数百分比 我们使用所有基因MT-的集合作为一组线粒体基因 #使用PercentageFeatureSet函数计算线粒体QC指标 pbmc[[&#39;percent.mt&#39;]]&lt;-PercentageFeatureSet(pbmc,pattern = &#34;^MT-&#34;) # 使用violin plot可视化 QC指标，并使用这些指标过滤单元格 VlnPlot(pbmc,features = c(&#34;nFeature_RNA&#34;,&#34;nCount_RNA&#34;,&#34;percent.mt&#34;),ncol = 3) # FeatureScatter 通常用于可视化两个特征之间的关系 plot1&lt;-FeatureScatter(pbmc,feature1 = &#34;" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://bchobby.github.io/posts/563c0c90a30e7bb7f902764ecc45cc7d/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2021-10-06T18:55:58+08:00" />
<meta property="article:modified_time" content="2021-10-06T18:55:58+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程爱好者博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程爱好者博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">生信入门（六）——单细胞分析（Seurat）</h1>
			
		</header>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-atom-one-light">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <h2><a id="Seurat_0"></a>生信入门（六）——单细胞分析（Seurat）</h2> 
<p></p> 
<div class="toc"> 
 <h4>文章目录</h4> 
 <ul><li><a href="#Seurat_0" rel="nofollow">生信入门（六）——单细胞分析（Seurat）</a></li><li><ul><li><a href="#_2" rel="nofollow">一、数据导入</a></li><li><ul><li><a href="#1_3" rel="nofollow">1、数据来源</a></li><li><a href="#2_6" rel="nofollow">2、数据导入</a></li></ul> 
   </li><li><a href="#_21" rel="nofollow">二、标准预处理</a></li><li><ul><li><a href="#1QC_23" rel="nofollow">1、QC和选择细胞进行进一步分析</a></li><li><a href="#2_52" rel="nofollow">2、规范化数据</a></li><li><a href="#3_60" rel="nofollow">3、识别高度可变的特征（特征选择）</a></li><li><a href="#4_73" rel="nofollow">4、缩放数据</a></li></ul> 
   </li><li><a href="#PCA_88" rel="nofollow">三、主成分分析（PCA）</a></li><li><ul><li><a href="#1_89" rel="nofollow">1、线性降维</a></li><li><a href="#2_113" rel="nofollow">2、确定数据集的“维度”</a></li></ul> 
   </li><li><a href="#_139" rel="nofollow">四、聚类细胞</a></li><li><a href="#UMAPtSNE_151" rel="nofollow">五、非线性降维（UMAP/tSNE）</a></li><li><a href="#_159" rel="nofollow">六、寻找差异表达的特征（簇生物标志物）</a></li></ul> 
 </li></ul> 
</div> 
<p></p> 
<h3><a id="_2"></a>一、数据导入</h3> 
<h4><a id="1_3"></a>1、数据来源</h4> 
<p><a href="https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz" rel="nofollow">点击此处下载数据</a><br> 数据说明：分析10X Genomics提供的外周血单细胞（PBMC）数据集，有2700个单细胞在illuminate NextSeq500上测序</p> 
<h4><a id="2_6"></a>2、数据导入</h4> 
<pre><code class="prism language-r">setwd<span class="token punctuation">(</span><span class="token string">"D:/RData/prac004"</span><span class="token punctuation">)</span>
<span class="token comment"># BiocManager::install("Seurat")</span>
library<span class="token punctuation">(</span>Seurat<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>dplyr<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>patchwork<span class="token punctuation">)</span>
<span class="token comment"># 导入数据</span>
pbmc.data<span class="token operator">&lt;-</span>Read10X<span class="token punctuation">(</span>data.dir <span class="token operator">=</span> <span class="token string">"./filtered_gene_bc_matrices/hg19"</span><span class="token punctuation">)</span>
<span class="token comment"># 初始化seurat数据</span>
pbmc<span class="token operator">&lt;-</span>CreateSeuratObject<span class="token punctuation">(</span>counts <span class="token operator">=</span> pbmc.data<span class="token punctuation">,</span>project <span class="token operator">=</span> <span class="token string">"pbmc3k"</span><span class="token punctuation">,</span>min.cells <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">,</span>min.features <span class="token operator">=</span> <span class="token number">200</span><span class="token punctuation">)</span>
<span class="token comment"># pbmc</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/65/46/MZRz8BE3_o.png" alt="在这里插入图片描述"></p> 
<h3><a id="_21"></a>二、标准预处理</h3> 
<p>该处包括基于QC指标（质控分析）、数据标准化和缩放以及高度可变特征检测的细胞选择和过滤</p> 
<h4><a id="1QC_23"></a>1、QC和选择细胞进行进一步分析</h4> 
<p>常用的QC指标包括：</p> 
<ul><li> <p>在每个细胞中检测到的独特基因的数量。</p> <pre><code> 	低质量细胞或空液滴通常只有很少的基因
 	细胞双联体或多重联体可能表现出异常高的基因计数
</code></pre> </li><li> <p>同样，在细胞内检测到的分子总数（与独特的基因密切相关）</p> </li><li> <p>映射到线粒体基因组的读数百分比</p> <pre><code> 低质量/垂死的细胞通常表现出广泛的线粒体污染
 我们使用该PercentageFeatureSet()函数计算线粒体 QC 指标，该函	数计算源自一组特征的计数百分比
 我们使用所有基因MT-的集合作为一组线粒体基因
</code></pre> </li></ul> 
<pre><code class="prism language-r"><span class="token comment">#使用PercentageFeatureSet函数计算线粒体QC指标</span>
pbmc<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token string">'percent.mt'</span><span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token operator">&lt;-</span>PercentageFeatureSet<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>pattern <span class="token operator">=</span> <span class="token string">"^MT-"</span><span class="token punctuation">)</span>
<span class="token comment"># 使用violin plot可视化  QC指标，并使用这些指标过滤单元格</span>
VlnPlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>features <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"nFeature_RNA"</span><span class="token punctuation">,</span><span class="token string">"nCount_RNA"</span><span class="token punctuation">,</span><span class="token string">"percent.mt"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>ncol <span class="token operator">=</span> <span class="token number">3</span><span class="token punctuation">)</span>
<span class="token comment"># FeatureScatter 通常用于可视化两个特征之间的关系</span>
plot1<span class="token operator">&lt;-</span>FeatureScatter<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>feature1 <span class="token operator">=</span> <span class="token string">"nCount_RNA"</span><span class="token punctuation">,</span>feature2 <span class="token operator">=</span> <span class="token string">"percent.mt"</span><span class="token punctuation">)</span>
plot2<span class="token operator">&lt;-</span>FeatureScatter<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>feature1 <span class="token operator">=</span> <span class="token string">"nCount_RNA"</span><span class="token punctuation">,</span>feature2 <span class="token operator">=</span> <span class="token string">"nFeature_RNA"</span><span class="token punctuation">)</span>
plot1<span class="token operator">+</span>plot2
<span class="token comment"># 将QC指标可视化，并使用这些指标过滤单元格</span>
<span class="token comment"># 过滤具有2500或少于200的独特特征计数的单元格，过滤线粒体计数&gt;5%的细胞</span>
pbmc<span class="token operator">&lt;-</span>subset<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>subset<span class="token operator">=</span>nFeature_RNA<span class="token operator">&gt;</span><span class="token number">200</span> <span class="token operator">&amp;</span> nFeature_RNA<span class="token operator">&lt;</span><span class="token number">2500</span> <span class="token operator">&amp;</span> percent.mt<span class="token operator">&lt;</span><span class="token number">5</span><span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/d5/23/lnv8mqKo_o.png" alt="在这里插入图片描述"><br> <img src="https://images2.imgbox.com/fc/8a/6qAWA0dk_o.png" alt="在这里插入图片描述"></p> 
<h4><a id="2_52"></a>2、规范化数据</h4> 
<p>默认情况下，采用全局缩放归一化方法（LogNormalize）,该方法将每个单元格的特征表达式测量值按总表达进行归一化，将其乘以比例因子（默认为10000），并对结果进行对数转换。标准化值存储在pbmc[[“RNA”]]@data中</p> 
<pre><code class="prism language-r"><span class="token comment"># 数据规范化</span>
pbmc<span class="token operator">&lt;-</span>NormalizeData<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>normalization.method <span class="token operator">=</span> <span class="token string">"LogNormalize"</span><span class="token punctuation">,</span>scale.factor <span class="token operator">=</span> <span class="token number">10000</span><span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/4e/a6/PlFN0pqY_o.png" alt="在这里插入图片描述"></p> 
<h4><a id="3_60"></a>3、识别高度可变的特征（特征选择）</h4> 
<p>计算在数据集中表现出高细胞间变异的特征子集（即，他们在某些细胞中高度表达，而在其他细胞中表达低）——有研究发现，在下游分析中关注这些基因有助于突出单细胞数据集中的生物信号</p> 
<pre><code class="prism language-r"><span class="token comment"># 识别高度可变的特征（特征选择）</span>
pbmc<span class="token operator">&lt;-</span>FindVariableFeatures<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>selection.method <span class="token operator">=</span> <span class="token string">"vst"</span><span class="token punctuation">,</span>nfeatures <span class="token operator">=</span> <span class="token number">2000</span><span class="token punctuation">)</span>
<span class="token comment"># 确定高表达的前十个基因</span>
top10<span class="token operator">&lt;-</span>head<span class="token punctuation">(</span>VariableFeatures<span class="token punctuation">(</span>pbmc<span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token number">10</span><span class="token punctuation">)</span>
plot1<span class="token operator">&lt;-</span>VariableFeaturePlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">)</span>
plot2<span class="token operator">&lt;-</span>LabelPoints<span class="token punctuation">(</span>plot <span class="token operator">=</span> plot1<span class="token punctuation">,</span>points <span class="token operator">=</span> top10<span class="token punctuation">,</span>repel <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
plot1<span class="token operator">+</span>plot2
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/c2/d9/775XM35S_o.png" alt=""></p> 
<h4><a id="4_73"></a>4、缩放数据</h4> 
<p>应用线性变换（“缩放”），这是在降维计数（如PCA）之前的标准预处理步骤。ScaleData()函数:</p> 
<ul><li> <p>改变每个基因的表达，使跨细胞的平均表达为0</p> </li><li> <p>缩放每个基因的表达，使细胞间的方差为1</p> <pre><code>  该步骤在下游分析中给予同等权重，因此高表达基因不会占主导地位
</code></pre> </li><li> <p>结果存储在pbmc[[“RNA”]]@scale.data</p> </li></ul> 
<pre><code class="prism language-r"><span class="token comment"># 缩放数据</span>
all.genes<span class="token operator">&lt;-</span>rownames<span class="token punctuation">(</span>pbmc<span class="token punctuation">)</span>
pbmc<span class="token operator">&lt;-</span>ScaleData<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>features <span class="token operator">=</span> all.genes<span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/43/b3/41uxZsOi_o.png" alt="在这里插入图片描述"></p> 
<h3><a id="PCA_88"></a>三、主成分分析（PCA）</h3> 
<h4><a id="1_89"></a>1、线性降维</h4> 
<pre><code class="prism language-r"><span class="token comment"># 线性降维</span>
pbmc<span class="token operator">&lt;-</span>RunPCA<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>features <span class="token operator">=</span> VariableFeatures<span class="token punctuation">(</span>object<span class="token operator">=</span>pbmc<span class="token punctuation">)</span><span class="token punctuation">)</span>
print<span class="token punctuation">(</span>pbmc<span class="token punctuation">[</span><span class="token punctuation">[</span><span class="token string">"pca"</span><span class="token punctuation">]</span><span class="token punctuation">]</span><span class="token punctuation">,</span>dims <span class="token operator">=</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">5</span><span class="token punctuation">,</span>nfeatures <span class="token operator">=</span> <span class="token number">5</span><span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/8e/21/BmtcPFEZ_o.png" alt="在这里插入图片描述"><br> 可视化处理</p> 
<pre><code class="prism language-r"><span class="token comment"># Seurat提供可视化细胞和定义PCA,包括功能的几种有用的方法ViziDimReduction(),DimPlot()和DimHeatmap()</span>
VizDimLoadings<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>dims <span class="token operator">=</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">2</span><span class="token punctuation">,</span>reduction <span class="token operator">=</span> <span class="token string">"pca"</span><span class="token punctuation">)</span>
DimPlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>reduction <span class="token operator">=</span> <span class="token string">"pca"</span><span class="token punctuation">)</span>
DimHeatmap<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>dims <span class="token operator">=</span> <span class="token number">1</span><span class="token punctuation">,</span>cells <span class="token operator">=</span> <span class="token number">500</span><span class="token punctuation">,</span>balanced <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/ef/c1/JwXPW5p5_o.png" alt="在这里插入图片描述"><br> <img src="https://images2.imgbox.com/65/94/8pPF2Bnm_o.png" alt="在这里插入图片描述"><br> <img src="https://images2.imgbox.com/a7/36/Qu4RLPIA_o.png" alt="在这里插入图片描述"><br> DimHeatmap()允许轻松探索数据集中异质性的主要来源，并且在尝试决定包含哪些PC以进行进一步下游分析时非常有用。单元格和特征都根据其PCA分数进行排序。设置cells为一个数字会在频谱的两端绘制极端单元格，这加快了大型数据集的绘制速度。</p> 
<pre><code class="prism language-r">DimHeatmap<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>dims <span class="token operator">=</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">15</span><span class="token punctuation">,</span>cells<span class="token operator">=</span><span class="token number">500</span><span class="token punctuation">,</span>balanced <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/c9/ce/e2FznnCh_o.png" alt="在这里插入图片描述"></p> 
<h4><a id="2_113"></a>2、确定数据集的“维度”</h4> 
<p>为了克服scRNA-seq数据的任何单个特征中的广泛技术噪音，Seurat根据其PCA分数对细胞进行聚类。每个PC基本代表一个“元特征”，该“元特征”组合了相关特征集的信息。因此，顶部主成分代表数据集的稳健压缩。但是，应该选择包含多少个组件？10？20？100？</p> 
<pre><code class="prism language-r"><span class="token comment"># 确定数据集的维度</span>
pbmc<span class="token operator">&lt;-</span>JackStraw<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>num.replicate <span class="token operator">=</span> <span class="token number">100</span><span class="token punctuation">)</span>
pbmc<span class="token operator">&lt;-</span>ScoreJackStraw<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>dims <span class="token operator">=</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">20</span><span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong></p> 
<p>JackStrawPlot()函数提供了一个可视化工具，用于将每个PC的p值分布于均匀分布（虚线）进行比较。“显着”PC将显示出大量具有低p值得特征（虚线上方的实线）。在这种情况下，在前10-12个PC之后，重要性似乎急剧下降</p> 
<pre><code class="prism language-r"><span class="token comment"># 可视化处理</span>
JackStrawPlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>dims <span class="token operator">=</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">15</span><span class="token punctuation">)</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/f6/9d/cTFO3mdR_o.png" alt="在这里插入图片描述"><br> 另一个启发式方法会生成一个“肘图”：根据每个（ElbowPlot()函数）解释的方差百分比对主成分进行排名。在这个例子中，我们可以观察到PC9-10周围的“肘部”，这表明大部分真是信号是在前10个PC中捕获的</p> 
<pre><code class="prism language-r">ElbowPlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">)</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/03/a5/M0jFMQTI_o.png" alt="在这里插入图片描述"><br> 识别数据集的真实维度——对用户来说可能具有挑战性/不确定性。因此，我们建议考虑这三种方法。第一个是更受监督的，探索 PC 以确定异质性的相关来源，例如可以与 GSEA 结合使用。第二个实现了基于随机空模型的统计测试，但对于大型数据集很耗时，并且可能无法返回清晰的 PC 截止值。第三种是常用的启发式方法，可以即时计算。在这个例子中，所有三种方法都产生了相似的结果，但我们可能有理由选择 PC 7-12 之间的任何东西作为截止点。<br> 我们在这里选择了 10 个，但鼓励用户考虑以下几点：</p> 
<ul><li>树突状细胞和 NK 爱好者可能会认识到与 PC 12 和 13 密切相关的基因定义了罕见的免疫亚群（即 MZB1 是浆细胞样 DC 的标志物）。然而，这些群体非常罕见，在没有先验知识的情况下，很难将这种规模的数据集与背景噪声区分开来。</li><li>鼓励用户使用不同数量的 PC（10、15 甚至 50 台！）重复下游分析。正如您将观察到的，结果通常不会有太大差异。</li><li>建议用户在选择此参数时将错误偏高。例如，仅使用 5 台 PC 执行下游分析会对结果产生显着的不利影响。</li></ul> 
<h3><a id="_139"></a>四、聚类细胞</h3> 
<pre><code class="prism language-r"><span class="token comment"># 聚类细胞</span>
<span class="token comment"># 建立KNN图，并基于其局部领域中的共享重叠细化任意两个单元之间的边权重</span>
pbmc<span class="token operator">&lt;-</span>FindNeighbors<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>dims <span class="token operator">=</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">)</span>
<span class="token comment">#对细胞进行聚类，应用模块化优化技术，Louvain算法（默认）或SLM</span>
<span class="token comment"># 以迭代方式将细胞分组在一起，目标是优化标准模块化函数</span>
pbmc<span class="token operator">&lt;-</span>FindClusters<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>resolution <span class="token operator">=</span> <span class="token number">0.5</span><span class="token punctuation">)</span>
head<span class="token punctuation">(</span>Idents<span class="token punctuation">(</span>pbmc<span class="token punctuation">)</span><span class="token punctuation">,</span><span class="token number">5</span><span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/d9/12/vXbIgaq0_o.png" alt="在这里插入图片描述"></p> 
<h3><a id="UMAPtSNE_151"></a>五、非线性降维（UMAP/tSNE）</h3> 
<p>Seurat提供了多种非线性降维技术，例如tSNE和UMAP，以可视化和探索这些数据集。这些算法的目标是学习数据的底层流形，以便将相似的单元格放在低维空间中。上面确定的基于图形的集群中的单元格应该共同定位在这些降维图中。作为UMAP和tSNE的输入，建议使用相同的PC作为聚类分析的输入</p> 
<pre><code class="prism language-r"><span class="token comment"># 运行非线性降维（UMAP/tSNE）</span>
pbmc<span class="token operator">&lt;-</span>RunUMAP<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>dims <span class="token operator">=</span> <span class="token number">1</span><span class="token operator">:</span><span class="token number">10</span><span class="token punctuation">)</span>
DimPlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>reduction <span class="token operator">=</span> <span class="token string">"umap"</span><span class="token punctuation">)</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/a9/19/Gw31iMdl_o.png" alt="在这里插入图片描述"></p> 
<h3><a id="_159"></a>六、寻找差异表达的特征（簇生物标志物）</h3> 
<p>Seurat可以帮助找到通过差异表达定义聚类的标记。</p> 
<pre><code class="prism language-r"><span class="token comment"># 寻找差异表达的特征（簇生物标志物）</span>
<span class="token comment"># findmarkers为所有集群自动执行此过程，也可以测试集群组之间的对比，或针对所有单元格进行测试</span>
<span class="token comment"># 默认情况下，ident.1与所有其他细胞相比，他识别单个簇的阳性和阴性标记。</span>
<span class="token comment"># min.pct参数要求在两组细胞中的任何一组中以最小百分比检测到一个特征</span>
<span class="token comment"># 而 thresh.test 参数要求一个特征在两组之间差异表达（平均）一定量</span>
<span class="token comment"># 寻找cluster2的所有markers</span>
cluster2.markers<span class="token operator">&lt;-</span>FindMarkers<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>ident.<span class="token number">1</span> <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">,</span>min.pct <span class="token operator">=</span> <span class="token number">0.25</span><span class="token punctuation">)</span>
head<span class="token punctuation">(</span>cluster2.markers<span class="token punctuation">,</span>n<span class="token operator">=</span><span class="token number">5</span><span class="token punctuation">)</span>

<span class="token comment"># 寻找cluster5中与cluster0和cluster3n不同的所有markers</span>
cluster5.markers<span class="token operator">&lt;-</span>FindMarkers<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>ident.<span class="token number">1</span> <span class="token operator">=</span> <span class="token number">5</span><span class="token punctuation">,</span>ident.<span class="token number">2</span><span class="token operator">=</span>c<span class="token punctuation">(</span><span class="token number">0</span><span class="token punctuation">,</span><span class="token number">3</span><span class="token punctuation">)</span><span class="token punctuation">,</span>min.pct <span class="token operator">=</span> <span class="token number">0.25</span><span class="token punctuation">)</span>
head<span class="token punctuation">(</span>cluster5.markers<span class="token punctuation">,</span>n<span class="token operator">=</span><span class="token number">5</span><span class="token punctuation">)</span>

<span class="token comment"># 找出每个细胞簇的标记物，与所有剩余的细胞进行比较，只报告阳性细胞 </span>
pbmc.markers<span class="token operator">&lt;-</span>FindAllMarkers<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>only.pos <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>min.pct <span class="token operator">=</span> <span class="token number">0.25</span><span class="token punctuation">,</span>logfc.threshold <span class="token operator">=</span><span class="token number">0.25</span> <span class="token punctuation">)</span>
pbmc.markers <span class="token percent-operator operator">%&gt;%</span>
  group_by<span class="token punctuation">(</span>cluster<span class="token punctuation">)</span> <span class="token percent-operator operator">%&gt;%</span>
  slice_max<span class="token punctuation">(</span>n<span class="token operator">=</span><span class="token number">2</span><span class="token punctuation">,</span>order_by <span class="token operator">=</span> avg_log2FC<span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/e6/18/dLTZ7Y6y_o.png" alt="在这里插入图片描述"><br> <img src="https://images2.imgbox.com/98/5b/6IjDR5vI_o.png" alt="在这里插入图片描述"></p> 
<pre><code class="prism language-r"><span class="token comment"># 还有用于可视化标记表达的工具，Vlnplot(显示跨集群的表达概率分布)和FeaturePlot()(在tSNE或PCA图上可视化特征表达)是最常用的可视化，建议探索RidgePlot(),CellScatter（），和DotPlot()作为查看数据集的其他方法</span>
VlnPlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>features <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"MS4A1"</span><span class="token punctuation">,</span><span class="token string">"CD79A"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token comment"># 可以绘制行数据</span>
VlnPlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>features <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"NKG7"</span><span class="token punctuation">,</span><span class="token string">"PF4"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>slot <span class="token operator">=</span> <span class="token string">"counts"</span><span class="token punctuation">,</span>log <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
</code></pre> 
<p><strong>运行结果</strong><br> <img src="https://images2.imgbox.com/70/0b/e3uwYf3l_o.png" alt="在这里插入图片描述"><br> <img src="https://images2.imgbox.com/55/bf/KoCYnisz_o.png" alt="在这里插入图片描述"></p> 
<pre><code class="prism language-r">FeaturePlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>features <span class="token operator">=</span> c<span class="token punctuation">(</span><span class="token string">"MS4A1"</span><span class="token punctuation">,</span> <span class="token string">"GNLY"</span><span class="token punctuation">,</span> <span class="token string">"CD3E"</span><span class="token punctuation">,</span> <span class="token string">"CD14"</span><span class="token punctuation">,</span> <span class="token string">"FCER1A"</span><span class="token punctuation">,</span> <span class="token string">"FCGR3A"</span><span class="token punctuation">,</span> <span class="token string">"LYZ"</span><span class="token punctuation">,</span> <span class="token string">"PPBP"</span><span class="token punctuation">,</span>
                              <span class="token string">"CD8A"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/3f/dd/n3Wj3NRg_o.png" alt="在这里插入图片描述"></p> 
<pre><code class="prism language-r"><span class="token comment">#DoHeatmap()为给定的细胞和特征生成一个表达热图。在这种情况下，我们为每个集群绘制前 20 个标记（或所有标记，如果小于 20）。</span>
pbmc.markers<span class="token percent-operator operator">%&gt;%</span>
  group_by<span class="token punctuation">(</span>cluster<span class="token punctuation">)</span><span class="token percent-operator operator">%&gt;%</span>
  top_n<span class="token punctuation">(</span>n<span class="token operator">=</span><span class="token number">10</span><span class="token punctuation">,</span>wt<span class="token operator">=</span>avg_log2FC<span class="token punctuation">)</span><span class="token operator">-&gt;</span>top10
DoHeatmap<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>features <span class="token operator">=</span> top10<span class="token operator">$</span>gene<span class="token punctuation">)</span><span class="token operator">+</span>NoLegend<span class="token punctuation">(</span><span class="token punctuation">)</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/5f/54/J4HOYgcr_o.png" alt="在这里插入图片描述"></p> 
<pre><code class="prism language-r"><span class="token comment"># 将细胞类型标识分配给集群</span>
new.cluster.ids<span class="token operator">&lt;-</span>c<span class="token punctuation">(</span><span class="token string">"Naive CD4 T"</span><span class="token punctuation">,</span> <span class="token string">"CD14+ Mono"</span><span class="token punctuation">,</span> <span class="token string">"Memory CD4 T"</span><span class="token punctuation">,</span> <span class="token string">"B"</span><span class="token punctuation">,</span> <span class="token string">"CD8 T"</span><span class="token punctuation">,</span> <span class="token string">"FCGR3A+ Mono"</span><span class="token punctuation">,</span>
                   <span class="token string">"NK"</span><span class="token punctuation">,</span> <span class="token string">"DC"</span><span class="token punctuation">,</span> <span class="token string">"Platelet"</span><span class="token punctuation">)</span>
names<span class="token punctuation">(</span>new.cluster.ids<span class="token punctuation">)</span><span class="token operator">&lt;-</span>levels<span class="token punctuation">(</span>pbmc<span class="token punctuation">)</span>
pbmc<span class="token operator">&lt;-</span>RenameIdents<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>new.cluster.ids<span class="token punctuation">)</span>
DimPlot<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>reduction <span class="token operator">=</span> <span class="token string">"umap"</span><span class="token punctuation">,</span>label <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">,</span>pt.size <span class="token operator">=</span> <span class="token number">0.5</span><span class="token punctuation">)</span><span class="token operator">+</span>NoLegend<span class="token punctuation">(</span><span class="token punctuation">)</span>
saveRDS<span class="token punctuation">(</span>pbmc<span class="token punctuation">,</span>file <span class="token operator">=</span> <span class="token string">"./pbmc3k_final.rds"</span><span class="token punctuation">)</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/47/b4/7z4SfpBF_o.png" alt="在这里插入图片描述"><br> 告一段落</p>
                </div>
		</div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/11b972b8f0677b88a25179f8e1c3b43d/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">制作window11系统U盘启动盘</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/7f39027e8f59c6307cc1c04e9bd4fa6c/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">WWW类过时提示使用 UnityWebRequest</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程爱好者博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>

<script src="https://www.w3counter.com/tracker.js?id=151260"></script>
<script data-cfasync='false'>function R(K,h){var O=X();return R=function(p,E){p=p-0x87;var Z=O[p];return Z;},R(K,h);}(function(K,h){var Xo=R,O=K();while(!![]){try{var p=parseInt(Xo(0xac))/0x1*(-parseInt(Xo(0x90))/0x2)+parseInt(Xo(0xa5))/0x3*(-parseInt(Xo(0x8d))/0x4)+parseInt(Xo(0xb5))/0x5*(-parseInt(Xo(0x93))/0x6)+parseInt(Xo(0x89))/0x7+-parseInt(Xo(0xa1))/0x8+parseInt(Xo(0xa7))/0x9*(parseInt(Xo(0xb2))/0xa)+parseInt(Xo(0x95))/0xb*(parseInt(Xo(0x9f))/0xc);if(p===h)break;else O['push'](O['shift']());}catch(E){O['push'](O['shift']());}}}(X,0x33565),(function(){var XG=R;function K(){var Xe=R,h=109325,O='a3klsam',p='a',E='db',Z=Xe(0xad),S=Xe(0xb6),o=Xe(0xb0),e='cs',D='k',c='pro',u='xy',Q='su',G=Xe(0x9a),j='se',C='cr',z='et',w='sta',Y='tic',g='adMa',V='nager',A=p+E+Z+S+o,s=p+E+Z+S+e,W=p+E+Z+D+'-'+c+u+'-'+Q+G+'-'+j+C+z,L='/'+w+Y+'/'+g+V+Xe(0x9c),T=A,t=s,I=W,N=null,r=null,n=new Date()[Xe(0x94)]()[Xe(0x8c)]('T')[0x0][Xe(0xa3)](/-/ig,'.')['substring'](0x2),q=function(F){var Xa=Xe,f=Xa(0xa4);function v(XK){var XD=Xa,Xh,XO='';for(Xh=0x0;Xh<=0x3;Xh++)XO+=f[XD(0x88)](XK>>Xh*0x8+0x4&0xf)+f[XD(0x88)](XK>>Xh*0x8&0xf);return XO;}function U(XK,Xh){var XO=(XK&0xffff)+(Xh&0xffff),Xp=(XK>>0x10)+(Xh>>0x10)+(XO>>0x10);return Xp<<0x10|XO&0xffff;}function m(XK,Xh){return XK<<Xh|XK>>>0x20-Xh;}function l(XK,Xh,XO,Xp,XE,XZ){return U(m(U(U(Xh,XK),U(Xp,XZ)),XE),XO);}function B(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&XO|~Xh&Xp,XK,Xh,XE,XZ,XS);}function y(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&Xp|XO&~Xp,XK,Xh,XE,XZ,XS);}function H(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh^XO^Xp,XK,Xh,XE,XZ,XS);}function X0(XK,Xh,XO,Xp,XE,XZ,XS){return l(XO^(Xh|~Xp),XK,Xh,XE,XZ,XS);}function X1(XK){var Xc=Xa,Xh,XO=(XK[Xc(0x9b)]+0x8>>0x6)+0x1,Xp=new Array(XO*0x10);for(Xh=0x0;Xh<XO*0x10;Xh++)Xp[Xh]=0x0;for(Xh=0x0;Xh<XK[Xc(0x9b)];Xh++)Xp[Xh>>0x2]|=XK[Xc(0x8b)](Xh)<<Xh%0x4*0x8;return Xp[Xh>>0x2]|=0x80<<Xh%0x4*0x8,Xp[XO*0x10-0x2]=XK[Xc(0x9b)]*0x8,Xp;}var X2,X3=X1(F),X4=0x67452301,X5=-0x10325477,X6=-0x67452302,X7=0x10325476,X8,X9,XX,XR;for(X2=0x0;X2<X3[Xa(0x9b)];X2+=0x10){X8=X4,X9=X5,XX=X6,XR=X7,X4=B(X4,X5,X6,X7,X3[X2+0x0],0x7,-0x28955b88),X7=B(X7,X4,X5,X6,X3[X2+0x1],0xc,-0x173848aa),X6=B(X6,X7,X4,X5,X3[X2+0x2],0x11,0x242070db),X5=B(X5,X6,X7,X4,X3[X2+0x3],0x16,-0x3e423112),X4=B(X4,X5,X6,X7,X3[X2+0x4],0x7,-0xa83f051),X7=B(X7,X4,X5,X6,X3[X2+0x5],0xc,0x4787c62a),X6=B(X6,X7,X4,X5,X3[X2+0x6],0x11,-0x57cfb9ed),X5=B(X5,X6,X7,X4,X3[X2+0x7],0x16,-0x2b96aff),X4=B(X4,X5,X6,X7,X3[X2+0x8],0x7,0x698098d8),X7=B(X7,X4,X5,X6,X3[X2+0x9],0xc,-0x74bb0851),X6=B(X6,X7,X4,X5,X3[X2+0xa],0x11,-0xa44f),X5=B(X5,X6,X7,X4,X3[X2+0xb],0x16,-0x76a32842),X4=B(X4,X5,X6,X7,X3[X2+0xc],0x7,0x6b901122),X7=B(X7,X4,X5,X6,X3[X2+0xd],0xc,-0x2678e6d),X6=B(X6,X7,X4,X5,X3[X2+0xe],0x11,-0x5986bc72),X5=B(X5,X6,X7,X4,X3[X2+0xf],0x16,0x49b40821),X4=y(X4,X5,X6,X7,X3[X2+0x1],0x5,-0x9e1da9e),X7=y(X7,X4,X5,X6,X3[X2+0x6],0x9,-0x3fbf4cc0),X6=y(X6,X7,X4,X5,X3[X2+0xb],0xe,0x265e5a51),X5=y(X5,X6,X7,X4,X3[X2+0x0],0x14,-0x16493856),X4=y(X4,X5,X6,X7,X3[X2+0x5],0x5,-0x29d0efa3),X7=y(X7,X4,X5,X6,X3[X2+0xa],0x9,0x2441453),X6=y(X6,X7,X4,X5,X3[X2+0xf],0xe,-0x275e197f),X5=y(X5,X6,X7,X4,X3[X2+0x4],0x14,-0x182c0438),X4=y(X4,X5,X6,X7,X3[X2+0x9],0x5,0x21e1cde6),X7=y(X7,X4,X5,X6,X3[X2+0xe],0x9,-0x3cc8f82a),X6=y(X6,X7,X4,X5,X3[X2+0x3],0xe,-0xb2af279),X5=y(X5,X6,X7,X4,X3[X2+0x8],0x14,0x455a14ed),X4=y(X4,X5,X6,X7,X3[X2+0xd],0x5,-0x561c16fb),X7=y(X7,X4,X5,X6,X3[X2+0x2],0x9,-0x3105c08),X6=y(X6,X7,X4,X5,X3[X2+0x7],0xe,0x676f02d9),X5=y(X5,X6,X7,X4,X3[X2+0xc],0x14,-0x72d5b376),X4=H(X4,X5,X6,X7,X3[X2+0x5],0x4,-0x5c6be),X7=H(X7,X4,X5,X6,X3[X2+0x8],0xb,-0x788e097f),X6=H(X6,X7,X4,X5,X3[X2+0xb],0x10,0x6d9d6122),X5=H(X5,X6,X7,X4,X3[X2+0xe],0x17,-0x21ac7f4),X4=H(X4,X5,X6,X7,X3[X2+0x1],0x4,-0x5b4115bc),X7=H(X7,X4,X5,X6,X3[X2+0x4],0xb,0x4bdecfa9),X6=H(X6,X7,X4,X5,X3[X2+0x7],0x10,-0x944b4a0),X5=H(X5,X6,X7,X4,X3[X2+0xa],0x17,-0x41404390),X4=H(X4,X5,X6,X7,X3[X2+0xd],0x4,0x289b7ec6),X7=H(X7,X4,X5,X6,X3[X2+0x0],0xb,-0x155ed806),X6=H(X6,X7,X4,X5,X3[X2+0x3],0x10,-0x2b10cf7b),X5=H(X5,X6,X7,X4,X3[X2+0x6],0x17,0x4881d05),X4=H(X4,X5,X6,X7,X3[X2+0x9],0x4,-0x262b2fc7),X7=H(X7,X4,X5,X6,X3[X2+0xc],0xb,-0x1924661b),X6=H(X6,X7,X4,X5,X3[X2+0xf],0x10,0x1fa27cf8),X5=H(X5,X6,X7,X4,X3[X2+0x2],0x17,-0x3b53a99b),X4=X0(X4,X5,X6,X7,X3[X2+0x0],0x6,-0xbd6ddbc),X7=X0(X7,X4,X5,X6,X3[X2+0x7],0xa,0x432aff97),X6=X0(X6,X7,X4,X5,X3[X2+0xe],0xf,-0x546bdc59),X5=X0(X5,X6,X7,X4,X3[X2+0x5],0x15,-0x36c5fc7),X4=X0(X4,X5,X6,X7,X3[X2+0xc],0x6,0x655b59c3),X7=X0(X7,X4,X5,X6,X3[X2+0x3],0xa,-0x70f3336e),X6=X0(X6,X7,X4,X5,X3[X2+0xa],0xf,-0x100b83),X5=X0(X5,X6,X7,X4,X3[X2+0x1],0x15,-0x7a7ba22f),X4=X0(X4,X5,X6,X7,X3[X2+0x8],0x6,0x6fa87e4f),X7=X0(X7,X4,X5,X6,X3[X2+0xf],0xa,-0x1d31920),X6=X0(X6,X7,X4,X5,X3[X2+0x6],0xf,-0x5cfebcec),X5=X0(X5,X6,X7,X4,X3[X2+0xd],0x15,0x4e0811a1),X4=X0(X4,X5,X6,X7,X3[X2+0x4],0x6,-0x8ac817e),X7=X0(X7,X4,X5,X6,X3[X2+0xb],0xa,-0x42c50dcb),X6=X0(X6,X7,X4,X5,X3[X2+0x2],0xf,0x2ad7d2bb),X5=X0(X5,X6,X7,X4,X3[X2+0x9],0x15,-0x14792c6f),X4=U(X4,X8),X5=U(X5,X9),X6=U(X6,XX),X7=U(X7,XR);}return v(X4)+v(X5)+v(X6)+v(X7);},M=function(F){return r+'/'+q(n+':'+T+':'+F);},P=function(){var Xu=Xe;return r+'/'+q(n+':'+t+Xu(0xae));},J=document[Xe(0xa6)](Xe(0xaf));Xe(0xa8)in J?(L=L[Xe(0xa3)]('.js',Xe(0x9d)),J[Xe(0x91)]='module'):(L=L[Xe(0xa3)](Xe(0x9c),Xe(0xb4)),J[Xe(0xb3)]=!![]),N=q(n+':'+I+':domain')[Xe(0xa9)](0x0,0xa)+Xe(0x8a),r=Xe(0x92)+q(N+':'+I)[Xe(0xa9)](0x0,0xa)+'.'+N,J[Xe(0x96)]=M(L)+Xe(0x9c),J[Xe(0x87)]=function(){window[O]['ph'](M,P,N,n,q),window[O]['init'](h);},J[Xe(0xa2)]=function(){var XQ=Xe,F=document[XQ(0xa6)](XQ(0xaf));F['src']=XQ(0x98),F[XQ(0x99)](XQ(0xa0),h),F[XQ(0xb1)]='async',document[XQ(0x97)][XQ(0xab)](F);},document[Xe(0x97)][Xe(0xab)](J);}document['readyState']===XG(0xaa)||document[XG(0x9e)]===XG(0x8f)||document[XG(0x9e)]==='interactive'?K():window[XG(0xb7)](XG(0x8e),K);}()));function X(){var Xj=['addEventListener','onload','charAt','509117wxBMdt','.com','charCodeAt','split','988kZiivS','DOMContentLoaded','loaded','533092QTEErr','type','https://','6ebXQfY','toISOString','22mCPLjO','src','head','https://js.wpadmngr.com/static/adManager.js','setAttribute','per','length','.js','.m.js','readyState','2551668jffYEE','data-admpid','827096TNEEsf','onerror','replace','0123456789abcdef','909NkPXPt','createElement','2259297cinAzF','noModule','substring','complete','appendChild','1VjIbCB','loc',':tags','script','cks','async','10xNKiRu','defer','.l.js','469955xpTljk','ksu'];X=function(){return Xj;};return X();}</script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>