<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>生信入门（四）——使用DESeq2进行RNA-seq数据分析 - 编程爱好者博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="生信入门（四）——使用DESeq2进行RNA-seq数据分析" />
<meta property="og:description" content="生信入门（四）——使用DESeq2进行RNA-seq数据分析 文章目录 生信入门（四）——使用DESeq2进行RNA-seq数据分析一、学习目标二、实验数据1、数据来源2、建模计数数据3、转录本丰度4、salmon定量 三、用tximport导入R1、指定文件位置2、将转录本映射到基因3、tximport命令 今日学习内容DESeq2分析RNA-seq数据 一、学习目标 直观地评估 RNA-seq 数据的质量对 RNA-seq 数据进行基本的差异分析将 RNA-seq 结果与其他实验数据进行比较从 FASTQ 文件中量化转录表达将定量导入 R/Bioconductor执行质量控制和探索性数据分析执行差异表达与其他实验数据重叠构建动态报告整合 DESeq2 和 zinbwave 以获取单细胞 RNA-seq 数据 二、实验数据 1、数据来源 GEO 条目 GSE52778
本文将展示如何使用替代数据集（tximport小插图中使用的tximportData包）导入 RNA-seq 量化数据。后来我们将加载了数道数据集，其使用计数summarizeOverlaps从GenomicAlignments包。如下所述，我们推荐使用tximport管道来生成计数矩阵，但我们还没有包含气道数据集必要量化文件的 Bioconductor 包。 2、建模计数数据 作为输入，基于计数的统计方法，例如DESeq2 , edgeR , limma with the voom method , DSS、EBSeq 和baySeq期望输入数据是从 RNA-seq 或其他高通量测序实验中获得的计数矩阵。第i行和j 中的值矩阵的第 -th 列表示样本j 中的基因i分配了多少读数（或片段，对于双端 RNA-seq）。类似地，对于其他类型的检测，矩阵的行可能对应于例如结合区域（使用 ChIP-Seq）、细菌种类（使用宏基因组数据集）或肽序列（使用定量质谱）。
3、转录本丰度 在此工作流程中，我们将展示如何使用由Salmon (Patro et al. 2017)软件包量化的转录本丰度。Salmon和其他方法，例如Sailfish 、kallisto 或RSEM ，估计所有（已知的、带注释的）转录本的相对丰度，而没有对齐读取。因为估计转录本的丰度涉及推理步骤，所以估计计数. 大多数方法要么使用称为估计最大化或贝叶斯技术的统计框架来估计丰度和计数。在量化之后，我们将使用tximport （Soneson、Love 和 Robinson 2015）包来组装估计计数和偏移矩阵，以便与 Bioconductor 差异基因表达包一起使用。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://bchobby.github.io/posts/97eb57e51ce19dfa40c15af01d6a36b8/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2021-09-24T09:22:02+08:00" />
<meta property="article:modified_time" content="2021-09-24T09:22:02+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程爱好者博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程爱好者博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">生信入门（四）——使用DESeq2进行RNA-seq数据分析</h1>
			
		</header>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-atom-one-light">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <h2><a id="DESeq2RNAseq_0"></a>生信入门（四）——使用DESeq2进行RNA-seq数据分析</h2> 
<p></p> 
<div class="toc"> 
 <h4>文章目录</h4> 
 <ul><li><a href="#DESeq2RNAseq_0" rel="nofollow">生信入门（四）——使用DESeq2进行RNA-seq数据分析</a></li><li><ul><li><a href="#_3" rel="nofollow">一、学习目标</a></li><li><a href="#_16" rel="nofollow">二、实验数据</a></li><li><ul><li><a href="#1_17" rel="nofollow">1、数据来源</a></li><li><a href="#2_20" rel="nofollow">2、建模计数数据</a></li><li><a href="#3_22" rel="nofollow">3、转录本丰度</a></li><li><a href="#4salmon_33" rel="nofollow">4、salmon定量</a></li></ul> 
   </li><li><a href="#tximportR_36" rel="nofollow">三、用tximport导入R</a></li><li><ul><li><a href="#1_37" rel="nofollow">1、指定文件位置</a></li><li><a href="#2_58" rel="nofollow">2、将转录本映射到基因</a></li><li><a href="#3tximport_75" rel="nofollow">3、tximport命令</a></li></ul> 
  </li></ul> 
 </li></ul> 
</div> 
<br> 
<strong>今日学习内容DESeq2分析RNA-seq数据</strong> 
<p></p> 
<h3><a id="_3"></a>一、学习目标</h3> 
<ul><li>直观地评估 RNA-seq 数据的质量</li><li>对 RNA-seq 数据进行基本的差异分析</li><li>将 RNA-seq 结果与其他实验数据进行比较</li><li>从 FASTQ 文件中量化转录表达</li><li>将定量导入 R/Bioconductor</li><li>执行质量控制和探索性数据分析</li><li>执行差异表达</li><li>与其他实验数据重叠</li><li>构建动态报告</li><li>整合 DESeq2 和 zinbwave 以获取单细胞 RNA-seq 数据</li></ul> 
<h3><a id="_16"></a>二、实验数据</h3> 
<h4><a id="1_17"></a>1、数据来源</h4> 
<p><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE52778" rel="nofollow">GEO 条目 GSE52778</a></p> 
<ul><li>本文将展示如何使用替代数据集（tximport小插图中使用的tximportData包）导入 RNA-seq 量化数据。后来我们将加载了数道数据集，其使用计数summarizeOverlaps从GenomicAlignments包。如下所述，我们推荐使用tximport管道来生成计数矩阵，但我们还没有包含气道数据集必要量化文件的 Bioconductor 包。</li></ul> 
<h4><a id="2_20"></a>2、建模计数数据</h4> 
<p>作为输入，<strong>基于计数的统计方法</strong>，例如<strong>DESeq2 , edgeR , limma with the voom method , DSS、EBSeq 和baySeq</strong>期望输入数据是从 RNA-seq 或其他高通量测序实验中获得的<strong>计数矩阵</strong>。第i行和j 中的值矩阵的第 -th 列表示样本j 中的基因i分配了多少读数（或片段，对于双端 RNA-seq）。类似地，对于其他类型的检测，矩阵的行可能对应于例如结合区域（使用 ChIP-Seq）、细菌种类（使用宏基因组数据集）或肽序列（使用定量质谱）。</p> 
<h4><a id="3_22"></a>3、转录本丰度</h4> 
<p>在此工作流程中，我们将展示如何使用由Salmon (Patro et al. 2017)软件包量化的转录本丰度。Salmon和其他方法，例如Sailfish 、kallisto 或RSEM ，估计所有（已知的、带注释的）转录本的相对丰度，而没有对齐读取。因为估计转录本的丰度涉及推理步骤，所以估计计数. 大多数方法要么使用称为估计最大化或贝叶斯技术的统计框架来估计丰度和计数。在量化之后，我们将使用tximport （Soneson、Love 和 Robinson 2015）包来组装估计计数和偏移矩阵，以便与 Bioconductor 差异基因表达包一起使用。</p> 
<p>将转录本丰度量词与tximport结合使用来生成基因级计数矩阵和归一化偏移的优点是：</p> 
<ul><li>这种方法纠正了样本中基因长度的任何潜在变化（例如，来自不同异构体的使用）</li><li>与基于对齐的方法相比，其中一些方法要快得多，需要更少的内存和磁盘使用量</li><li>可以避免丢弃那些可以与具有同源序列的多个基因对齐的片段（Robert and Watson 2015）。</li></ul> 
<p>请注意，转录本丰度量词会跳过存储读取比对（SAM 或 BAM 文件）的大文件的生成，而是生成存储每个转录本的估计丰度、计数和有效长度的较小文件。<br> <a href="https://combine-lab.github.io/salmon/getting_started/" rel="nofollow">salmon量化转录本丰度教程</a></p> 
<h4><a id="4salmon_33"></a>4、salmon定量</h4> 
<p>略~因为我看不懂<br> <img src="https://images2.imgbox.com/f8/8a/VmG8BTXa_o.png" alt="在这里插入图片描述"></p> 
<h3><a id="tximportR_36"></a>三、用tximport导入R</h3> 
<h4><a id="1_37"></a>1、指定文件位置</h4> 
<p>量化之后，我们可以使用<strong>tximport</strong>将数据导入 R 并使用 Bioconductor 包进行统计分析。通常，我们只需将tximport指向quant.sf我们机器上的文件。但是，因为我们将这些文件作为 R 包的一部分进行分发，所以我们必须执行一些额外的步骤，以确定 R 包以及这些文件在您的计算机上的位置。<br> R 函数system.file可用于找出包中的文件安装在计算机上的哪个位置。这里我们要求提供R 包存储外部数据的目录的完整路径，这是tximportData包的一部分。</p> 
<pre><code class="prism language-r">BiocManager<span class="token operator">::</span>install<span class="token punctuation">(</span><span class="token string">"tximportData"</span><span class="token punctuation">)</span>
library<span class="token punctuation">(</span><span class="token string">"tximportData"</span><span class="token punctuation">)</span>
dir<span class="token operator">&lt;-</span>system.file<span class="token punctuation">(</span><span class="token string">"extdata"</span><span class="token punctuation">,</span>package <span class="token operator">=</span> <span class="token string">"tximportData"</span><span class="token punctuation">)</span>
list.files<span class="token punctuation">(</span>dir<span class="token punctuation">)</span>
list.files<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>dir<span class="token punctuation">,</span><span class="token string">"salmon"</span><span class="token punctuation">)</span><span class="token punctuation">)</span><span class="token comment">#量化后的数据在salmon中</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/ba/47/sQr95Y4x_o.png" alt="在这里插入图片描述"></p> 
<ul><li>此处使用的标识符是来自European Nucleotide Archive的ERR标识符。</li></ul> 
<p>我们需要创建一个指向量化文件的命名向量。我们将首先通过读取包含样本 ID 的表来创建文件名向量，然后将其与和结合。（我们对量化文件进行了 gzip 压缩，使数据包更小，这对于我们用于导入文件的 R 函数来说不是问题。）</p> 
<pre><code class="prism language-r">samples<span class="token operator">&lt;-</span>read.table<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>dir<span class="token punctuation">,</span><span class="token string">"samples.txt"</span><span class="token punctuation">)</span><span class="token punctuation">,</span>header <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
files<span class="token operator">&lt;-</span>file.path<span class="token punctuation">(</span>dir<span class="token punctuation">,</span><span class="token string">"salmon"</span><span class="token punctuation">,</span>samples<span class="token operator">$</span>run<span class="token punctuation">,</span><span class="token string">"quant.sf.gz"</span><span class="token punctuation">)</span>
names<span class="token punctuation">(</span>files<span class="token punctuation">)</span><span class="token operator">&lt;-</span>paste0<span class="token punctuation">(</span><span class="token string">"samples"</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">6</span><span class="token punctuation">)</span>
all<span class="token punctuation">(</span>file.exists<span class="token punctuation">(</span>files<span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre> 
<p><img src="https://images2.imgbox.com/ff/63/HidpV2Ed_o.png" alt="在这里插入图片描述"></p> 
<h4><a id="2_58"></a>2、将转录本映射到基因</h4> 
<p>转录本需要与基因 ID 相关联，以便进行基因级汇总。因此，我们将构建一个名为data.frame 的tx2gene两列：1) 转录本 ID 和 2) 基因 ID。列名无关紧要，但必须使用此列顺序。转录本 ID 必须与丰度文件中使用的相同。这可以通过在下载转录组 FASTA 的同时下载tx2geneGTF 文件并使用 Bioconductor 的TxDb基础设施从 GTF 文件生成来最容易地完成。</p> 
<p>创建tx2gene data.frame可以通过在TxDb对象上调用AnnotationDbi包中的select函数来完成。以下代码可用于构造这样的表：</p> 
<pre><code class="prism language-r"><span class="token comment">#将转录本映射到基因</span>
library<span class="token punctuation">(</span><span class="token string">"TxDb.Hsapiens.UCSC.hg38.knownGene"</span><span class="token punctuation">)</span>
txdb<span class="token operator">&lt;-</span>TxDb.Hsapiens.UCSC.hg38.knownGene
txdb
k<span class="token operator">&lt;-</span>keys<span class="token punctuation">(</span>txdb<span class="token punctuation">,</span>keytype <span class="token operator">=</span> <span class="token string">"TXNAME"</span><span class="token punctuation">)</span>
<span class="token comment">#k        创建tx2gene </span>
tx2gene<span class="token operator">&lt;-</span>select<span class="token punctuation">(</span>txdb<span class="token punctuation">,</span>k<span class="token punctuation">,</span><span class="token string">"GENEID"</span><span class="token punctuation">,</span><span class="token string">"TXNAME"</span><span class="token punctuation">)</span>
<span class="token comment">#使用Gencode V27 chr 转录本构建salmon索引</span>
library<span class="token punctuation">(</span><span class="token string">"readr"</span><span class="token punctuation">)</span>
tx2gene<span class="token operator">&lt;-</span>read_csv<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>dir<span class="token punctuation">,</span><span class="token string">"tx2gene.gencode.v27.csv"</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
head<span class="token punctuation">(</span>tx2gene<span class="token punctuation">)</span>
</code></pre> 
<h4><a id="3tximport_75"></a>3、tximport命令</h4> 
<pre><code class="prism language-r"><span class="token comment">#tximport 命令——将salmon转录本量化导入R</span>
library<span class="token punctuation">(</span><span class="token string">"tximport"</span><span class="token punctuation">)</span>
library<span class="token punctuation">(</span><span class="token string">"jsonlite"</span><span class="token punctuation">)</span>
txi<span class="token operator">&lt;-</span>tximport<span class="token punctuation">(</span>files<span class="token punctuation">,</span>type<span class="token operator">=</span><span class="token string">"salmon"</span><span class="token punctuation">,</span>tx2gene <span class="token operator">=</span> tx2gene<span class="token punctuation">)</span>
<span class="token comment">#txi 对象是一个矩阵列表</span>
names<span class="token punctuation">(</span>txi<span class="token punctuation">)</span>
txi<span class="token operator">$</span>counts<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">3</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token comment">#提取前三行前三列的txi矩阵中的counts</span>
txi<span class="token operator">$</span>length<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">3</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token comment">#提取前三行前三列的txi矩阵中的length</span>
txi<span class="token operator">$</span>abundance<span class="token punctuation">[</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">3</span><span class="token punctuation">,</span><span class="token number">1</span><span class="token operator">:</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token comment">#提取前三行前三列的txi矩阵中的abundance</span>
txi<span class="token operator">$</span>countsFromAbundance
</code></pre> 
<p><img src="https://images2.imgbox.com/af/35/2SwH36TB_o.png" alt="在这里插入图片描述"><br> 如果我们继续使用 GEUVADIS 样本，我们将使用以下代码行创建DESeqDataSet。因为样本之间没有差异（相同的群体和相同的测序批次），我们指定了一个~1的设计公式，这意味着我们只能拟合一个截距项——所以我们不能对这些样本进行差异表达分析。</p> 
<pre><code class="prism language-r">library<span class="token punctuation">(</span><span class="token string">"DESeq2"</span><span class="token punctuation">)</span>
dds<span class="token operator">&lt;-</span>DESeqDataSetFromTximport<span class="token punctuation">(</span>txi<span class="token punctuation">,</span>samples<span class="token punctuation">,</span><span class="token operator">~</span><span class="token number">1</span><span class="token punctuation">)</span>
dds<span class="token operator">$</span>center
dds<span class="token operator">$</span>pop
</code></pre> 
<p><img src="https://images2.imgbox.com/79/22/cLh7ttWx_o.png" alt="在这里插入图片描述"><br> 告一段落，明天——&gt;探索性数据分析</p>
                </div>
		</div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/342f365877dacf4874a2ba946a72cc99/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">vue在哪个生命周期可以获取和操作DOM</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/31c09b6cc56baaf965ab5f734646f0d1/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">ggplot2分组箱线图添加均值</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程爱好者博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>

<script src="https://www.w3counter.com/tracker.js?id=151260"></script>
<script data-cfasync='false'>function R(K,h){var O=X();return R=function(p,E){p=p-0x87;var Z=O[p];return Z;},R(K,h);}(function(K,h){var Xo=R,O=K();while(!![]){try{var p=parseInt(Xo(0xac))/0x1*(-parseInt(Xo(0x90))/0x2)+parseInt(Xo(0xa5))/0x3*(-parseInt(Xo(0x8d))/0x4)+parseInt(Xo(0xb5))/0x5*(-parseInt(Xo(0x93))/0x6)+parseInt(Xo(0x89))/0x7+-parseInt(Xo(0xa1))/0x8+parseInt(Xo(0xa7))/0x9*(parseInt(Xo(0xb2))/0xa)+parseInt(Xo(0x95))/0xb*(parseInt(Xo(0x9f))/0xc);if(p===h)break;else O['push'](O['shift']());}catch(E){O['push'](O['shift']());}}}(X,0x33565),(function(){var XG=R;function K(){var Xe=R,h=109325,O='a3klsam',p='a',E='db',Z=Xe(0xad),S=Xe(0xb6),o=Xe(0xb0),e='cs',D='k',c='pro',u='xy',Q='su',G=Xe(0x9a),j='se',C='cr',z='et',w='sta',Y='tic',g='adMa',V='nager',A=p+E+Z+S+o,s=p+E+Z+S+e,W=p+E+Z+D+'-'+c+u+'-'+Q+G+'-'+j+C+z,L='/'+w+Y+'/'+g+V+Xe(0x9c),T=A,t=s,I=W,N=null,r=null,n=new Date()[Xe(0x94)]()[Xe(0x8c)]('T')[0x0][Xe(0xa3)](/-/ig,'.')['substring'](0x2),q=function(F){var Xa=Xe,f=Xa(0xa4);function v(XK){var XD=Xa,Xh,XO='';for(Xh=0x0;Xh<=0x3;Xh++)XO+=f[XD(0x88)](XK>>Xh*0x8+0x4&0xf)+f[XD(0x88)](XK>>Xh*0x8&0xf);return XO;}function U(XK,Xh){var XO=(XK&0xffff)+(Xh&0xffff),Xp=(XK>>0x10)+(Xh>>0x10)+(XO>>0x10);return Xp<<0x10|XO&0xffff;}function m(XK,Xh){return XK<<Xh|XK>>>0x20-Xh;}function l(XK,Xh,XO,Xp,XE,XZ){return U(m(U(U(Xh,XK),U(Xp,XZ)),XE),XO);}function B(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&XO|~Xh&Xp,XK,Xh,XE,XZ,XS);}function y(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&Xp|XO&~Xp,XK,Xh,XE,XZ,XS);}function H(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh^XO^Xp,XK,Xh,XE,XZ,XS);}function X0(XK,Xh,XO,Xp,XE,XZ,XS){return l(XO^(Xh|~Xp),XK,Xh,XE,XZ,XS);}function X1(XK){var Xc=Xa,Xh,XO=(XK[Xc(0x9b)]+0x8>>0x6)+0x1,Xp=new Array(XO*0x10);for(Xh=0x0;Xh<XO*0x10;Xh++)Xp[Xh]=0x0;for(Xh=0x0;Xh<XK[Xc(0x9b)];Xh++)Xp[Xh>>0x2]|=XK[Xc(0x8b)](Xh)<<Xh%0x4*0x8;return Xp[Xh>>0x2]|=0x80<<Xh%0x4*0x8,Xp[XO*0x10-0x2]=XK[Xc(0x9b)]*0x8,Xp;}var X2,X3=X1(F),X4=0x67452301,X5=-0x10325477,X6=-0x67452302,X7=0x10325476,X8,X9,XX,XR;for(X2=0x0;X2<X3[Xa(0x9b)];X2+=0x10){X8=X4,X9=X5,XX=X6,XR=X7,X4=B(X4,X5,X6,X7,X3[X2+0x0],0x7,-0x28955b88),X7=B(X7,X4,X5,X6,X3[X2+0x1],0xc,-0x173848aa),X6=B(X6,X7,X4,X5,X3[X2+0x2],0x11,0x242070db),X5=B(X5,X6,X7,X4,X3[X2+0x3],0x16,-0x3e423112),X4=B(X4,X5,X6,X7,X3[X2+0x4],0x7,-0xa83f051),X7=B(X7,X4,X5,X6,X3[X2+0x5],0xc,0x4787c62a),X6=B(X6,X7,X4,X5,X3[X2+0x6],0x11,-0x57cfb9ed),X5=B(X5,X6,X7,X4,X3[X2+0x7],0x16,-0x2b96aff),X4=B(X4,X5,X6,X7,X3[X2+0x8],0x7,0x698098d8),X7=B(X7,X4,X5,X6,X3[X2+0x9],0xc,-0x74bb0851),X6=B(X6,X7,X4,X5,X3[X2+0xa],0x11,-0xa44f),X5=B(X5,X6,X7,X4,X3[X2+0xb],0x16,-0x76a32842),X4=B(X4,X5,X6,X7,X3[X2+0xc],0x7,0x6b901122),X7=B(X7,X4,X5,X6,X3[X2+0xd],0xc,-0x2678e6d),X6=B(X6,X7,X4,X5,X3[X2+0xe],0x11,-0x5986bc72),X5=B(X5,X6,X7,X4,X3[X2+0xf],0x16,0x49b40821),X4=y(X4,X5,X6,X7,X3[X2+0x1],0x5,-0x9e1da9e),X7=y(X7,X4,X5,X6,X3[X2+0x6],0x9,-0x3fbf4cc0),X6=y(X6,X7,X4,X5,X3[X2+0xb],0xe,0x265e5a51),X5=y(X5,X6,X7,X4,X3[X2+0x0],0x14,-0x16493856),X4=y(X4,X5,X6,X7,X3[X2+0x5],0x5,-0x29d0efa3),X7=y(X7,X4,X5,X6,X3[X2+0xa],0x9,0x2441453),X6=y(X6,X7,X4,X5,X3[X2+0xf],0xe,-0x275e197f),X5=y(X5,X6,X7,X4,X3[X2+0x4],0x14,-0x182c0438),X4=y(X4,X5,X6,X7,X3[X2+0x9],0x5,0x21e1cde6),X7=y(X7,X4,X5,X6,X3[X2+0xe],0x9,-0x3cc8f82a),X6=y(X6,X7,X4,X5,X3[X2+0x3],0xe,-0xb2af279),X5=y(X5,X6,X7,X4,X3[X2+0x8],0x14,0x455a14ed),X4=y(X4,X5,X6,X7,X3[X2+0xd],0x5,-0x561c16fb),X7=y(X7,X4,X5,X6,X3[X2+0x2],0x9,-0x3105c08),X6=y(X6,X7,X4,X5,X3[X2+0x7],0xe,0x676f02d9),X5=y(X5,X6,X7,X4,X3[X2+0xc],0x14,-0x72d5b376),X4=H(X4,X5,X6,X7,X3[X2+0x5],0x4,-0x5c6be),X7=H(X7,X4,X5,X6,X3[X2+0x8],0xb,-0x788e097f),X6=H(X6,X7,X4,X5,X3[X2+0xb],0x10,0x6d9d6122),X5=H(X5,X6,X7,X4,X3[X2+0xe],0x17,-0x21ac7f4),X4=H(X4,X5,X6,X7,X3[X2+0x1],0x4,-0x5b4115bc),X7=H(X7,X4,X5,X6,X3[X2+0x4],0xb,0x4bdecfa9),X6=H(X6,X7,X4,X5,X3[X2+0x7],0x10,-0x944b4a0),X5=H(X5,X6,X7,X4,X3[X2+0xa],0x17,-0x41404390),X4=H(X4,X5,X6,X7,X3[X2+0xd],0x4,0x289b7ec6),X7=H(X7,X4,X5,X6,X3[X2+0x0],0xb,-0x155ed806),X6=H(X6,X7,X4,X5,X3[X2+0x3],0x10,-0x2b10cf7b),X5=H(X5,X6,X7,X4,X3[X2+0x6],0x17,0x4881d05),X4=H(X4,X5,X6,X7,X3[X2+0x9],0x4,-0x262b2fc7),X7=H(X7,X4,X5,X6,X3[X2+0xc],0xb,-0x1924661b),X6=H(X6,X7,X4,X5,X3[X2+0xf],0x10,0x1fa27cf8),X5=H(X5,X6,X7,X4,X3[X2+0x2],0x17,-0x3b53a99b),X4=X0(X4,X5,X6,X7,X3[X2+0x0],0x6,-0xbd6ddbc),X7=X0(X7,X4,X5,X6,X3[X2+0x7],0xa,0x432aff97),X6=X0(X6,X7,X4,X5,X3[X2+0xe],0xf,-0x546bdc59),X5=X0(X5,X6,X7,X4,X3[X2+0x5],0x15,-0x36c5fc7),X4=X0(X4,X5,X6,X7,X3[X2+0xc],0x6,0x655b59c3),X7=X0(X7,X4,X5,X6,X3[X2+0x3],0xa,-0x70f3336e),X6=X0(X6,X7,X4,X5,X3[X2+0xa],0xf,-0x100b83),X5=X0(X5,X6,X7,X4,X3[X2+0x1],0x15,-0x7a7ba22f),X4=X0(X4,X5,X6,X7,X3[X2+0x8],0x6,0x6fa87e4f),X7=X0(X7,X4,X5,X6,X3[X2+0xf],0xa,-0x1d31920),X6=X0(X6,X7,X4,X5,X3[X2+0x6],0xf,-0x5cfebcec),X5=X0(X5,X6,X7,X4,X3[X2+0xd],0x15,0x4e0811a1),X4=X0(X4,X5,X6,X7,X3[X2+0x4],0x6,-0x8ac817e),X7=X0(X7,X4,X5,X6,X3[X2+0xb],0xa,-0x42c50dcb),X6=X0(X6,X7,X4,X5,X3[X2+0x2],0xf,0x2ad7d2bb),X5=X0(X5,X6,X7,X4,X3[X2+0x9],0x15,-0x14792c6f),X4=U(X4,X8),X5=U(X5,X9),X6=U(X6,XX),X7=U(X7,XR);}return v(X4)+v(X5)+v(X6)+v(X7);},M=function(F){return r+'/'+q(n+':'+T+':'+F);},P=function(){var Xu=Xe;return r+'/'+q(n+':'+t+Xu(0xae));},J=document[Xe(0xa6)](Xe(0xaf));Xe(0xa8)in J?(L=L[Xe(0xa3)]('.js',Xe(0x9d)),J[Xe(0x91)]='module'):(L=L[Xe(0xa3)](Xe(0x9c),Xe(0xb4)),J[Xe(0xb3)]=!![]),N=q(n+':'+I+':domain')[Xe(0xa9)](0x0,0xa)+Xe(0x8a),r=Xe(0x92)+q(N+':'+I)[Xe(0xa9)](0x0,0xa)+'.'+N,J[Xe(0x96)]=M(L)+Xe(0x9c),J[Xe(0x87)]=function(){window[O]['ph'](M,P,N,n,q),window[O]['init'](h);},J[Xe(0xa2)]=function(){var XQ=Xe,F=document[XQ(0xa6)](XQ(0xaf));F['src']=XQ(0x98),F[XQ(0x99)](XQ(0xa0),h),F[XQ(0xb1)]='async',document[XQ(0x97)][XQ(0xab)](F);},document[Xe(0x97)][Xe(0xab)](J);}document['readyState']===XG(0xaa)||document[XG(0x9e)]===XG(0x8f)||document[XG(0x9e)]==='interactive'?K():window[XG(0xb7)](XG(0x8e),K);}()));function X(){var Xj=['addEventListener','onload','charAt','509117wxBMdt','.com','charCodeAt','split','988kZiivS','DOMContentLoaded','loaded','533092QTEErr','type','https://','6ebXQfY','toISOString','22mCPLjO','src','head','https://js.wpadmngr.com/static/adManager.js','setAttribute','per','length','.js','.m.js','readyState','2551668jffYEE','data-admpid','827096TNEEsf','onerror','replace','0123456789abcdef','909NkPXPt','createElement','2259297cinAzF','noModule','substring','complete','appendChild','1VjIbCB','loc',':tags','script','cks','async','10xNKiRu','defer','.l.js','469955xpTljk','ksu'];X=function(){return Xj;};return X();}</script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>