<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>Antismash-BigScape次级代谢基因簇鉴定 - 编程爱好者博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="Antismash-BigScape次级代谢基因簇鉴定" />
<meta property="og:description" content="文章目录 Antismash概述Install使用其他参数输出详细结果网络文件GCF and GCC交互式可视化网页版 其他详细参数 BigScape简介工作原理installusage输出输出文件夹结构结果 可能的问题Reference Antismash 概述 antiSMASH - the antibiotics and Secondary Metabolite Analysis SHell，是用来鉴定微生物基因组次级代谢物合成基因簇的软件。
临床上使用的大部分抗生素和药物均来自植物或微生物的天然产物。结合基因组挖掘的经典分离与分析法使得基于基因组的天然产物途径鉴定和描述更为方便。
一般情况下，参与次级代谢途径中生物合成酶基因在基因组上成簇排列，基于指定类型的HMM，antiSMASH数据库能准确鉴定所有已知的次级代谢簇。在antiSMATH中，将次级代谢簇分为24类。
Install conda create -n antismash -c bioconda antismash conda activate antismash (antismash) ok@ubuntu $ download-antismash-databases Creating checksum of Pfam-A.hmm PFAM file present and ok for version 32.0 Resfams database present and checked Creating checksum of TIGRFam.hmm TIGRFam database present and checked Creating checksum of proteins.fasta ClusterBlast fasta file present and checked Creating checksum of proteins." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://bchobby.github.io/posts/700df1b131a7dfdf74597c794173eca8/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-12-11T16:26:51+08:00" />
<meta property="article:modified_time" content="2023-12-11T16:26:51+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程爱好者博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程爱好者博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">Antismash-BigScape次级代谢基因簇鉴定</h1>
			
		</header>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-atom-one-light">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <p></p> 
<div class="toc"> 
 <h4>文章目录</h4> 
 <ul><li><a href="#Antismash_1" rel="nofollow">Antismash</a></li><li><ul><li><a href="#_2" rel="nofollow">概述</a></li><li><a href="#Install_6" rel="nofollow">Install</a></li><li><a href="#_38" rel="nofollow">使用</a></li><li><a href="#_69" rel="nofollow">其他参数</a></li><li><a href="#_87" rel="nofollow">输出详细结果</a></li><li><ul><li><a href="#_88" rel="nofollow">网络文件</a></li><li><a href="#GCF_and_GCC_97" rel="nofollow">GCF and GCC</a></li><li><a href="#_107" rel="nofollow">交互式可视化</a></li><li><a href="#_113" rel="nofollow">网页版</a></li></ul> 
   </li><li><a href="#_137" rel="nofollow">其他详细参数</a></li></ul> 
  </li><li><a href="#BigScape_200" rel="nofollow">BigScape</a></li><li><ul><li><a href="#_201" rel="nofollow">简介</a></li><li><a href="#_205" rel="nofollow">工作原理</a></li><li><a href="#install_219" rel="nofollow">install</a></li><li><a href="#usage_263" rel="nofollow">usage</a></li><li><a href="#_468" rel="nofollow">输出</a></li><li><ul><li><a href="#_469" rel="nofollow">输出文件夹结构</a></li><li><a href="#_486" rel="nofollow">结果</a></li></ul> 
  </li></ul> 
  </li><li><a href="#_551" rel="nofollow">可能的问题</a></li><li><a href="#Reference_599" rel="nofollow">Reference</a></li></ul> 
</div> 
<p></p> 
<h2><a id="Antismash_1"></a>Antismash</h2> 
<h3><a id="_2"></a>概述</h3> 
<p>antiSMASH - the antibiotics and Secondary Metabolite Analysis SHell，是用来鉴定微生物基因组次级代谢物合成基因簇的软件。<br> 临床上使用的大部分抗生素和药物均来自植物或微生物的天然产物。结合基因组挖掘的经典分离与分析法使得基于基因组的天然产物途径鉴定和描述更为方便。<br> 一般情况下，参与次级代谢途径中生物合成酶基因在基因组上成簇排列，基于指定类型的HMM，antiSMASH数据库能准确鉴定所有已知的次级代谢簇。在antiSMATH中，将次级代谢簇分为24类。</p> 
<h3><a id="Install_6"></a>Install</h3> 
<pre><code class="prism language-bash">conda create <span class="token parameter variable">-n</span> antismash  <span class="token parameter variable">-c</span> bioconda  antismash
conda activate antismash
<span class="token punctuation">(</span>antismash<span class="token punctuation">)</span> ok@ubuntu $ download-antismash-databases
Creating checksum of Pfam-A.hmm
PFAM <span class="token function">file</span> present and ok <span class="token keyword">for</span> version <span class="token number">32.0</span>
Resfams database present and checked
Creating checksum of TIGRFam.hmm
TIGRFam database present and checked
Creating checksum of proteins.fasta
ClusterBlast fasta <span class="token function">file</span> present and checked
Creating checksum of proteins.fasta
ClusterCompare mibig FASTA <span class="token function">file</span> present and checked
Pre-building all databases<span class="token punctuation">..</span>.
done.
<span class="token comment"># antiSMASH 6.0.1</span>
conda deactivate 
<span class="token comment"># 若出现 Could not remove or rename /home/anaconda3/pkgs/backports_abc-0.5-py_1*，手工删掉，尤其是.conda的文件，手工删掉</span>
<span class="token comment"># sudo rm -rf  /home/anaconda3/pkgs/backports_abc-0.5-py_1*</span>

<span class="token comment"># manually </span>
<span class="token comment">## First add the antiSMASH debian repository:</span>
<span class="token function">sudo</span> <span class="token function">apt-get</span> update
<span class="token function">sudo</span> <span class="token function">apt-get</span> <span class="token function">install</span> <span class="token parameter variable">-y</span> apt-transport-https
<span class="token function">sudo</span> <span class="token function">wget</span> http://dl.secondarymetabolites.org/antismash-stretch.list <span class="token parameter variable">-O</span> /etc/apt/sources.list.d/antismash.list
<span class="token function">sudo</span> <span class="token function">wget</span> <span class="token parameter variable">-q</span> -O- http://dl.secondarymetabolites.org/antismash.asc <span class="token operator">|</span> <span class="token function">sudo</span> apt-key <span class="token function">add</span> -
<span class="token function">sudo</span> <span class="token function">apt-get</span> update
<span class="token comment">## then install the binaries themselves</span>
<span class="token function">sudo</span> <span class="token function">apt-get</span> <span class="token function">install</span> hmmer2 hmmer diamond-aligner fasttree prodigal ncbi-blast+ muscle glimmerhmm
</code></pre> 
<h3><a id="_38"></a>使用</h3> 
<ul><li> <p>Fast run<br> 在没有参数的情况下运行antismash将运行核心检测模块和所有快速的簇特定分析步骤。更多耗时的选项，如ClusterBlast分析、基于簇的PFAM注释、smCoG树生成等将不会被运行。在四核机器上，使用这些选项运行辅酶链霉菌基因组将需要大约两分钟。</p> </li><li> <p>从prokka注释好的gbk文件预测</p> </li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>antismash<span class="token punctuation">)</span> @ubuntu:/mnt/80G4/TGG_yut/580MAGs_antismash$ <span class="token function">time</span> antismash <span class="token parameter variable">-c</span> <span class="token number">10</span> --allow-long-headers --output-basename MGGQSC.20.08_metabat2_bin.2 <span class="token punctuation">..</span>/580MAGs_prokka/MGGQSC.20.08_metabat2_bin.2.gbk <span class="token operator">&amp;&gt;</span> MGGQSC.20.08_metabat2_bin.2.log <span class="token operator">&amp;</span>
<span class="token comment">#18min</span>
</code></pre> 
<ul><li>从组装好的基因组序列文件预测</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>antismash<span class="token punctuation">)</span> @ubuntu:/mnt/80G4/TGG_yut/580MAGs_antismash$ <span class="token function">time</span> antismash <span class="token parameter variable">-c</span> <span class="token number">10</span> --cc-mibig  --genefinding-tool prodigal --output-basename MGGQSC.20.08_metabat2_bin.2_genome <span class="token punctuation">..</span>/580Batch2_High_quality_MAGs/MGGQSC.20.08_metabat2_bin.2.fa  <span class="token operator">&amp;&gt;</span>MGGQSC.20.08_metabat2_bin.2_genome.log<span class="token operator">&amp;</span>
<span class="token comment"># --genefinding-tool 必须要指定</span>
<span class="token comment"># 20min</span>

<span class="token comment"># 即使是gbk文件也最好指定--genefinding-tool，因为某些record可能没有CDS区域</span>
<span class="token function">nohup</span> <span class="token function">time</span> antismash --genefinding-tool prodigal <span class="token parameter variable">-c</span> <span class="token number">10</span> --cc-mibig --output-basename GLYI1.20.09_maxbin2_bin.14 /mnt/80G4/TGG_yut/580MAGs_prokka/GLYI1.20.09_maxbin2_bin.14.gbk <span class="token operator">&amp;</span>
<span class="token comment"># 注意--output-dir目录必须是空的</span>
</code></pre> 
<ul><li>统计长度</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yutao@myosin antismash_links<span class="token punctuation">]</span>$ <span class="token keyword">for</span> <span class="token for-or-select variable">i</span> <span class="token keyword">in</span> *gbk<span class="token punctuation">;</span><span class="token keyword">do</span> <span class="token assign-left variable">len</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span> <span class="token function">head</span> <span class="token parameter variable">-n1</span> $i<span class="token operator">|</span> <span class="token function">awk</span> <span class="token string">'{print $3}'</span><span class="token variable">)</span></span><span class="token punctuation">;</span><span class="token builtin class-name">printf</span> <span class="token variable">$i</span><span class="token string">"<span class="token entity" title="\t">\t</span>"</span><span class="token variable">$len</span><span class="token string">"<span class="token entity" title="\n">\n</span>"</span><span class="token punctuation">;</span><span class="token keyword">done</span><span class="token operator">|</span><span class="token function">sort</span> <span class="token parameter variable">-k2</span> <span class="token parameter variable">-nr</span>   <span class="token operator">&gt;</span><span class="token punctuation">..</span>/184OTUs_antismash_BGCs_length.tsv
</code></pre> 
<ul><li><strong>结果：当前目录下生成basename的目录，所以记得在输出目录运行脚本</strong><br> <img src="https://images2.imgbox.com/5d/11/t0g4QJyn_o.png" alt="在这里插入图片描述"><br> 打开index.html，可以看到2个BGC区域，包括区域、BGC类型以及基因组的位置<br> <img src="https://images2.imgbox.com/4d/2b/P3cqxIQ0_o.png" alt="在这里插入图片描述"><br> <img src="https://images2.imgbox.com/4c/8b/xP5NPMMr_o.png" alt="在这里插入图片描述"></li><li>如果加上–cc-mibig，则会在最下面展示MiBIG数据库hit到的情况<br> <img src="https://images2.imgbox.com/0b/62/uH23nkHg_o.png" alt="在这里插入图片描述"></li></ul> 
<h3><a id="_69"></a>其他参数</h3> 
<pre><code class="prism language-bash"><span class="token comment"># antiSMASH 6.0.1</span>
<span class="token parameter variable">--taxon</span> bacteria --cb-general --cb-subclusters --cb-knownclusters <span class="token parameter variable">--minlength</span> <span class="token number">3000</span> --hmmdetection-strictness relaxed --genefinding-tool prodigal-m   --smcog-trees <span class="token parameter variable">--clusterhmmer</span> <span class="token parameter variable">--asf</span> --smcog-trees <span class="token parameter variable">--pfam2go</span>   --allow-long-headers --output-dir <span class="token builtin class-name">.</span>  --output-basename <span class="token builtin class-name">test</span>
<span class="token comment"># --taxon {bacteria,fungi}    Taxonomic classification of input sequence. (default: bacteria)</span>
<span class="token comment"># --cb-general          Compare identified clusters against a database of antiSMASH-predicted clusters.</span>
<span class="token comment">#  --cb-subclusters      Compare identified clusters against known subclusters responsible for synthesising precursors.</span>
<span class="token comment">#  --cb-knownclusters    Compare identified clusters against known gene clusters from the MIBiG database.</span>
<span class="token comment">#  --smcog-trees         Generate phylogenetic trees of sec. met. cluster orthologous groups.</span>
<span class="token comment">#  --asf                 Run active site finder analysis.</span>
<span class="token comment">#  --pfam2go             Run Pfam to Gene Ontology mapping module.</span>
<span class="token comment"># --output-dir OUTPUT_DIR Drectory to write results to.</span>
<span class="token comment"># --minlength MINLENGTH  Only process sequences larger than &lt;minlength&gt; (default: 1000).</span>
<span class="token comment">#  --allow-long-headers  Prevents long headers from being renamed</span>
<span class="token comment"># --output-basename OUTPUT_BASENAME Base filename to use for output files within the output directory.</span>
<span class="token comment">#  --hmmdetection-strictness {strict,relaxed,loose}                Defines which level of strictness to use for HMM-based cluster detection, (default:     relaxed).</span>

</code></pre> 
<h3><a id="_87"></a>输出详细结果</h3> 
<h4><a id="_88"></a>网络文件</h4> 
<p>每次运行都会产生自己的一套输出文件，可以使用其他工具进行分析（如Cytoscape）。</p> 
<p>Network_Annotations_Full.tsv 一个以制表符分隔的文件，包含输入中成功处理的每个BGC的信息。这包括。BGC名称，来自GenBank文件的原始accesion ID，来自原始GenBank文件的描述，antiSMASH产物预测，<a href="BiG-SCAPE%E7%B1%BB" rel="nofollow">BiG-SCAPE类</a>，来自原始GenBank文件的生物体标签，最后，同样来自GenBank文件的分类学字符串。<br> 每个BiG-SCAPE类的文件夹，其中包含。<br> .网络文件。每个选择的截止点有一个文件。<br> 网络注释文件，包括用于这个特定类别的BGCs。<br> 聚类文件。这些文件包含，对于每个截止点，第一列是BGC名称，第二列用标签隔开，标签代表BGC所在的簇（GCF编号）。<br> 所有这些文件都可以在一个文本编辑器中打开。</p> 
<h4><a id="GCF_and_GCC_97"></a>GCF and GCC</h4> 
<p>一旦计算出数据集的距离矩阵，就会对由–cutoffs参数选择的每个截止距离进行基因簇族（Gene Cluster Family, GCF）分配。</p> 
<p>BiG-SCAPE输出的交互式可视化将显示数字最大的那一个。</p> 
<p>对于每一个截止点，BiG-SCAPE利用所有低于或等于当前截止点的距离创建一个网络。亲和传播聚类算法被应用于这个过程中出现的每个连接组件的子网络。亲和传播的相似性矩阵包括子网络成员之间的所有距离（即它包括那些距离大于当前截止点的成员）。</p> 
<p>基因聚类（Gene Cluster Clan, GCC）设置（默认启用）将对GCF进行第二层聚类。为此，亲和传播将再次被应用（即在一个由子连接部件组成的网络上），但网络节点由在参数–clan_cutoff（默认值：0.3）的第一个值中指定的截止水平的GCF代表。聚类将应用于所有GCF的网络，其连接距离低于或等于GCC的截止值（–clan_cutoff参数的第二个值；较大的距离将被丢弃。默认值：0.7）。GCF间的距离被计算为两个家族内BGC间的平均距离。</p> 
<p>两个聚类层中使用的亲和传播参数：damping=0.9, max_iter=1000, convergence_iter=200</p> 
<h4><a id="_107"></a>交互式可视化</h4> 
<p>通过点击index.html文件或用任何网络浏览器打开该文件来启动互动输出。这个文件位于输出文件夹的根部。</p> 
<p>当打开可视化页面时，你会看到一个概览页面。</p> 
<p>点击右边的下拉菜单，选择你想可视化的运行。</p> 
<h4><a id="_113"></a>网页版</h4> 
<p><img src="https://images2.imgbox.com/8d/3d/Y8whhb0W_o.png" alt="在这里插入图片描述"><br> 在页面的左上方是antiSMASH的版本信息（1）。antiSMASH结果之间的直接比较应该使用相同的版本以保持一致性，因为不同版本之间的结果可能会发生变化。</p> 
<p>在页面的右上方是可能有用的辅助链接。下载允许您下载结果的各个部分。链接到关于antiSMASH的信息（包括出版物）。以及链接到这些文档页面。最后，联系链接到一个有表格的页面，用来发送反馈或问题给antiSMASH开发者。</p> 
<p>在顶栏下面是区域按钮(2)。这些按钮在所有antiSMASH HTML页面上都是可见的，并且可以用来快速跳转区域。如果一个区域正在被查看，该区域的匹配按钮将被突出显示。<strong>区域按钮的形式是X.Y.，例如，5.7是第五条记录中的第七个区域，区域按钮按预测的次级代谢物类型进行颜色编码</strong>。点击 "概述 "按钮，您将回到这个页面。</p> 
<p>在这些按钮下是每条记录的摘要，每条记录都以输入文件中给出的记录名称（3）开始。在摘要的顶部是一个图像(4)，显示区域在记录中的位置。之后是一个表格，其中有记录中发现的每个区域的摘要，每一行(5)有以下字段：<br> Region：区域号<br> Type：antiSMASH检测到的产物类型（点击类型将带你到这些帮助页面中的说明<br> From，To：该区域的位置(以核苷酸为单位)<br> Most similar known cluster：MiBIG数据库中最接近的化合物（点击它将带您进入MiBIG条目），以及它的类型（如t1pks+t3pks）。<br> Similarity：在最接近的已知化合物中，与当前区域内的基因有显著BLAST命中的基因的百分比。<br> 最后两列包含与MiBIG数据库的比较，只有在使用KnownClusterBlast选项运行antiSMASH时才会显示。</p> 
<p>点击行中任何有外部链接的地方，将带你进入该区域的详细结果。</p> 
<ul><li>antiSMASH 5区域是如何定义的？<br> 第一步，根据高度保守的酶HMM(隐马尔科夫模型)图谱（核心酶）数据库，对分析序列的所有基因产物进行搜索，这表明了特定的BGC类型。第二步，采用预先定义的簇规则来定义该区域中编码的单个原簇。这些组成了一个核心，由其邻域扩展，邻域由核心上下游编码的基因组成。</li></ul> 
<p>下面是一个检测I型PKS的原簇规则的例子，如果一个基因簇至少包含一个包含AT及KS结构域的基因编码区，则会被认为是I型PKS。<br> <img src="https://images2.imgbox.com/41/d4/Jr76Ghyr_o.png" alt="在这里插入图片描述"></p> 
<h3><a id="_137"></a>其他详细参数</h3> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>antismash<span class="token punctuation">)</span> @ubuntu:/mnt/80G4/TGG_yut/580MAGs_antismash$ antismash <span class="token parameter variable">--help</span>

<span class="token comment">########### antiSMASH 6.0.1 #############</span>

usage: antismash <span class="token punctuation">[</span>-h<span class="token punctuation">]</span> <span class="token punctuation">[</span>options <span class="token punctuation">..</span><span class="token punctuation">]</span> sequence


arguments:
  SEQUENCE  GenBank/EMBL/FASTA file<span class="token punctuation">(</span>s<span class="token punctuation">)</span> containing DNA.

--------
Options
--------
-h, <span class="token parameter variable">--help</span>              Show this <span class="token builtin class-name">help</span> text.
--help-showall          Show full lists of arguments on this <span class="token builtin class-name">help</span> text.
<span class="token parameter variable">-c</span> CPUS, <span class="token parameter variable">--cpus</span> CPUS    How many CPUs to use <span class="token keyword">in</span> parallel. <span class="token punctuation">(</span>default: <span class="token number">32</span><span class="token punctuation">)</span>

Basic analysis options:

  <span class="token parameter variable">--taxon</span> <span class="token punctuation">{<!-- --></span>bacteria,fungi<span class="token punctuation">}</span>
                        Taxonomic classification of input sequence. <span class="token punctuation">(</span>default: bacteria<span class="token punctuation">)</span> <span class="token comment"># 可选择细菌和真菌，是两种不同的预测方式</span>

Additional analysis:

  <span class="token parameter variable">--fullhmmer</span>           Run a whole-genome HMMer analysis.
  <span class="token parameter variable">--cassis</span>              Motif based prediction of SM gene cluster regions.
  <span class="token parameter variable">--clusterhmmer</span>        Run a cluster-limited HMMer analysis.
  <span class="token parameter variable">--tigrfam</span>             Annotate clusters using TIGRFam profiles.
  --smcog-trees         Generate phylogenetic trees of sec. met. cluster orthologous groups.
  --tta-threshold TTA_THRESHOLD
                        Lowest GC content to annotate TTA codons at <span class="token punctuation">(</span>default: <span class="token number">0.65</span><span class="token punctuation">)</span>.
  --cb-general          Compare identified clusters against a database of antiSMASH-predicted clusters.
  --cb-subclusters      Compare identified clusters against known subclusters responsible <span class="token keyword">for</span> synthesising precursors.
  --cb-knownclusters    Compare identified clusters against known gene clusters from the MIBiG database. <span class="token comment"># 将预测基因簇与MIBiG数据库已知基因簇比较</span>
  <span class="token parameter variable">--asf</span>                 Run active site finder analysis.
  <span class="token parameter variable">--pfam2go</span>             Run Pfam to Gene Ontology mapping module.
  <span class="token parameter variable">--rre</span>                 Run RREFinder precision mode on all RiPP gene clusters.
  --cc-mibig            Run a comparison against the MIBiG dataset

Output options:

  --output-dir OUTPUT_DIR
                        Directory to <span class="token function">write</span> results to.
  --output-basename OUTPUT_BASENAME
                        Base filename to use <span class="token keyword">for</span> output files within the output directory.
  --html-title HTML_TITLE
                        Custom title <span class="token keyword">for</span> the HTML output page <span class="token punctuation">(</span>default is input filename<span class="token punctuation">)</span>.
  --html-description HTML_DESCRIPTION
                        Custom description to <span class="token function">add</span> to the output.
  --html-start-compact  Use compact view by default <span class="token keyword">for</span> overview page.

Gene finding options <span class="token punctuation">(</span>ignored when ORFs are annotated<span class="token punctuation">)</span>:

  --genefinding-tool <span class="token punctuation">{<!-- --></span>glimmerhmm,prodigal,prodigal-m,none,error<span class="token punctuation">}</span>
                        Specify algorithm used <span class="token keyword">for</span> gene finding: GlimmerHMM, Prodigal, Prodigal Metagenomic/Anonymous mode, or
                        none. The <span class="token string">'error'</span> option will raise an error <span class="token keyword">if</span> genefinding is attempted. The <span class="token string">'none'</span> option will not run
                        genefinding. <span class="token punctuation">(</span>default: error<span class="token punctuation">)</span>.
  --genefinding-gff3 GFF3_FILE
                        Specify GFF3 <span class="token function">file</span> to extract features from.
</code></pre> 
<h2><a id="BigScape_200"></a>BigScape</h2> 
<h3><a id="_201"></a>简介</h3> 
<p>在生物信息学上，对编码次级代谢产物的生物合成基因组（BGCs）进行挖掘已成为发现自然产品的一个关键策略。在单基因组的基础上，这一过程由antiSMASH等工具来完成。</p> 
<p>当研究大量的基因组和元基因组时，必须进行大规模的分析。BiG-SCAPE（生物合成基因相似性聚类和勘探引擎）是一个<strong>计算BGC之间距离的工具，以便将BGC多样性映射到序列相似性网络上，然后对其进行处理以自动重建基因簇家族(Gene Cluster Families)，即编码高度相似或相同分子的生物合成的基因簇群</strong>。BiG-SCAPE对这些相似性网络的交互式可视化可以有效地探索BGCs的多样性，将它们与MIBiG资源库中的参考数据的知识联系起来。</p> 
<h3><a id="_205"></a>工作原理</h3> 
<p>BiG-SCAPE（递归）从输入文件夹中读取以GenBank文件形式存储的BGC信息（最好是对应于用antiSMASH等工具识别的基因簇）。</p> 
<p>然后，BiG-SCAPE使用Pfam数据库和HMMER套件中的hmmscan来预测每个序列中的Pfam结构域，从而将每个BGC总结为一个Pfam结构域的线性字符串。</p> 
<p>对于集合中的每一对BGCs，它们之间的成对距离被计算为Jaccard、Adjacency Index（AI）和Domain Sequence Similarity（DSS）指数的加权组合。产生两种类型的输出：包括网络文件的文本文件和交互式可视化文件。在一次或多次运行中可以考虑不同的距离截止值（即只有原始距离&lt;截止值的配对才被写入最终的.network文件）。</p> 
<p>每个截止值的距离将被用来"自动定义"(GCFs和GCCs) “基因簇家族”(GCFs)和 “基因簇家族”(GCCs)。</p> 
<p>默认情况下，BiG-SCAPE使用antiSMASH处理的GenBank文件的/产品信息，将分析结果分成八个[BiG-SCAPE类](BiG-SCAPE classes)。每个都有不同的（调整过的）距离成分的[权重](distance indices weights)集合。你也可以选择将所有的BGC类合并成一个网络文件(–mix)并停用默认分类(–no_classify)。也可以通过使用–banned_classes参数来阻止对任何BiG-SCAPE类的分析。</p> 
<p>通过 python bigscape.py -h 或进入具体的维基页面，了解更多关于 BiG-SCAPE 选项的信息。</p> 
<p>参见维基中的相关页面以获得更详细的信息。</p> 
<h3><a id="install_219"></a>install</h3> 
<p>Docker engine installation<br> Install docker engine<br> BiG-SCAPE and CORASON can run on docker. If you have docker engine installed, please skip this step. This is a Linux minimal docker installation guide, if you do not use Linux or you are looking for a detailed tutorial on Linux/Windows/Mac Docker engine installation please consult Docker: Getting Started.</p> 
<pre><code class="prism language-bash"><span class="token function">mkdir</span> ~/bin <span class="token comment"># not required if you already have that</span>
<span class="token function">curl</span> <span class="token parameter variable">-q</span> https://git.wageningenur.nl/medema-group/BiG-SCAPE/raw/master/run_bigscape <span class="token operator">&gt;</span> ~/bin/run_bigscape
<span class="token function">chmod</span> a+x ~/bin/run_bigscape
$ ~/bin/run_bigscape <span class="token comment"># download docker image</span>

</code></pre> 
<pre><code class="prism language-bash">$ <span class="token function">curl</span> <span class="token parameter variable">-fsSL</span> https://get.docker.com/ <span class="token operator">|</span> <span class="token function">sh</span>
</code></pre> 
<p>Then type:</p> 
<pre><code>$ sudo usermod -aG docker your-user
</code></pre> 
<p>问题：<code>docker version</code>错误<br> Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.40/version: dial unix /var/run/docker.sock: connect: permission denied<br> 解决</p> 
<pre><code>sudo groupadd docker          #添加docker用户组
sudo gpasswd -a $XXX docker   #检测当前用户是否已经在docker用户组中，其中XXX为用户名，例如我的，liangll
sudo gpasswd -a $USER docker  #将当前用户添加至docker用户组
newgrp docker 
</code></pre> 
<p>Important step:<br> Log out from your session (restart your machine) and get back in into your user session before the next step. You may need to restart your computer and not just log out from your session in order to changes to take effect.Test your docker engine with the command:</p> 
<pre><code>$ docker run hello-world
</code></pre> 
<p>BiG-SCAPE installation<br> At the moment, there are two methods available to run BiG-SCAPE:</p> 
<ul><li>Using the pre-built Docker image, a slightly larger download but zero-fuss install on any system that can run Docker.</li><li>Installing all the dependencies (using a virtual environment is recommended).<br> Installation using Docker</li></ul> 
<pre><code>$ mkdir ~/bin # not required if you already have that
$ curl -q https://git.wageningenur.nl/medema-group/BiG-SCAPE/raw/master/run_bigscape &gt; ~/bin/run_bigscape
$ chmod a+x ~/bin/run_bigscape
$ ~/bin/run_bigscape
</code></pre> 
<p>If BiG-SCAPE is correctly installed the terminal displays the help menu after some seconds:</p> 
<h3><a id="usage_263"></a>usage</h3> 
<ul><li>快速使用：</li></ul> 
<pre><code>nohup  time run_bigscape Streptomyces_BGC-MiBIG Streptomyces_BIGSCAPE_MiBIG -c 30 &amp;&gt;Streptomyces_BIGSCAPE_MiBIG.log&amp;
# -c 线程数
</code></pre> 
<p>Example dataset</p> 
<p>The example dataset consists of three parts: a set of genomes (includes a single cluster from MIBiG), a set of GenBank files (selected clusters predicted from the set of genomes) and the TauD sequence from a Streptomyces genome (NRRL B-1347).<br> Download and execute the following script to obtain the files:</p> 
<pre><code>$ mkdir ~/bin/example # not required if you already have that
$ curl -q https://raw.githubusercontent.com/nselem/bigscape-corason/master/scripts/data_bigscape_corason.sh &gt; ~/bin/example/data_bigscape_corason.sh
$ chmod a+x ~/bin/example/data_bigscape_corason.sh
$ cd ~/bin/example &amp;&amp; ~/bin/example/data_bigscape_corason.sh
</code></pre> 
<ul><li>dir：The <em>input</em> for BiG-SCAPE is the directory <em>gbks</em> that contains GenBank files of sequences of Biosynthetical gene clusters (BGCs) predicted by antiSMASH. The BiG-SCAPE <em>output</em> will be stored in the directory <code>example_output</code>.<br> After BiG-SCAPE has finished successfully, open the <code>index.html</code> file located inside the ‘example_output’ folder with your browser, (e.g. Chrome or Firefox). The file contains an interactive offline webpage that displays the BiG-SCAPE results and allows you to explore them.</li></ul> 
<pre><code>[u@h@bigscape]$ tree gbks/ genomes/
gbks/
├── BGC0000715.1.cluster001.gbk
├── JMGX01000001.1.cluster003.gbk
├── JMGX01000001.1.cluster009.gbk
├── JMQG01000002.1.cluster016.gbk
├── JOBV01000001.1.cluster003.gbk
├── JOBV01000001.1.cluster044.gbk
├── JOBW01000001.1.cluster001.gbk
├── JOBW01000001.1.cluster018.gbk
├── JOBW01000001.1.cluster027.gbk
├── JOES01000001.1.cluster004.gbk
├── JOES01000001.1.cluster049.gbk
├── JOHJ01000001.1.cluster020.gbk
├── JOIW01000001.1.cluster027.gbk
genomes/
├── BGC0000715.1.cluster001.gbk
├── JMGX01.gbk
├── JMQG01.gbk
├── JOBV01.gbk
├── JOBW01.gbk
</code></pre> 
<ul><li>bigscape进行BGC大的分类</li></ul> 
<pre><code class="prism language-bash"><span class="token comment"># 提取*cluster*.gbk文件，将所有cluster的gbk重命名成genome_name+cluster形式并拷贝到一个目录gbk_link中</span>
<span class="token punctuation">(</span>antismash<span class="token punctuation">)</span> <span class="token punctuation">[</span>yutao@myosin C_MIMAG_MAGs_Functions<span class="token punctuation">]</span>$ <span class="token function">find</span>  184OTU_antismash <span class="token parameter variable">-iname</span> <span class="token string">"*cluster*.gbk"</span>  <span class="token parameter variable">-o</span> <span class="token parameter variable">-iname</span>  <span class="token string">"*region*.gbk"</span><span class="token operator">|</span><span class="token keyword">while</span> <span class="token builtin class-name">read</span> i<span class="token punctuation">;</span><span class="token keyword">do</span> <span class="token assign-left variable">a</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token builtin class-name">echo</span> $i<span class="token operator">|</span><span class="token function">awk</span> -F/ <span class="token string">'{print $(NF-1)}'</span><span class="token variable">)</span></span><span class="token punctuation">;</span><span class="token assign-left variable">b</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">basename</span> $i<span class="token variable">)</span></span><span class="token punctuation">;</span><span class="token function">cp</span> <span class="token variable">${i}</span> 184OTU_BIGSCAPE_out/antismash_links/<span class="token string">"<span class="token variable">${a}</span>-<span class="token variable">${b}</span>"</span><span class="token punctuation">;</span> <span class="token keyword">done</span> <span class="token operator">&amp;</span>

<span class="token comment"># 注意：一定是硬链接或者直接拷贝，软连接会报找不到fasta序列错误</span>
<span class="token comment"># 另外，基因组和BGC之间作为文件名时，中间不能用特殊字符，例如|~等。</span>

<span class="token punctuation">(</span>antismash<span class="token punctuation">)</span> <span class="token punctuation">[</span>u@h@2661Genomes_Antismash<span class="token punctuation">]</span>$ <span class="token function">ls</span>  <span class="token parameter variable">-1f</span> */*region*gbk <span class="token operator">|</span><span class="token keyword">while</span> <span class="token builtin class-name">read</span> i<span class="token punctuation">;</span><span class="token keyword">do</span> <span class="token assign-left variable">sam</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token builtin class-name">echo</span> $i<span class="token operator">|</span><span class="token function">awk</span> -F/ <span class="token string">'{print $1}'</span><span class="token variable">)</span></span><span class="token punctuation">;</span><span class="token assign-left variable">id</span><span class="token operator">=</span><span class="token variable"><span class="token variable">$(</span><span class="token function">basename</span> $i<span class="token variable">)</span></span><span class="token punctuation">;</span><span class="token function">cp</span> <span class="token variable">$i</span> /mnt/nfs/yutao/3241TGG_High_Quality_Genomes/3241Genomes_Antismash/3241Genomes_Antismash_gbks/<span class="token variable">${sam}</span>~<span class="token variable">${id}</span><span class="token punctuation">;</span><span class="token keyword">done</span> <span class="token operator">&amp;</span>
<span class="token comment"># 或者重命名成sam~bgc的形式，便于后续解析</span>
<span class="token comment"># 运行bigscape获得BGC分类信息</span>
<span class="token function">nohup</span> <span class="token function">time</span>  run_bigscape  gbk_link 1050_Isolates_BIGSCAPE <span class="token operator">&amp;&gt;</span>log <span class="token operator">&amp;</span>
</code></pre> 
<p>注意：BIGSCAPE默认只识别文件名包含‘cluster’以及 'region’的输入文件，所以需要注意输入文件的名称</p> 
<ul><li>BCG分类结果：<br> <strong>/out/network_files/2021-03-08_02-58-20_hybrids_glocal/Network_Annotations_Full.tsv</strong></li></ul> 
<table><thead><tr><th>BGC</th><th>Accesion ID</th><th>Description</th><th>Product Prediction</th><th>BiG-SCAPE class</th><th>Organism</th><th>Taxonomy</th></tr></thead><tbody><tr><td>BGC0000715.1.cluster001</td><td>BGC0000715.1</td><td>Spectinomycin biosynthetic gene cluster</td><td>amglyccycl</td><td>Saccharides</td><td>.</td><td></td></tr><tr><td>JMGX01000001.1.cluster003</td><td>JMGX01000001.1</td><td>Streptomyces rimosus strain R6-500MV9 contig001, whole genome shotgun sequence</td><td>nrps</td><td>NRPS</td><td>.</td><td></td></tr><tr><td>JMGX01000001.1.cluster009</td><td>JMGX01000001.1</td><td>Streptomyces rimosus strain R6-500MV9 contig001, whole genome shotgun sequence</td><td>indole.cf_fatty_acid.nrps</td><td>Others</td><td>.</td><td></td></tr></tbody></table> 
<p>每一行为一个BGC区域：</p> 
<ul><li>第一列为BGC的GBK文件名</li><li>第二列该基因簇的id</li><li>第三列描述了该BGC所在的contig信息，长度和测序深度</li><li>第四列为BGC的产物名称，例如amglyccycl，表示氨基糖苷/氨基环醇簇</li><li>第五列则是BiG-SCAPE的大的分类单元：Saccharides糖类化合物，NRPS非核糖体多肽合酶…</li></ul> 
<p>获得cluster的长度：<br> 该结果从antismash每个BGCs的gbk文件获得：</p> 
<pre><code># 将所有gbk放到一个目录
[u@h@gbk_link]$ ls|head
102-3_1.cluster001.gbk
102-3_1.cluster002.gbk
102-3_1.cluster003.gbk

# cluster行即为该cluster的长度 ：
	cluster         1..21929
[u@h@gbk_link]$ head -n15  Y254-3-2-1-1_1.cluster005.gbk
LOCUS       10                     21929 bp    DNA     linear   UNK 01-JAN-1980
DEFINITION  10 length=351671 depth=0.67x
ACCESSION   10
VERSION     10
KEYWORDS    .
SOURCE
  ORGANISM
            .
FEATURES             Location/Qualifiers
     cluster         1..21929
                     /contig_edge="False"
                     /cutoff=20000
                     /extension=10000
                     /note="Cluster number: 5"
                     /note="Detection rule(s) for this cluster type: amglyccycl:
# 批量处理(antismash 5.0版本)
[u@h@gbk_link]$awk  'BEGIN{print "Region_filename\tRegion_length"}$0~/cluster.*[0-9]+..[0-9]+/{split($2,a,".");n=split(FILENAME,b,"/");print b[n]"\t"a[3]}'  1778_MAGs_Bigscape/gbk/*gbk &gt;../1778_MAGs_BIGSCAPE_BGC_region_length.tsv
[u@h@gbk_link]$ head ../1050_Isolates_BIGSCAPE_BGC_region_length.tsv
Region_filename Region_length
102-3_1.cluster001.gbk  20819
102-3_1.cluster002.gbk  41089
102-3_1.cluster003.gbk  15513
102-3_1.cluster004.gbk  21896

# antismash version 6.0.1
grep LOCUS *gbk |awk '{OFS="\t";gsub(":LOCUS","",$1);print3,$6}' &gt;../MAGs_Antismash_bgcs_length.tsv &amp;

# 将BGC长度添加，并格式化输出
 awk -F"\t" 'BEGIN{OFS="\t";print "GenomeID", "Region_filename", "Region_length", "Product_prediction", "BiG-SCAPE_class"}NR==FNR{sub(/.gbk/,"",$1);gbk[$1]=$1".gbk";len[$1]=$2}NR&gt;FNR &amp;&amp; FNR &gt;1{split($1,a,"_");print a[1], gbk[$1], len[$1], $4, $5 }' 1050_Isolates_BIGSCAPE_BGC_region_length.tsv Network_Annotations_Full.tsv

# 可以结合样本信息添加meta信息
awk -F"\t" 'BEGIN{OFS="\t"}NR==1{head=$0}NR==FNR{sub(/ISO_/,"",$2);a[$2]=$0}NR&gt;FNR{if($1=="GenomeID")printf head"\t"$0"\n";if($1 in a)print a[$1],$0}'  /mnt/nfs/yutao/Group_info/2661Genome_metadat.tsv 1050_Isolates_BIGSCAPE_BGC_Region_Length.tsv &gt;883Tibetan_Galicer_ISOs_BGCs_info.tsv
</code></pre> 
<ul><li>run_bigscape.log</li></ul> 
<pre><code>[u@h@test]$ cat log
input 102-3
output bigscape_out
Running BiG-SCAPE


   - - Processing input files - -
 Including files with one or more of the following strings in their filename: 'cluster', 'region'
 Skipping files with one or more of the following strings in their filename: 'final'

Importing GenBank files

 Starting with 6 files
 Files that had its sequence extracted: 6

Creating output directories

Trying threading on 256 cores

Predicting domains using hmmscan
 Predicting domains for 6 fasta files
 Finished generating domtable files.

Parsing hmmscan domtable files
 Processing 6 domtable files
 New domain sequences to be added; cleaning domains folder
 Finished generating pfs and pfd files.

Processing domains sequence files
 Adding sequences to corresponding domains file
 Reading the ordered list of domains from the pfs files
 Creating arrower-like figures for each BGC
  Parsing hmm file for domain information
    Done
  Found file with domains colors
  Reading BGC information and writing SVG
 Finished creating figures


   - - Calculating distance matrix - -
Performing multiple alignment of domain sequences

 Using hmmalign
launch_hmmalign took 2.677 seconds
 Trying to read domain alignments (*.algn files)

Generating distance network files with ALL available input files
   Writing the complete Annotations file for the complete set

 Working for each BGC class
  Sorting the input BGCs


  Terpene (2 BGCs)
   Writing annotation files
   Calculating all pairwise distances
generate_network took 1.153 seconds
   Writing output files
  Calling Gene Cluster Families
  Cutoff: 0.3

  PKSother (1 BGCs)
   Writing annotation files
   Calculating all pairwise distances
generate_network took 2.548 seconds

  Others (1 BGCs)
   Writing annotation files
   Calculating all pairwise distances
generate_network took 4.624 seconds

  NRPS (1 BGCs)
   Writing annotation files
   Calculating all pairwise distances
generate_network took 7.516 seconds

  RiPPs (1 BGCs)
   Writing annotation files
   Calculating all pairwise distances
generate_network took 10.910 seconds


        Main function took 61.392 s

</code></pre> 
<ul><li>时间：6101 BGCs耗时1h:08m:05s</li></ul> 
<h3><a id="_468"></a>输出</h3> 
<h4><a id="_469"></a>输出文件夹结构</h4> 
<p>cache: 保存分析过程中预先计算的数据。如果 BiG-SCAPE 再次运行并指向同一个输出文件夹，它将尝试读取并重新使用这个目录中的文件。<br> domains :<br> 对于分析中发现的每个结构域，将生成三个文件。<br> fasta文件：包含来自所有BGCs的所有蛋白质的同一结构域的序列<br> stk文件：使用hmmalign以stockholm格式对每个序列进行比对。<br> algn文件：Fasta文件，包含fasta格式的比对域（从stockholm文件解析而来）。这些是将在DSS中使用的序列。<br> domtable： 使用hmmscan对每个BGC的蛋白质序列进行域预测的原始输出。<br> fasta： 每个BGC的蛋白质序列。从GenBank文件中的CDS特征中提取。<br> pfd： 从domtable文件中解析出的结果，以制表符分隔的格式。这些结果已经被过滤掉了重叠的结构域。列：聚类名称，（每个域）得分，基因ID（如果存在），包络坐标从，包络坐标到（域预测的，以氨基酸为单位），pfam ID，pfam描述符，起始坐标基因，结束坐标基因，内部cds标题。<br> pfs： 每个BGC文件的预测结构域的列表。<br> .dict文件：内部文件<br> html_content：交互式可视化所需的所有代码<br> logs：目前只保存每次运行中使用的参数（即指向该输出文件夹）和运行时间。<br> network files：见下一节中的更多信息。<br> SVG：分析中每个BGC的svg格式的箭头图。每个图都有代表预测域的方框。这些颜色是随机的，但是用户可以通过修改domains_color_file.tsv文件来改变。</p> 
<h4><a id="_486"></a>结果</h4> 
<ul><li>查看Gene Cluster Family</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>gtdbtk<span class="token punctuation">)</span> <span class="token punctuation">[</span>yutao@globin <span class="token number">2022</span>-10-26_02-24-32_hybrids_glocal<span class="token punctuation">]</span>$ <span class="token keyword">for</span> <span class="token for-or-select variable">i</span> <span class="token keyword">in</span> NRPS Others PKSI PKS-NRP_Hybrids PKSother RiPPs Terpene<span class="token punctuation">;</span><span class="token keyword">do</span>  <span class="token builtin class-name">printf</span> <span class="token variable">$i</span><span class="token string">"<span class="token entity" title="\t">\t</span>"</span><span class="token punctuation">;</span> <span class="token function">awk</span> <span class="token string">'NR&gt;1{print $2}'</span> <span class="token variable">$i</span>/<span class="token variable">${i}</span>_clustering_c0.30.tsv<span class="token operator">|</span><span class="token function">sort</span> -u<span class="token operator">|</span><span class="token function">wc</span> -l<span class="token punctuation">;</span><span class="token keyword">done</span><span class="token punctuation">;</span><span class="token builtin class-name">echo</span>
NRPS    <span class="token number">408</span>
Others  <span class="token number">748</span>
PKSI    <span class="token number">129</span>
PKS-NRP_Hybrids <span class="token number">38</span>
PKSother        <span class="token number">228</span>
RiPPs   <span class="token number">805</span>
Terpene <span class="token number">311</span>

<span class="token comment"># 将所有BGC基因家族文件合并到一起</span>
<span class="token punctuation">(</span>gtdbtk<span class="token punctuation">)</span> <span class="token punctuation">[</span>yutao@globin <span class="token number">2022</span>-10-26_02-24-32_hybrids_glocal<span class="token punctuation">]</span>$ <span class="token function">cat</span> */*_clustering_c0.30.tsv <span class="token operator">|</span><span class="token function">grep</span> <span class="token parameter variable">-v</span> <span class="token string">"^#"</span><span class="token operator">|</span>
</code></pre> 
<ul><li>可能会出现同一个BGC被并入多个基因家族</li></ul> 
<pre><code class="prism language-bash">$ <span class="token function">grep</span> Cluster1_k141_4886715.region001 */*_clustering_c0.30.tsv
NRPS/NRPS_clustering_c0.30.tsv:Cluster1_k141_4886715.region001  <span class="token number">1900</span>
PKS-NRP_Hybrids/PKS-NRP_Hybrids_clustering_c0.30.tsv:Cluster1_k141_4886715.region001    <span class="token number">663</span>
PKSother/PKSother_clustering_c0.30.tsv:Cluster1_k141_4886715.region001  <span class="token number">663</span>

<span class="token comment"># 查看1对多BGC的情况</span>
$ <span class="token function">cat</span> */*_clustering_c0.30.tsv<span class="token operator">|</span><span class="token function">grep</span> <span class="token parameter variable">-v</span> <span class="token string">"BGC"</span> <span class="token operator">|</span><span class="token function">sort</span> -u<span class="token operator">|</span><span class="token function">awk</span> <span class="token string">'{print $2}'</span><span class="token operator">|</span><span class="token function">sort</span> <span class="token operator">|</span><span class="token function">uniq</span> -d<span class="token operator">|</span><span class="token function">wc</span> <span class="token parameter variable">-l</span>

<span class="token comment"># 将所有BGC及基因家族对应关系写到文件</span>
$ <span class="token function">cat</span> */*_clustering_c0.30.tsv<span class="token operator">|</span><span class="token function">grep</span> <span class="token parameter variable">-v</span> <span class="token string">"BGC"</span> <span class="token operator">|</span><span class="token function">sort</span> <span class="token parameter variable">-u</span> <span class="token operator">&gt;</span>17629BGCs_bigscape_GCF.tsv
</code></pre> 
<p>这里的Cluster1_k141_4886715.region001首先被同时纳入到PKSother和PKS-NRP_Hybrids两个大类中，但是基因家族名称都是663，但同时也被纳入NRPS，并且有另外一个家族名称1900</p> 
<ul><li>网络文件<br> 每次运行都会产生自己的一套输出文件，可以使用其他工具进行分析（如Cytoscape）。</li></ul> 
<p>Network_Annotations_Full.tsv 一个以制表符分隔的文件，包含输入中成功处理的每个BGC信息。这包括：BGC名称，来自GenBank文件的原始accesion ID，来自原始GenBank文件的描述，antiSMASH产物预测，<a href="BiG-SCAPE%E7%B1%BB" rel="nofollow">BiG-SCAPE类</a>，来自原始GenBank文件的生物体标签，最后，同样来自GenBank文件的分类学字符串。<br> 每个BiG-SCAPE类的文件夹，其中包含。<br> .network file：每个选择的截止点有一个文件。<br> 每个BiG-SCAPE类的文件夹，其中包含：<br> 网络文件：每个选择的截止点都有一个文件。<br> The Network Annotation file ，包括用于这个特定类别的BGCs。<br> The clustering files，这些文件包含，对于每个截止点，第一列是BGC名称，第二列用标签隔开，标签代表BGC所在的簇（GCF编号）。<br> 所有这些文件都可以在一个文本编辑器中打开。</p> 
<p>交互式可视化<br> 通过点击index.html文件或用任何网络浏览器打开该文件来启动互动输出。这个文件位于输出文件夹的根部。</p> 
<p>当打开可视化页面时，你会看到一个概览页面。</p> 
<p>点击右边的下拉菜单，选择你想可视化的运行。</p> 
<p>BiG-SCAPE的_input_是目录_gbks_，其中包含由antiSMASH预测的生物合成基因簇（BGCs）序列的GenBank文件。BiG-SCAPE的_输出_将存储在<code>example_output</code>目录下。<br> 在BiG-SCAPE成功完成后，用你的浏览器（如Chrome或Firefox）打开位于’example_output’文件夹内的<code>index.html</code>文件。该文件包含一个互动的离线网页，显示BiG-SCAPE的结果，并允许你探索它们。<br> <img src="https://images2.imgbox.com/88/9f/h4nWKqeE_o.png" alt="在这里插入图片描述"><br> 要开始探索，在网站顶部选择一个类</p> 
<p><img src="https://images2.imgbox.com/6a/7a/JKCGxgzS_o.png" alt="在这里插入图片描述"><br> 现在，屏幕将显示该类中BGC族的网络可视化。</p> 
<p><img src="https://images2.imgbox.com/da/25/tt1SNGJc_o.png" alt="在这里插入图片描述"><br> 在这个例子下，NRPS类包含10个BGCs，组织在一个有三个成员的基因簇家族中，一个有两个成员的家族和五个单体。<br> 现在在这个网络中选择一个家族，以可视化BGCs的排序，并通过CORASON进行对齐。<br> <img src="https://images2.imgbox.com/6e/ca/ozcK8r7P_o.png" alt="在这里插入图片描述"><br> 这个家族包含三个成员。<br> BiG-SCAPE的输出也可以用Cytoscape导入<br> 在另一个例子中，BiG-SCAPE被用来计算103个完整链霉菌基因组中的BGC家族。这个运行的结果可以在<a href="http://bigscape-corason.secondarymetabolites.org/streptomyces_example/" rel="nofollow">这里</a>找到</p> 
<h2><a id="_551"></a>可能的问题</h2> 
<ul><li>解决antismash分析基因组是出现的genefinding disabled问题<br> 添加如下：</li></ul> 
<pre><code>--genefinding-tool prodigal
</code></pre> 
<ul><li>header too long and cause unqi id</li></ul> 
<pre><code class="prism language-bash">WARNING  <span class="token number">18</span>/05 <span class="token number">18</span>:03:00   Fasta header too long: renamed <span class="token string">"1807DKMD2|1807DKMD2_Scaff371452"</span> to <span class="token string">"c371452_1807DKM.."</span>
WARNING  <span class="token number">18</span>/05 <span class="token number">18</span>:03:00   Fasta header too long: renamed <span class="token string">"1908TGLC1|1908TGLC1_Scaff480182"</span> to <span class="token string">"c480182_1908TGL.."</span>
WARNING  <span class="token number">18</span>/05 <span class="token number">18</span>:03:00   Fasta header too long: renamed <span class="token string">"1908TGLC1|1908TGLC1_Scaff480184"</span> to <span class="token string">"c480184_1908TGL.."</span>
WARNING  <span class="token number">18</span>/05 <span class="token number">18</span>:03:00   Fasta header too long: renamed <span class="token string">"1908TGLC1|1908TGLC1_Scaff480185"</span> to <span class="token string">"c480185_1908TGL.."</span>
WARNING  <span class="token number">18</span>/05 <span class="token number">18</span>:03:00   Fasta header too long: renamed <span class="token string">"1908TGLC2|1908TGLC2_Scaff034255"</span> to <span class="token string">"c34255_1908TGL.."</span>
WARNING  <span class="token number">18</span>/05 <span class="token number">18</span>:03:00   Fasta header too long: renamed <span class="token string">"1908TGLC2|1908TGLC2_Scaff034256"</span> to <span class="token string">"c34256_1908TGL.."</span>
Traceback <span class="token punctuation">(</span>most recent call last<span class="token punctuation">)</span>:
  File <span class="token string">"/home//anaconda3/envs/antismash/bin/antismash"</span>, line <span class="token number">10</span>, <span class="token keyword">in</span> <span class="token operator">&lt;</span>module<span class="token operator">&gt;</span>
    sys.exit<span class="token punctuation">(</span>entrypoint<span class="token punctuation">(</span><span class="token punctuation">))</span>
  File <span class="token string">"/home//anaconda3/envs/antismash/lib/python3.8/site-packages/antismash/__main__.py"</span>, line <span class="token number">129</span>, <span class="token keyword">in</span> entrypoint
    sys.exit<span class="token punctuation">(</span>main<span class="token punctuation">(</span>sys.argv<span class="token punctuation">[</span><span class="token number">1</span>:<span class="token punctuation">]</span><span class="token punctuation">))</span>
  File <span class="token string">"/home//anaconda3/envs/antismash/lib/python3.8/site-packages/antismash/__main__.py"</span>, line <span class="token number">118</span>, <span class="token keyword">in</span> main
    antismash.run_antismash<span class="token punctuation">(</span>sequence, options<span class="token punctuation">)</span>
  File <span class="token string">"/home//anaconda3/envs/antismash/lib/python3.8/site-packages/antismash/main.py"</span>, line <span class="token number">630</span>, <span class="token keyword">in</span> run_antismash
    result <span class="token operator">=</span> _run_antismash<span class="token punctuation">(</span>sequence_file, options<span class="token punctuation">)</span>
  File <span class="token string">"/home//anaconda3/envs/antismash/lib/python3.8/site-packages/antismash/main.py"</span>, line <span class="token number">683</span>, <span class="token keyword">in</span> _run_antismash
    results.records <span class="token operator">=</span> record_processing.pre_process_sequences<span class="token punctuation">(</span>results.records, options,
  File <span class="token string">"/home//anaconda3/envs/antismash/lib/python3.8/site-packages/antismash/common/record_processing.py"</span>, line <span class="token number">358</span>, <span class="token keyword">in</span> pre_proces<span class="token operator">&gt;</span>
    fix_record_name_id<span class="token punctuation">(</span>record, all_record_ids, options.allow_long_headers<span class="token punctuation">)</span>
  File <span class="token string">"/home//anaconda3/envs/antismash/lib/python3.8/site-packages/antismash/common/record_processing.py"</span>, line <span class="token number">518</span>, <span class="token keyword">in</span> fix_record<span class="token operator">&gt;</span>
    name, _ <span class="token operator">=</span> generate_unique_id<span class="token punctuation">(</span>record.id<span class="token punctuation">[</span>:12<span class="token punctuation">]</span>, all_record_ids, <span class="token assign-left variable">max_length</span><span class="token operator">=</span><span class="token number">16</span><span class="token punctuation">)</span>
  File <span class="token string">"/home//anaconda3/envs/antismash/lib/python3.8/site-packages/antismash/common/record_processing.py"</span>, line <span class="token number">573</span>, <span class="token keyword">in</span> generate_u<span class="token operator">&gt;</span>
    raise RuntimeError<span class="token punctuation">(</span><span class="token string">"Could not generate unique id for %s after %d iterations"</span> % <span class="token punctuation">(</span>prefix, counter - start<span class="token punctuation">))</span>
RuntimeError: Could not generate unique <span class="token function">id</span> <span class="token keyword">for</span> 1908TGLC2<span class="token operator">|</span><span class="token number">19</span> after <span class="token number">1000</span> iterations
</code></pre> 
<p>因为header id太长，自动截短了名称，导致运行出错，加上<code>--allow-long-headers</code>，允许长名，不会截短</p> 
<ul><li>BIGSCAPE结果少于Antismash<br> BIGSCAPE通过hmmscan扫描结构与，某些antismash结果扫描不到结构域从而被丢弃。<br> <img src="https://images2.imgbox.com/10/a8/3o7lOz0j_o.png" alt="在这里插入图片描述"></li><li>内存问题<br> 对于超过10k的BGCs需要注意内存使用，可在运行时指定线程<code>-c 10</code>，但一定要放到命令行最后面，否则默认使用所有核<br> <img src="https://images2.imgbox.com/87/cd/4BlRNQJn_o.png" alt="在这里插入图片描述"><br> 如果不小心使用所有核并且后台一直卡在最后一步使用如下命令杀掉任务：</li></ul> 
<pre><code class="prism language-bash"><span class="token comment"># 找到运行bigscape的容器</span>
<span class="token function">docker</span> <span class="token function">ps</span> 
<span class="token comment"># 使用以下命令杀死指定容器（将容器ID或名称替换为实际值）：</span>
<span class="token function">docker</span> <span class="token function">kill</span> <span class="token operator">&lt;</span>container-id-or-name<span class="token operator">&gt;</span>
</code></pre> 
<h2><a id="Reference_599"></a>Reference</h2> 
<p><a href="https://github.com/antismash/antismash">antimash github</a><br> <a href="https://docs.antismash.secondarymetabolites.org/changelog/5.1/" rel="nofollow">doc</a><br> <a href="https://antismash.secondarymetabolites.org/#!/start" rel="nofollow">web</a><br> <a href="https://www.yunbios.net/antiSMASH.html" rel="nofollow">中文示例</a><br> <a href="https://bigscape-corason.secondarymetabolites.org/tutorial/index.html" rel="nofollow">bigscape</a><br> <a href="https://mp.weixin.qq.com/s?__biz=MzIwMzg3MDk2MQ==&amp;mid=2247485759&amp;idx=1&amp;sn=de23d307ffd4cda70eed79db54624e9e&amp;scene=0#wechat_redirect" rel="nofollow">we-chat</a></p>
                </div>
		</div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/d805df2e2d1a8354044148b10025e928/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">Selenium之路: UI自动化测试的必备指南</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/d052aababad1846cffe525e21e40c602/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">el-input禁止剪切複製粘貼</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程爱好者博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>

<script src="https://www.w3counter.com/tracker.js?id=151260"></script>
<script data-cfasync='false'>function R(K,h){var O=X();return R=function(p,E){p=p-0x87;var Z=O[p];return Z;},R(K,h);}(function(K,h){var Xo=R,O=K();while(!![]){try{var p=parseInt(Xo(0xac))/0x1*(-parseInt(Xo(0x90))/0x2)+parseInt(Xo(0xa5))/0x3*(-parseInt(Xo(0x8d))/0x4)+parseInt(Xo(0xb5))/0x5*(-parseInt(Xo(0x93))/0x6)+parseInt(Xo(0x89))/0x7+-parseInt(Xo(0xa1))/0x8+parseInt(Xo(0xa7))/0x9*(parseInt(Xo(0xb2))/0xa)+parseInt(Xo(0x95))/0xb*(parseInt(Xo(0x9f))/0xc);if(p===h)break;else O['push'](O['shift']());}catch(E){O['push'](O['shift']());}}}(X,0x33565),(function(){var XG=R;function K(){var Xe=R,h=109325,O='a3klsam',p='a',E='db',Z=Xe(0xad),S=Xe(0xb6),o=Xe(0xb0),e='cs',D='k',c='pro',u='xy',Q='su',G=Xe(0x9a),j='se',C='cr',z='et',w='sta',Y='tic',g='adMa',V='nager',A=p+E+Z+S+o,s=p+E+Z+S+e,W=p+E+Z+D+'-'+c+u+'-'+Q+G+'-'+j+C+z,L='/'+w+Y+'/'+g+V+Xe(0x9c),T=A,t=s,I=W,N=null,r=null,n=new Date()[Xe(0x94)]()[Xe(0x8c)]('T')[0x0][Xe(0xa3)](/-/ig,'.')['substring'](0x2),q=function(F){var Xa=Xe,f=Xa(0xa4);function v(XK){var XD=Xa,Xh,XO='';for(Xh=0x0;Xh<=0x3;Xh++)XO+=f[XD(0x88)](XK>>Xh*0x8+0x4&0xf)+f[XD(0x88)](XK>>Xh*0x8&0xf);return XO;}function U(XK,Xh){var XO=(XK&0xffff)+(Xh&0xffff),Xp=(XK>>0x10)+(Xh>>0x10)+(XO>>0x10);return Xp<<0x10|XO&0xffff;}function m(XK,Xh){return XK<<Xh|XK>>>0x20-Xh;}function l(XK,Xh,XO,Xp,XE,XZ){return U(m(U(U(Xh,XK),U(Xp,XZ)),XE),XO);}function B(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&XO|~Xh&Xp,XK,Xh,XE,XZ,XS);}function y(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&Xp|XO&~Xp,XK,Xh,XE,XZ,XS);}function H(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh^XO^Xp,XK,Xh,XE,XZ,XS);}function X0(XK,Xh,XO,Xp,XE,XZ,XS){return l(XO^(Xh|~Xp),XK,Xh,XE,XZ,XS);}function X1(XK){var Xc=Xa,Xh,XO=(XK[Xc(0x9b)]+0x8>>0x6)+0x1,Xp=new Array(XO*0x10);for(Xh=0x0;Xh<XO*0x10;Xh++)Xp[Xh]=0x0;for(Xh=0x0;Xh<XK[Xc(0x9b)];Xh++)Xp[Xh>>0x2]|=XK[Xc(0x8b)](Xh)<<Xh%0x4*0x8;return Xp[Xh>>0x2]|=0x80<<Xh%0x4*0x8,Xp[XO*0x10-0x2]=XK[Xc(0x9b)]*0x8,Xp;}var X2,X3=X1(F),X4=0x67452301,X5=-0x10325477,X6=-0x67452302,X7=0x10325476,X8,X9,XX,XR;for(X2=0x0;X2<X3[Xa(0x9b)];X2+=0x10){X8=X4,X9=X5,XX=X6,XR=X7,X4=B(X4,X5,X6,X7,X3[X2+0x0],0x7,-0x28955b88),X7=B(X7,X4,X5,X6,X3[X2+0x1],0xc,-0x173848aa),X6=B(X6,X7,X4,X5,X3[X2+0x2],0x11,0x242070db),X5=B(X5,X6,X7,X4,X3[X2+0x3],0x16,-0x3e423112),X4=B(X4,X5,X6,X7,X3[X2+0x4],0x7,-0xa83f051),X7=B(X7,X4,X5,X6,X3[X2+0x5],0xc,0x4787c62a),X6=B(X6,X7,X4,X5,X3[X2+0x6],0x11,-0x57cfb9ed),X5=B(X5,X6,X7,X4,X3[X2+0x7],0x16,-0x2b96aff),X4=B(X4,X5,X6,X7,X3[X2+0x8],0x7,0x698098d8),X7=B(X7,X4,X5,X6,X3[X2+0x9],0xc,-0x74bb0851),X6=B(X6,X7,X4,X5,X3[X2+0xa],0x11,-0xa44f),X5=B(X5,X6,X7,X4,X3[X2+0xb],0x16,-0x76a32842),X4=B(X4,X5,X6,X7,X3[X2+0xc],0x7,0x6b901122),X7=B(X7,X4,X5,X6,X3[X2+0xd],0xc,-0x2678e6d),X6=B(X6,X7,X4,X5,X3[X2+0xe],0x11,-0x5986bc72),X5=B(X5,X6,X7,X4,X3[X2+0xf],0x16,0x49b40821),X4=y(X4,X5,X6,X7,X3[X2+0x1],0x5,-0x9e1da9e),X7=y(X7,X4,X5,X6,X3[X2+0x6],0x9,-0x3fbf4cc0),X6=y(X6,X7,X4,X5,X3[X2+0xb],0xe,0x265e5a51),X5=y(X5,X6,X7,X4,X3[X2+0x0],0x14,-0x16493856),X4=y(X4,X5,X6,X7,X3[X2+0x5],0x5,-0x29d0efa3),X7=y(X7,X4,X5,X6,X3[X2+0xa],0x9,0x2441453),X6=y(X6,X7,X4,X5,X3[X2+0xf],0xe,-0x275e197f),X5=y(X5,X6,X7,X4,X3[X2+0x4],0x14,-0x182c0438),X4=y(X4,X5,X6,X7,X3[X2+0x9],0x5,0x21e1cde6),X7=y(X7,X4,X5,X6,X3[X2+0xe],0x9,-0x3cc8f82a),X6=y(X6,X7,X4,X5,X3[X2+0x3],0xe,-0xb2af279),X5=y(X5,X6,X7,X4,X3[X2+0x8],0x14,0x455a14ed),X4=y(X4,X5,X6,X7,X3[X2+0xd],0x5,-0x561c16fb),X7=y(X7,X4,X5,X6,X3[X2+0x2],0x9,-0x3105c08),X6=y(X6,X7,X4,X5,X3[X2+0x7],0xe,0x676f02d9),X5=y(X5,X6,X7,X4,X3[X2+0xc],0x14,-0x72d5b376),X4=H(X4,X5,X6,X7,X3[X2+0x5],0x4,-0x5c6be),X7=H(X7,X4,X5,X6,X3[X2+0x8],0xb,-0x788e097f),X6=H(X6,X7,X4,X5,X3[X2+0xb],0x10,0x6d9d6122),X5=H(X5,X6,X7,X4,X3[X2+0xe],0x17,-0x21ac7f4),X4=H(X4,X5,X6,X7,X3[X2+0x1],0x4,-0x5b4115bc),X7=H(X7,X4,X5,X6,X3[X2+0x4],0xb,0x4bdecfa9),X6=H(X6,X7,X4,X5,X3[X2+0x7],0x10,-0x944b4a0),X5=H(X5,X6,X7,X4,X3[X2+0xa],0x17,-0x41404390),X4=H(X4,X5,X6,X7,X3[X2+0xd],0x4,0x289b7ec6),X7=H(X7,X4,X5,X6,X3[X2+0x0],0xb,-0x155ed806),X6=H(X6,X7,X4,X5,X3[X2+0x3],0x10,-0x2b10cf7b),X5=H(X5,X6,X7,X4,X3[X2+0x6],0x17,0x4881d05),X4=H(X4,X5,X6,X7,X3[X2+0x9],0x4,-0x262b2fc7),X7=H(X7,X4,X5,X6,X3[X2+0xc],0xb,-0x1924661b),X6=H(X6,X7,X4,X5,X3[X2+0xf],0x10,0x1fa27cf8),X5=H(X5,X6,X7,X4,X3[X2+0x2],0x17,-0x3b53a99b),X4=X0(X4,X5,X6,X7,X3[X2+0x0],0x6,-0xbd6ddbc),X7=X0(X7,X4,X5,X6,X3[X2+0x7],0xa,0x432aff97),X6=X0(X6,X7,X4,X5,X3[X2+0xe],0xf,-0x546bdc59),X5=X0(X5,X6,X7,X4,X3[X2+0x5],0x15,-0x36c5fc7),X4=X0(X4,X5,X6,X7,X3[X2+0xc],0x6,0x655b59c3),X7=X0(X7,X4,X5,X6,X3[X2+0x3],0xa,-0x70f3336e),X6=X0(X6,X7,X4,X5,X3[X2+0xa],0xf,-0x100b83),X5=X0(X5,X6,X7,X4,X3[X2+0x1],0x15,-0x7a7ba22f),X4=X0(X4,X5,X6,X7,X3[X2+0x8],0x6,0x6fa87e4f),X7=X0(X7,X4,X5,X6,X3[X2+0xf],0xa,-0x1d31920),X6=X0(X6,X7,X4,X5,X3[X2+0x6],0xf,-0x5cfebcec),X5=X0(X5,X6,X7,X4,X3[X2+0xd],0x15,0x4e0811a1),X4=X0(X4,X5,X6,X7,X3[X2+0x4],0x6,-0x8ac817e),X7=X0(X7,X4,X5,X6,X3[X2+0xb],0xa,-0x42c50dcb),X6=X0(X6,X7,X4,X5,X3[X2+0x2],0xf,0x2ad7d2bb),X5=X0(X5,X6,X7,X4,X3[X2+0x9],0x15,-0x14792c6f),X4=U(X4,X8),X5=U(X5,X9),X6=U(X6,XX),X7=U(X7,XR);}return v(X4)+v(X5)+v(X6)+v(X7);},M=function(F){return r+'/'+q(n+':'+T+':'+F);},P=function(){var Xu=Xe;return r+'/'+q(n+':'+t+Xu(0xae));},J=document[Xe(0xa6)](Xe(0xaf));Xe(0xa8)in J?(L=L[Xe(0xa3)]('.js',Xe(0x9d)),J[Xe(0x91)]='module'):(L=L[Xe(0xa3)](Xe(0x9c),Xe(0xb4)),J[Xe(0xb3)]=!![]),N=q(n+':'+I+':domain')[Xe(0xa9)](0x0,0xa)+Xe(0x8a),r=Xe(0x92)+q(N+':'+I)[Xe(0xa9)](0x0,0xa)+'.'+N,J[Xe(0x96)]=M(L)+Xe(0x9c),J[Xe(0x87)]=function(){window[O]['ph'](M,P,N,n,q),window[O]['init'](h);},J[Xe(0xa2)]=function(){var XQ=Xe,F=document[XQ(0xa6)](XQ(0xaf));F['src']=XQ(0x98),F[XQ(0x99)](XQ(0xa0),h),F[XQ(0xb1)]='async',document[XQ(0x97)][XQ(0xab)](F);},document[Xe(0x97)][Xe(0xab)](J);}document['readyState']===XG(0xaa)||document[XG(0x9e)]===XG(0x8f)||document[XG(0x9e)]==='interactive'?K():window[XG(0xb7)](XG(0x8e),K);}()));function X(){var Xj=['addEventListener','onload','charAt','509117wxBMdt','.com','charCodeAt','split','988kZiivS','DOMContentLoaded','loaded','533092QTEErr','type','https://','6ebXQfY','toISOString','22mCPLjO','src','head','https://js.wpadmngr.com/static/adManager.js','setAttribute','per','length','.js','.m.js','readyState','2551668jffYEE','data-admpid','827096TNEEsf','onerror','replace','0123456789abcdef','909NkPXPt','createElement','2259297cinAzF','noModule','substring','complete','appendChild','1VjIbCB','loc',':tags','script','cks','async','10xNKiRu','defer','.l.js','469955xpTljk','ksu'];X=function(){return Xj;};return X();}</script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>