<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>ncbi-datasets-cli-高效便捷下载NCBI数据 - 编程爱好者博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="ncbi-datasets-cli-高效便捷下载NCBI数据" />
<meta property="og:description" content="文章目录 简介安装```datasets download```下载基因组/基因序列按照GCA list文件编号下载下载大基因组genome完整参数gene参数 ```datasets summary```下载元数据```dataformat```将json转换成表格格式通过json文件解析其他字段问题 简介 NCBI Datasets 可以轻松从 NCBI 数据库中收集数据。使用命令行界面（CLI）工具或 NCBI Datasets 网页界面查找和下载基因和基因组的序列、注释和元数据。如下是可用的工具：
安装 使用conda安装Datasets CLI tools, datasetsand dataformat: # 注意不是datasets而是ncbi-datasets-cli $ conda install -c conda-forge ncbi-datasets-cli (base) [yut@io02 ~]$ datasets --version datasets version: 15.25.0 datasets download下载基因组/基因序列 datasets从 NCBI 下载所有生命领域的生物序列数据，dataformat将前者下载的数据包中的元数据从 JSON Lines 格式转换为其他格式。
使用datasets下载人类参考基因组 GRCh38 的基因组数据包：
$ datasets download genome taxon human --reference --filename human-reference.zip 使用 dataformat从下载的人类参考基因组 GRCh38 数据包中提取选定的元数据字段：
$ dataformat tsv genome --package human-reference.zip --fields organism-name,assminfo-name,accession,assminfo-submitter Organism name	Assembly Name	Assembly Accession	Assembly Submitter Homo sapiens	GRCh38." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://bchobby.github.io/posts/a7654ad7f3ca29d9176408bb7e306825/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-11-16T17:42:48+08:00" />
<meta property="article:modified_time" content="2023-11-16T17:42:48+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程爱好者博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程爱好者博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">ncbi-datasets-cli-高效便捷下载NCBI数据</h1>
			
		</header>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-atom-one-light">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <p></p> 
<div class="toc"> 
 <h4>文章目录</h4> 
 <ul><li><a href="#_1" rel="nofollow">简介</a></li><li><a href="#_5" rel="nofollow">安装</a></li><li><a href="#datasets_download_14" rel="nofollow">```datasets download```下载基因组/基因序列</a></li><li><ul><li><a href="#GCA_list_30" rel="nofollow">按照GCA list文件编号下载</a></li><li><a href="#_54" rel="nofollow">下载大基因组</a></li><li><a href="#genome_71" rel="nofollow">genome完整参数</a></li><li><a href="#gene_129" rel="nofollow">gene参数</a></li></ul> 
  </li><li><a href="#datasets_summary_175" rel="nofollow">```datasets summary```下载元数据</a></li><li><a href="#dataformatjson_221" rel="nofollow">```dataformat```将json转换成表格格式</a></li><li><a href="#json_342" rel="nofollow">通过json文件解析其他字段</a></li><li><a href="#_412" rel="nofollow">问题</a></li></ul> 
</div> 
<p></p> 
<h2><a id="_1"></a>简介</h2> 
<p>NCBI Datasets 可以轻松从 NCBI 数据库中收集数据。使用命令行界面（CLI）工具或 NCBI Datasets 网页界面查找和下载基因和基因组的序列、注释和元数据。如下是可用的工具：<br> <img src="https://images2.imgbox.com/1e/51/ix63o07T_o.png" alt="在这里插入图片描述"></p> 
<h2><a id="_5"></a>安装</h2> 
<ul><li>使用conda安装Datasets CLI tools, <code>datasets</code>and d<code>ataformat</code>:</li></ul> 
<pre><code class="prism language-bash"><span class="token comment"># 注意不是datasets而是ncbi-datasets-cli</span>
$ conda <span class="token function">install</span> <span class="token parameter variable">-c</span> conda-forge ncbi-datasets-cli
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ datasets <span class="token parameter variable">--version</span>
datasets version: <span class="token number">15.25</span>.0
</code></pre> 
<h2><a id="datasets_download_14"></a><code>datasets download</code>下载基因组/基因序列</h2> 
<p><code>datasets</code>从 NCBI 下载所有生命领域的生物序列数据，<code>dataformat</code>将前者下载的数据包中的元数据从 JSON Lines 格式转换为其他格式。</p> 
<p>使用<code>datasets</code>下载人类参考基因组 GRCh38 的基因组数据包：</p> 
<pre><code class="prism language-bash">$ datasets download genome taxon human <span class="token parameter variable">--reference</span> <span class="token parameter variable">--filename</span> human-reference.zip

</code></pre> 
<p>使用 <code>dataformat</code>从下载的人类参考基因组 GRCh38 数据包中提取选定的元数据字段：</p> 
<pre><code class="prism language-bash">$ dataformat tsv genome <span class="token parameter variable">--package</span> human-reference.zip <span class="token parameter variable">--fields</span> organism-name,assminfo-name,accession,assminfo-submitter
Organism name	Assembly Name	Assembly Accession	Assembly Submitter
Homo sapiens	GRCh38.p14	GCF_000001405.40	Genome Reference Consortium
</code></pre> 
<h3><a id="GCA_list_30"></a>按照GCA list文件编号下载</h3> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 02_Glacier_new_taxa<span class="token punctuation">]</span>$ <span class="token function">head</span> <span class="token number">3</span>.gca
GCF_020042285.1
GCF_020783315.1
GCF_024343615.1
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 02_Glacier_new_taxa<span class="token punctuation">]</span>$ <span class="token function">time</span> datasets download genome accession <span class="token parameter variable">--inputfile</span> <span class="token number">3</span>.gca <span class="token parameter variable">--include</span> gff3,rna,cds,protein,genome,seq-report <span class="token parameter variable">--filename</span>  3genome.zip
<span class="token comment"># --inputfile:输入GCA号的list，每行一个</span>
<span class="token comment"># --filename：输出zip包名称，默认ncbi-dataset.zip</span>

New version of client <span class="token punctuation">(</span><span class="token number">15.27</span>.1<span class="token punctuation">)</span> available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets
Collecting <span class="token number">3</span> genome records <span class="token punctuation">[</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token punctuation">]</span> <span class="token number">100</span>% <span class="token number">3</span>/3
Downloading: 3genome.zip    <span class="token number">10</span>.4MB valid <span class="token function">zip</span> archive
Validating package files <span class="token punctuation">[</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token punctuation">]</span> <span class="token number">100</span>% <span class="token number">18</span>/18

real    0m8.208s
user    0m0.652s
sys     0m0.234s
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 02_Glacier_new_taxa<span class="token punctuation">]</span>$ <span class="token function">ls</span>
3genome.zip  download.log 
</code></pre> 
<h3><a id="_54"></a>下载大基因组</h3> 
<p>下载大量基因组，首先下载压缩包，然后分三步访问数据。</p> 
<ul><li>1.下载人基因组压缩包</li></ul> 
<pre><code class="prism language-bash">datasets download genome accession GCF_000001405.40 <span class="token parameter variable">--dehydrated</span> <span class="token parameter variable">--filename</span> human_GRCh38_dataset.zip
</code></pre> 
<ul><li>2.解压</li></ul> 
<pre><code class="prism language-bash"><span class="token function">unzip</span> human_GRCh38_dataset.zip <span class="token parameter variable">-d</span> my_human_dataset
</code></pre> 
<ul><li>3.转换格式</li></ul> 
<pre><code class="prism language-bash">datasets rehydrate <span class="token parameter variable">--directory</span> my_human_dataset/
</code></pre> 
<h3><a id="genome_71"></a>genome完整参数</h3> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ datasets download genome  <span class="token parameter variable">--help</span>

Download a genome data package. Genome data packages may include genome, transcript and protein sequences, annotation and one or <span class="token function">more</span> data reports. Data packages are downloaded as a <span class="token function">zip</span> archive.

The default genome data package includes the following files:
  * <span class="token operator">&lt;</span>accession<span class="token operator">&gt;</span>_<span class="token operator">&lt;</span>assembly_name<span class="token operator">&gt;</span>_genomic.fna <span class="token punctuation">(</span>genomic sequences<span class="token punctuation">)</span>
  * assembly_data_report.jsonl <span class="token punctuation">(</span>data report with genome assembly and annotation metadata<span class="token punctuation">)</span>
  * dataset_catalog.json <span class="token punctuation">(</span>a list of files and <span class="token function">file</span> types included <span class="token keyword">in</span> the data package<span class="token punctuation">)</span>

Usage
  datasets download genome <span class="token punctuation">[</span>flags<span class="token punctuation">]</span>
  datasets download genome <span class="token punctuation">[</span>command<span class="token punctuation">]</span>

Sample Commands
  datasets download genome accession GCF_000001405.40 <span class="token parameter variable">--chromosomes</span> X,Y <span class="token parameter variable">--include</span> genome,gff3,rna
  datasets download genome taxon <span class="token string">"bos taurus"</span> <span class="token parameter variable">--dehydrated</span>
  datasets download genome taxon human --assembly-level chromosome,complete <span class="token parameter variable">--dehydrated</span>
  datasets download genome taxon mouse <span class="token parameter variable">--search</span> C57BL/6J <span class="token parameter variable">--search</span> <span class="token string">"Broad Institute"</span> <span class="token parameter variable">--dehydrated</span>

Available Commands
  accession   Download a genome data package by Assembly or BioProject accession
  taxon       Download a genome data package by taxon <span class="token punctuation">(</span>NCBI Taxonomy ID, scientific or common name at any tax rank<span class="token punctuation">)</span>

Flags
      <span class="token parameter variable">--annotated</span>                Limit to annotated genomes
      --assembly-level string    Limit to genomes at one or <span class="token function">more</span> assembly levels <span class="token punctuation">(</span>comma-separated<span class="token punctuation">)</span>:
                                   * chromosome
                                   * complete
                                   * contig
                                   * scaffold
                                    <span class="token punctuation">(</span>default <span class="token string">"[]"</span><span class="token punctuation">)</span>
      --assembly-source string   Limit to <span class="token string">'RefSeq'</span> <span class="token punctuation">(</span>GCF_<span class="token punctuation">)</span> or <span class="token string">'GenBank'</span> <span class="token punctuation">(</span>GCA_<span class="token punctuation">)</span> genomes <span class="token punctuation">(</span>default <span class="token string">"all"</span><span class="token punctuation">)</span>
      <span class="token parameter variable">--chromosomes</span> strings      Limit to a specified, comma-delimited list of chromosomes, or <span class="token string">'all'</span> <span class="token keyword">for</span> all chromosomes
      <span class="token parameter variable">--dehydrated</span>               Download a dehydrated <span class="token function">zip</span> archive including the data report and locations of data files <span class="token punctuation">(</span>use the rehydrate <span class="token builtin class-name">command</span> to retrieve data files<span class="token punctuation">)</span>.
      --exclude-atypical         Exclude atypical assemblies
      <span class="token parameter variable">--mag</span> string               Limit to metagenome assembled genomes <span class="token punctuation">(</span>only<span class="token punctuation">)</span> or remove them from the results <span class="token punctuation">(</span>exclude<span class="token punctuation">)</span> <span class="token punctuation">(</span>default <span class="token string">"all"</span><span class="token punctuation">)</span>
      <span class="token parameter variable">--preview</span>                  Show information about the requested data package
      <span class="token parameter variable">--reference</span>                Limit to reference genomes
      --released-after string    Limit to genomes released on or after a specified <span class="token function">date</span> <span class="token punctuation">(</span>MM/DD/YYYY<span class="token punctuation">)</span>
      --released-before string   Limit to genomes released on or before a specified <span class="token function">date</span> <span class="token punctuation">(</span>MM/DD/YYYY<span class="token punctuation">)</span>
      <span class="token parameter variable">--search</span> strings           Limit results to genomes with specified text <span class="token keyword">in</span> the searchable fields:
                                 species and infraspecies, assembly name and submitter.
                                 To search multiple strings, use the flag multiple times.


Global Flags
      --api-key string    Specify an NCBI API key
      <span class="token parameter variable">--debug</span>             Emit debugging info
      <span class="token parameter variable">--filename</span> string   Specify a custom <span class="token function">file</span> name <span class="token keyword">for</span> the downloaded data package <span class="token punctuation">(</span>default <span class="token string">"ncbi_dataset.zip"</span><span class="token punctuation">)</span>
      <span class="token parameter variable">--help</span>              Print detailed <span class="token builtin class-name">help</span> about a datasets <span class="token builtin class-name">command</span>
      --no-progressbar    Hide progress bar
      <span class="token parameter variable">--version</span>           Print version of datasets

Use datasets download genome <span class="token operator">&lt;</span>command<span class="token operator">&gt;</span> <span class="token parameter variable">--help</span> <span class="token keyword">for</span> detailed <span class="token builtin class-name">help</span> about a command.

</code></pre> 
<h3><a id="gene_129"></a>gene参数</h3> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ datasets download gene <span class="token parameter variable">--help</span>

Download a gene data package.  Gene data packages include gene, transcript and protein sequences and one or <span class="token function">more</span> data reports. Data packages are downloaded as a <span class="token function">zip</span> archive.

The default gene data package <span class="token keyword">for</span> NM, NR, NP, XM, XR, XP and YP accessions:
  * rna.fna <span class="token punctuation">(</span>transcript sequences<span class="token punctuation">)</span>
  * protein.faa <span class="token punctuation">(</span>protein sequences<span class="token punctuation">)</span>
  * data_report.jsonl <span class="token punctuation">(</span>data report with gene metadata<span class="token punctuation">)</span>
  * dataset_catalog.json <span class="token punctuation">(</span>a list of files and <span class="token function">file</span> types included <span class="token keyword">in</span> the data package<span class="token punctuation">)</span>

Usage
  datasets download gene <span class="token punctuation">[</span>flags<span class="token punctuation">]</span>
  datasets download gene <span class="token punctuation">[</span>command<span class="token punctuation">]</span>

Sample Commands
  datasets download gene gene-id <span class="token number">672</span>
  datasets download gene symbol brca1 <span class="token parameter variable">--taxon</span> mouse
  datasets download gene accession NP_000483.3
  datasets download gene gene-id <span class="token number">2778</span> --fasta-filter NC_000020.11,NM_001077490.3,NP_001070958.1

Available Commands
  gene-id     Download a gene data package by NCBI Gene ID
  symbol      Download a gene data package by gene symbol
  accession   Download a gene data package by RefSeq nucleotide or protein accession
  taxon       Download a gene data package by taxon <span class="token punctuation">(</span>NCBI Taxonomy ID, scientific or common name at any tax rank<span class="token punctuation">)</span>

Flags
      --fasta-filter strings       Limit protein and RNA sequence files to the specified RefSeq nucleotide and protein accessions
      --fasta-filter-file string   Limit protein and RNA sequence files to the specified RefSeq nucleotide and protein accessions included <span class="token keyword">in</span> the specified <span class="token function">file</span>
      <span class="token parameter variable">--preview</span>                    Show information about the requested data package


Global Flags
      --api-key string    Specify an NCBI API key
      <span class="token parameter variable">--debug</span>             Emit debugging info
      <span class="token parameter variable">--filename</span> string   Specify a custom <span class="token function">file</span> name <span class="token keyword">for</span> the downloaded data package <span class="token punctuation">(</span>default <span class="token string">"ncbi_dataset.zip"</span><span class="token punctuation">)</span>
      <span class="token parameter variable">--help</span>              Print detailed <span class="token builtin class-name">help</span> about a datasets <span class="token builtin class-name">command</span>
      --no-progressbar    Hide progress bar
      <span class="token parameter variable">--version</span>           Print version of datasets

Use datasets download gene <span class="token operator">&lt;</span>command<span class="token operator">&gt;</span> <span class="token parameter variable">--help</span> <span class="token keyword">for</span> detailed <span class="token builtin class-name">help</span> about a command.

</code></pre> 
<h2><a id="datasets_summary_175"></a><code>datasets summary</code>下载元数据</h2> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ datasets summary <span class="token parameter variable">--help</span>

Print a data report containing gene, genome or virus metadata <span class="token keyword">in</span> JSON format.

Usage
  datasets summary <span class="token punctuation">[</span>flags<span class="token punctuation">]</span>
  datasets summary <span class="token punctuation">[</span>command<span class="token punctuation">]</span>

Sample Commands
  datasets summary genome accession GCF_000001405.40
  datasets summary genome taxon <span class="token string">"mus musculus"</span>
  datasets summary gene gene-id <span class="token number">672</span>
  datasets summary gene symbol brca1 <span class="token parameter variable">--taxon</span> mouse
  datasets summary gene accession NP_000483.3
  datasets summary virus genome accession NC_045512.2
  datasets summary virus genome taxon sars-cov-2 <span class="token parameter variable">--host</span> dog

Available Commands
  gene        Print a summary of a gene dataset
  genome      Print a data report containing genome metadata
  virus       Print a data report containing virus genome metadata

Global Flags
      --api-key string   Specify an NCBI API key
      <span class="token parameter variable">--debug</span>            Emit debugging info
      <span class="token parameter variable">--help</span>             Print detailed <span class="token builtin class-name">help</span> about a datasets <span class="token builtin class-name">command</span>
      <span class="token parameter variable">--version</span>          Print version of datasets

Use datasets summary <span class="token operator">&lt;</span>command<span class="token operator">&gt;</span> <span class="token parameter variable">--help</span> <span class="token keyword">for</span> detailed <span class="token builtin class-name">help</span> about a command.

</code></pre> 
<ul><li>实例</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ datasets summary genome accession GCF_000001405.40
New version of client <span class="token punctuation">(</span><span class="token number">15.27</span>.1<span class="token punctuation">)</span> available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets
<span class="token punctuation">{<!-- --></span><span class="token string">"reports"</span><span class="token builtin class-name">:</span> <span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"accession"</span><span class="token builtin class-name">:</span><span class="token string">"GCF_000001405.40"</span>,<span class="token string">"annotation_info"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"busco"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"busco_lineage"</span><span class="token builtin class-name">:</span><span class="token string">"primates_odb10"</span>,<span class="token string">"busco_ver"</span><span class="token builtin class-name">:</span><span class="token string">"4.1.4"</span>,<span class="token string">"complete"</span>:0.99187225,<span class="token string">"duplicated"</span>:0.007256894,<span class="token string">"fragmented"</span>:0.0015239477,<span class="token string">"missing"</span>:0.0066037737,<span class="token string">"single_copy"</span>:0.9846154,<span class="token string">"total_count"</span><span class="token builtin class-name">:</span><span class="token string">"13780"</span><span class="token punctuation">}</span>,<span class="token string">"method"</span><span class="token builtin class-name">:</span><span class="token string">"Best-placed RefSeq; Gnomon; RefSeqFE; cmsearch; tRNAscan-SE"</span>,<span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"GCF_000001405.40-RS_2023_10"</span>,<span class="token string">"pipeline"</span><span class="token builtin class-name">:</span><span class="token string">"NCBI eukaryotic genome annotation pipeline"</span>,<span class="token string">"provider"</span><span class="token builtin class-name">:</span><span class="token string">"NCBI RefSeq"</span>,<span class="token string">"release_date"</span><span class="token builtin class-name">:</span><span class="token string">"2023-10-02"</span>,<span class="token string">"report_url"</span><span class="token builtin class-name">:</span><span class="token string">"https://www.ncbi.nlm.nih.gov/genome/annotation_euk/Homo_sapiens/GCF_000001405.40-RS_2023_10.html"</span>,<span class="token string">"software_version"</span><span class="token builtin class-name">:</span><span class="token string">"10.2"</span>,<span class="token string">"stats"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"gene_counts"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"non_coding"</span>:22158,<span class="token string">"other"</span>:413,<span class="token string">"protein_coding"</span>:20080,<span class="token string">"pseudogene"</span>:17001,<span class="token string">"total"</span>:59652<span class="token punctuation">}</span><span class="token punctuation">}</span>,<span class="token string">"status"</span><span class="token builtin class-name">:</span><span class="token string">"Updated annotation"</span><span class="token punctuation">}</span>,<span class="token string">"assembly_info"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"assembly_level"</span><span class="token builtin class-name">:</span><span class="token string">"Chromosome"</span>,<span class="token string">"assembly_name"</span><span class="token builtin class-name">:</span><span class="token string">"GRCh38.p14"</span>,<span class="token string">"assembly_status"</span><span class="token builtin class-name">:</span><span class="token string">"current"</span>,<span class="token string">"assembly_type"</span><span class="token builtin class-name">:</span><span class="token string">"haploid-with-alt-loci"</span>,<span class="token string">"bioproject_accession"</span><span class="token builtin class-name">:</span><span class="token string">"PRJNA31257"</span>,<span class="token string">"bioproject_lineage"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"bioprojects"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"accession"</span><span class="token builtin class-name">:</span><span class="token string">"PRJNA31257"</span>,<span class="token string">"title"</span><span class="token builtin class-name">:</span><span class="token string">"The Human Genome Project, currently maintained by the Genome Reference Consortium (GRC)"</span><span class="token punctuation">}</span><span class="token punctuation">]</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"blast_url"</span><span class="token builtin class-name">:</span><span class="token string">"https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastSearch<span class="token entity" title="\u0026">\u0026</span>PROG_DEF=blastn<span class="token entity" title="\u0026">\u0026</span>BLAST_SPEC=GDH_GCF_000001405.40"</span>,<span class="token string">"description"</span><span class="token builtin class-name">:</span><span class="token string">"Genome Reference Consortium Human Build 38 patch release 14 (GRCh38.p14)"</span>,<span class="token string">"paired_assembly"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"accession"</span><span class="token builtin class-name">:</span><span class="token string">"GCA_000001405.29"</span>,<span class="token string">"only_genbank"</span><span class="token builtin class-name">:</span><span class="token string">"4 unlocalized and unplaced scaffolds."</span>,<span class="token string">"status"</span><span class="token builtin class-name">:</span><span class="token string">"current"</span><span class="token punctuation">}</span>,<span class="token string">"refseq_category"</span><span class="token builtin class-name">:</span><span class="token string">"reference genome"</span>,<span class="token string">"release_date"</span><span class="token builtin class-name">:</span><span class="token string">"2022-02-03"</span>,<span class="token string">"submitter"</span><span class="token builtin class-name">:</span><span class="token string">"Genome Reference Consortium"</span>,<span class="token string">"synonym"</span><span class="token builtin class-name">:</span><span class="token string">"hg38"</span><span class="token punctuation">}</span>,<span class="token string">"assembly_stats"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"contig_l50"</span>:18,<span class="token string">"contig_n50"</span>:57879411,<span class="token string">"gaps_between_scaffolds_count"</span>:349,<span class="token string">"gc_count"</span><span class="token builtin class-name">:</span><span class="token string">"1374283647"</span>,<span class="token string">"gc_percent"</span>:41,<span class="token string">"number_of_component_sequences"</span>:35611,<span class="token string">"number_of_contigs"</span>:996,<span class="token string">"number_of_organelles"</span>:1,<span class="token string">"number_of_scaffolds"</span>:470,<span class="token string">"scaffold_l50"</span>:16,<span class="token string">"scaffold_n50"</span>:67794873,<span class="token string">"total_number_of_chromosomes"</span>:24,<span class="token string">"total_sequence_length"</span><span class="token builtin class-name">:</span><span class="token string">"3099441038"</span>,<span class="token string">"total_ungapped_length"</span><span class="token builtin class-name">:</span><span class="token string">"2948318359"</span><span class="token punctuation">}</span>,<span class="token string">"current_accession"</span><span class="token builtin class-name">:</span><span class="token string">"GCF_000001405.40"</span>,<span class="token string">"organelle_info"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"description"</span><span class="token builtin class-name">:</span><span class="token string">"Mitochondrion"</span>,<span class="token string">"submitter"</span><span class="token builtin class-name">:</span><span class="token string">"Genome Reference Consortium"</span>,<span class="token string">"total_seq_length"</span><span class="token builtin class-name">:</span><span class="token string">"16569"</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"organism"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"common_name"</span><span class="token builtin class-name">:</span><span class="token string">"human"</span>,<span class="token string">"organism_name"</span><span class="token builtin class-name">:</span><span class="token string">"Homo sapiens"</span>,<span class="token string">"tax_id"</span>:9606<span class="token punctuation">}</span>,<span class="token string">"paired_accession"</span><span class="token builtin class-name">:</span><span class="token string">"GCA_000001405.29"</span>,<span class="token string">"source_database"</span><span class="token builtin class-name">:</span><span class="token string">"SOURCE_DATABASE_REFSEQ"</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"total_count"</span><span class="token builtin class-name">:</span> <span class="token number">1</span><span class="token punctuation">}</span>

<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ datasets summary gene gene-id <span class="token number">672</span>
New version of client <span class="token punctuation">(</span><span class="token number">15.27</span>.1<span class="token punctuation">)</span> available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets
<span class="token punctuation">{<!-- --></span><span class="token string">"reports"</span><span class="token builtin class-name">:</span> <span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"gene"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"annotations"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"annotation_name"</span><span class="token builtin class-name">:</span><span class="token string">"GCF_000001405.40-RS_2023_10"</span>,<span class="token string">"annotation_release_date"</span><span class="token builtin class-name">:</span><span class="token string">"2023-10-02"</span>,<span class="token string">"assembly_accession"</span><span class="token builtin class-name">:</span><span class="token string">"GCF_000001405.40"</span>,<span class="token string">"assembly_name"</span><span class="token builtin class-name">:</span><span class="token string">"GRCh38.p14"</span>,<span class="token string">"genomic_locations"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"genomic_accession_version"</span><span class="token builtin class-name">:</span><span class="token string">"NC_000017.11"</span>,<span class="token string">"genomic_range"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"begin"</span><span class="token builtin class-name">:</span><span class="token string">"43044295"</span>,<span class="token string">"end"</span><span class="token builtin class-name">:</span><span class="token string">"43170327"</span>,<span class="token string">"orientation"</span><span class="token builtin class-name">:</span><span class="token string">"minus"</span><span class="token punctuation">}</span>,<span class="token string">"sequence_name"</span><span class="token builtin class-name">:</span><span class="token string">"17"</span><span class="token punctuation">}</span><span class="token punctuation">]</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"annotation_name"</span><span class="token builtin class-name">:</span><span class="token string">"GCF_009914755.1-RS_2023_10"</span>,<span class="token string">"annotation_release_date"</span><span class="token builtin class-name">:</span><span class="token string">"2023-10-02"</span>,<span class="token string">"assembly_accession"</span><span class="token builtin class-name">:</span><span class="token string">"GCF_009914755.1"</span>,<span class="token string">"assembly_name"</span><span class="token builtin class-name">:</span><span class="token string">"T2T-CHM13v2.0"</span>,<span class="token string">"genomic_locations"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"genomic_accession_version"</span><span class="token builtin class-name">:</span><span class="token string">"NC_060941.1"</span>,<span class="token string">"genomic_range"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"begin"</span><span class="token builtin class-name">:</span><span class="token string">"43902857"</span>,<span class="token string">"end"</span><span class="token builtin class-name">:</span><span class="token string">"44029084"</span>,<span class="token string">"orientation"</span><span class="token builtin class-name">:</span><span class="token string">"minus"</span><span class="token punctuation">}</span>,<span class="token string">"sequence_name"</span><span class="token builtin class-name">:</span><span class="token string">"17"</span><span class="token punctuation">}</span><span class="token punctuation">]</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"chromosomes"</span>:<span class="token punctuation">[</span><span class="token string">"17"</span><span class="token punctuation">]</span>,<span class="token string">"common_name"</span><span class="token builtin class-name">:</span><span class="token string">"human"</span>,<span class="token string">"description"</span><span class="token builtin class-name">:</span><span class="token string">"BRCA1 DNA repair associated"</span>,<span class="token string">"ensembl_gene_ids"</span>:<span class="token punctuation">[</span><span class="token string">"ENSG00000012048"</span><span class="token punctuation">]</span>,<span class="token string">"gene_groups"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"id"</span><span class="token builtin class-name">:</span><span class="token string">"672"</span>,<span class="token string">"method"</span><span class="token builtin class-name">:</span><span class="token string">"NCBI Ortholog"</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"gene_id"</span><span class="token builtin class-name">:</span><span class="token string">"672"</span>,<span class="token string">"nomenclature_authority"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"authority"</span><span class="token builtin class-name">:</span><span class="token string">"HGNC"</span>,<span class="token string">"identifier"</span><span class="token builtin class-name">:</span><span class="token string">"HGNC:1100"</span><span class="token punctuation">}</span>,<span class="token string">"omim_ids"</span>:<span class="token punctuation">[</span><span class="token string">"113705"</span><span class="token punctuation">]</span>,<span class="token string">"orientation"</span><span class="token builtin class-name">:</span><span class="token string">"minus"</span>,<span class="token string">"protein_count"</span>:368,<span class="token string">"reference_standards"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"gene_range"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"accession_version"</span><span class="token builtin class-name">:</span><span class="token string">"NG_005905.2"</span>,<span class="token string">"range"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"begin"</span><span class="token builtin class-name">:</span><span class="token string">"92501"</span>,<span class="token string">"end"</span><span class="token builtin class-name">:</span><span class="token string">"173689"</span>,<span class="token string">"orientation"</span><span class="token builtin class-name">:</span><span class="token string">"plus"</span><span class="token punctuation">}</span><span class="token punctuation">]</span><span class="token punctuation">}</span>,<span class="token string">"type"</span><span class="token builtin class-name">:</span><span class="token string">"REFSEQ_GENE"</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"swiss_prot_accessions"</span>:<span class="token punctuation">[</span><span class="token string">"P38398"</span><span class="token punctuation">]</span>,<span class="token string">"symbol"</span><span class="token builtin class-name">:</span><span class="token string">"BRCA1"</span>,<span class="token string">"synonyms"</span>:<span class="token punctuation">[</span><span class="token string">"IRIS"</span>,<span class="token string">"PSCP"</span>,<span class="token string">"BRCAI"</span>,<span class="token string">"BRCC1"</span>,<span class="token string">"FANCS"</span>,<span class="token string">"PNCA4"</span>,<span class="token string">"RNF53"</span>,<span class="token string">"BROVCA1"</span>,<span class="token string">"PPP1R53"</span><span class="token punctuation">]</span>,<span class="token string">"tax_id"</span><span class="token builtin class-name">:</span><span class="token string">"9606"</span>,<span class="token string">"taxname"</span><span class="token builtin class-name">:</span><span class="token string">"Homo sapiens"</span>,<span class="token string">"transcript_count"</span>:368,<span class="token string">"transcript_type_counts"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"count"</span>:368,<span class="token string">"type"</span><span class="token builtin class-name">:</span><span class="token string">"PROTEIN_CODING"</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"type"</span><span class="token builtin class-name">:</span><span class="token string">"PROTEIN_CODING"</span><span class="token punctuation">}</span>,<span class="token string">"query"</span>:<span class="token punctuation">[</span><span class="token string">"672"</span><span class="token punctuation">]</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"total_count"</span><span class="token builtin class-name">:</span> <span class="token number">1</span><span class="token punctuation">}</span>
</code></pre> 
<ul><li>下载结果为json格式</li></ul> 
<h2><a id="dataformatjson_221"></a><code>dataformat</code>将json转换成表格格式</h2> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ dataformat tsv

Convert data to TSV format.

Refer to NCBI's <span class="token punctuation">[</span>download and install<span class="token punctuation">]</span><span class="token punctuation">(</span>https://www.ncbi.nlm.nih.gov/datasets/docs/v2/download-and-install/<span class="token punctuation">)</span> documentation <span class="token keyword">for</span> information about getting started with the command-line tools.

Usage
  dataformat tsv <span class="token punctuation">[</span>command<span class="token punctuation">]</span>

Report Commands
  genome             Convert Genome Assembly Data Report into TSV <span class="token function">format</span>
  genome-seq         Convert Genome Assembly Sequence Report into TSV <span class="token function">format</span>
  gene               Convert Gene Report into TSV <span class="token function">format</span>
  gene-product       Convert Gene Product Report into TSV <span class="token function">format</span>
  virus-genome       Convert Virus Data Report into TSV <span class="token function">format</span>
  virus-annotation   Convert Virus Annotation Report into TSV <span class="token function">format</span>
  microbigge         Convert MicroBIGG-E Data Report into TSV <span class="token function">format</span>
  prok-gene          Convert Prokaryote Gene Report into TSV <span class="token function">format</span>
  prok-gene-location Convert Prokaryote Gene Location Report into TSV <span class="token function">format</span>
  genome-annotations Convert Genome Annotation Report into TSV <span class="token function">format</span>

Flags
      --elide-header   Do not output header
  -h, <span class="token parameter variable">--help</span>           <span class="token builtin class-name">help</span> <span class="token keyword">for</span> tsv



Global Flags
      <span class="token parameter variable">--force</span>   Force dataformat to run without <span class="token builtin class-name">type</span> check prompt

Use dataformat tsv <span class="token operator">&lt;</span>command<span class="token operator">&gt;</span> <span class="token parameter variable">--help</span> <span class="token keyword">for</span> detailed <span class="token builtin class-name">help</span> about a command.

<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ dataformat tsv gene
Error: <span class="token parameter variable">--inputfile</span> and/or <span class="token parameter variable">--packagefile</span> must be specified, or data can be <span class="token builtin class-name">read</span> from standard input
Usage
  dataformat tsv gene <span class="token punctuation">[</span>flags<span class="token punctuation">]</span>

Examples
  dataformat tsv gene <span class="token parameter variable">--inputfile</span> gene_package/ncbi_dataset/data/data_report.jsonl
  dataformat tsv gene <span class="token parameter variable">--package</span> genes.zip

Flags
      <span class="token parameter variable">--fields</span> strings     Comma-separated list of fields <span class="token punctuation">(</span>default annotation-assembly-accession,annotation-assembly-name,annotation-genomic-range-accession,annotation-genomic-range-exon-order,annotation-genomic-range-exon-orientation,annotation-genomic-range-exon-start,annotation-genomic-range-exon-stop,annotation-genomic-range-range-order,annotation-genomic-range-range-orientation,annotation-genomic-range-range-start,annotation-genomic-range-range-stop,annotation-genomic-range-seq-name,annotation-release-date,annotation-release-name,chromosomes,common-name,description,ensembl-geneids,gene-id,gene-type,genomic-region-gene-range-accession,genomic-region-gene-range-range-order,genomic-region-gene-range-range-orientation,genomic-region-gene-range-range-start,genomic-region-gene-range-range-stop,genomic-region-genomic-region-type,group-id,group-method,name-authority,name-id,omim-ids,orientation,protein-count,ref-standard-gene-range-accession,ref-standard-gene-range-range-order,ref-standard-gene-range-range-orientation,ref-standard-gene-range-range-start,ref-standard-gene-range-range-stop,ref-standard-genomic-region-type,replaced-gene-id,rna-type,swissprot-accessions,symbol,synonyms,tax-id,tax-name,transcript-count<span class="token punctuation">)</span>
                               - annotation-assembly-accession
                               - annotation-assembly-name
                               - annotation-genomic-range-accession
                               - annotation-genomic-range-exon-order
                               - annotation-genomic-range-exon-orientation
                               - annotation-genomic-range-exon-start
                               - annotation-genomic-range-exon-stop
                               - annotation-genomic-range-range-order
                               - annotation-genomic-range-range-orientation
                               - annotation-genomic-range-range-start
                               - annotation-genomic-range-range-stop
                               - annotation-genomic-range-seq-name
                               - annotation-release-date
                               - annotation-release-name
                               - chromosomes
                               - common-name
                               - description
                               - ensembl-geneids
                               - gene-id
                               - gene-type
                               - genomic-region-gene-range-accession
                               - genomic-region-gene-range-range-order
                               - genomic-region-gene-range-range-orientation
                               - genomic-region-gene-range-range-start
                               - genomic-region-gene-range-range-stop
                               - genomic-region-genomic-region-type
                               - group-id
                               - group-method
                               - name-authority
                               - name-id
                               - omim-ids
                               - orientation
                               - protein-count
                               - ref-standard-gene-range-accession
                               - ref-standard-gene-range-range-order
                               - ref-standard-gene-range-range-orientation
                               - ref-standard-gene-range-range-start
                               - ref-standard-gene-range-range-stop
                               - ref-standard-genomic-region-type
                               - replaced-gene-id
                               - rna-type
                               - swissprot-accessions
                               - symbol
                               - synonyms
                               - tax-id
                               - tax-name
                               - transcript-count
  -h, <span class="token parameter variable">--help</span>               <span class="token builtin class-name">help</span> <span class="token keyword">for</span> gene
      <span class="token parameter variable">--inputfile</span> string   Input <span class="token function">file</span> <span class="token punctuation">(</span>default <span class="token string">"ncbi_dataset/data/data_report.jsonl"</span><span class="token punctuation">)</span>
      <span class="token parameter variable">--package</span> string     Data package <span class="token punctuation">(</span>zip archive<span class="token punctuation">)</span>, inputfile parameter is relative to the root path inside the archive



Global Flags
      --elide-header   Do not output header
      <span class="token parameter variable">--force</span>          Force dataformat to run without <span class="token builtin class-name">type</span> check prompt

</code></pre> 
<ul><li>实例</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ datasets summary gene gene-id <span class="token number">672</span>  --as-json-lines <span class="token operator">|</span>dataformat tsv gene
New version of client <span class="token punctuation">(</span><span class="token number">15.27</span>.1<span class="token punctuation">)</span> available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets
Annotation Assembly Accession   Annotation Assembly Name        Annotation Genomic Range Accession      Annotation Genomic Range Exons Order    Annotation Genomic Range Exons Orientation      Annotation Genomic Range Exons Start    Annotation Genomic Range Exons Stop     Annotation Genomic Range Order        Annotation Genomic Range Orientation    Annotation Genomic Range Start  Annotation Genomic Range Stop   Annotation Genomic Range Seq Name       Annotation Release Date Annotation Release Name Chromosomes     Common Name     Description     Ensembl GeneIDs NCBI GeneID   Gene Type       Genomic Region Gene Range Sequence Accession    Genomic Region Gene Range Order Genomic Region Gene Range Orientation   Genomic Region Gene Range Start Genomic Region Gene Range Stop  Genomic Region Genomic Region Type      Gene Group Identifier   Gene Group Method     Nomenclature Authority  Nomenclature ID OMIM IDs        Orientation     Proteins        Reference Standard Gene Range Sequence Accession        Reference Standard Gene Range Order     Reference Standard Gene Range Orientation       Reference Standard Gene Range Start     Reference Standard Gene Range Stop    Reference Standard Genomic Region Type  Replaced NCBI GeneID    RNA Type        SwissProt Accessions    Symbol  Synonyms        Taxonomic ID    Taxonomic Name  Transcripts
GCF_000001405.40        GRCh38.p14      NC_000017.11                                            minus   <span class="token number">43044295</span>        <span class="token number">43170327</span>        <span class="token number">17</span>      <span class="token number">2023</span>-10-02      GCF_000001405.40-RS_2023_10     <span class="token number">17</span>      human   BRCA1 DNA repair associated     ENSG00000012048 <span class="token number">672</span>     PROTEIN_CODING       <span class="token number">672</span>      NCBI Ortholog   HGNC    HGNC:1100       <span class="token number">113705</span>  minus   <span class="token number">368</span>     NG_005905.2             plus    <span class="token number">92501</span>   <span class="token number">173689</span>  REFSEQ_GENE                     P38398  BRCA1   IRIS,PSCP,BRCAI,BRCC1,FANCS,PNCA4,RNF53,BROVCA1,PPP1R53 <span class="token number">9606</span>    Homo sapiens    <span class="token number">368</span>
GCF_009914755.1 T2T-CHM13v2.0   NC_060941.1                                             minus   <span class="token number">43902857</span>        <span class="token number">44029084</span>        <span class="token number">17</span>      <span class="token number">2023</span>-10-02      GCF_009914755.1-RS_2023_10      <span class="token number">17</span>      human   BRCA1 DNA repair associated     ENSG00000012048 <span class="token number">672</span>     PROTEIN_CODING               <span class="token number">672</span>      NCBI Ortholog   HGNC    HGNC:1100       <span class="token number">113705</span>  minus   <span class="token number">368</span>     NG_005905.2             plus    <span class="token number">92501</span>   <span class="token number">173689</span>  REFSEQ_GENE                     P38398  BRCA1   IRIS,PSCP,BRCAI,BRCC1,FANCS,PNCA4,RNF53,BROVCA1,PPP1R53 <span class="token number">9606</span>    Homo sapiens    <span class="token number">368</span>

<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ datasets summary gene gene-id <span class="token number">672</span>  --as-json-lines <span class="token operator">|</span>dataformat tsv gene <span class="token parameter variable">--fields</span> gene-id,gene-type,symbol
New version of client <span class="token punctuation">(</span><span class="token number">15.27</span>.1<span class="token punctuation">)</span> available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets
NCBI GeneID     Gene Type       Symbol
<span class="token number">672</span>     PROTEIN_CODING  BRCA1

<span class="token comment"># --as-json-lines必须加上</span>
<span class="token comment"># --fields指定需要的字段，多个空格隔开</span>
</code></pre> 
<h2><a id="json_342"></a>通过json文件解析其他字段</h2> 
<ul><li>某些字段无法通过<code>dataformat</code>提取出来，可先保存成json文件，然后通过下面脚本解析：</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 ~<span class="token punctuation">]</span>$ <span class="token function">cat</span> dataset.json
<span class="token punctuation">{<!-- --></span><span class="token string">"accession"</span><span class="token builtin class-name">:</span><span class="token string">"GCA_013141435.1"</span>,<span class="token string">"annotation_info"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"method"</span><span class="token builtin class-name">:</span><span class="token string">"Best-placed reference protein set; GeneMarkS-2+"</span>,<span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"NCBI Prokaryotic Genome Annotation Pipeline (PGAP)"</span>,<span class="token string">"pipeline"</span><span class="token builtin class-name">:</span><span class="token string">"NCBI Prokaryotic Genome Annotation Pipeline (PGAP)"</span>,<span class="token string">"provider"</span><span class="token builtin class-name">:</span><span class="token string">"NCBI"</span>,<span class="token string">"release_date"</span><span class="token builtin class-name">:</span><span class="token string">"2020-05-14"</span>,<span class="token string">"software_version"</span><span class="token builtin class-name">:</span><span class="token string">"4.11"</span>,<span class="token string">"stats"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"gene_counts"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"non_coding"</span>:27,<span class="token string">"protein_coding"</span>:2566,<span class="token string">"pseudogene"</span>:17,<span class="token string">"total"</span>:2610<span class="token punctuation">}</span><span class="token punctuation">}</span><span class="token punctuation">}</span>,<span class="token string">"assembly_info"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"assembly_level"</span><span class="token builtin class-name">:</span><span class="token string">"Contig"</span>,<span class="token string">"assembly_method"</span><span class="token builtin class-name">:</span><span class="token string">"MetaSPAdes v. 3.10.1"</span>,<span class="token string">"assembly_name"</span><span class="token builtin class-name">:</span><span class="token string">"ASM1314143v1"</span>,<span class="token string">"assembly_status"</span><span class="token builtin class-name">:</span><span class="token string">"current"</span>,<span class="token string">"assembly_type"</span><span class="token builtin class-name">:</span><span class="token string">"haploid"</span>,<span class="token string">"bioproject_accession"</span><span class="token builtin class-name">:</span><span class="token string">"PRJNA622654"</span>,<span class="token string">"bioproject_lineage"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"bioprojects"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"accession"</span><span class="token builtin class-name">:</span><span class="token string">"PRJNA622654"</span>,<span class="token string">"title"</span><span class="token builtin class-name">:</span><span class="token string">"Metagenomic profiling of ammonia and methane-oxidizing microorganisms in a Dutch drinking water treatment plant"</span><span class="token punctuation">}</span><span class="token punctuation">]</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"biosample"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"accession"</span><span class="token builtin class-name">:</span><span class="token string">"SAMN14539096"</span>,<span class="token string">"attributes"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"isolation_source"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"Primary rapid sand filter"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"collection_date"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"not applicable"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"geo_loc_name"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"Netherlands"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"lat_lon"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"not applicable"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"isolate"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"P-RSF-IL-07"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"depth"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"not applicable"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"env_broad_scale"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"drinking water treatment plant"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"env_local_scale"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"Primary rapid sand filter"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"env_medium"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"not applicable"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"metagenomic"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"1"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"environmental-sample"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"1"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"sample_type"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"metagenomic assembly"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"metagenome-source"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"drinking water metagenome"</span><span class="token punctuation">}</span>,<span class="token punctuation">{<!-- --></span><span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"derived_from"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"This BioSample is a metagenomic assembly obtained from the drinking water metagenome BioSample:SAMN14524263, SAMN14524264, SAMN14524265, SAMN14524266"</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"bioprojects"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"accession"</span><span class="token builtin class-name">:</span><span class="token string">"PRJNA622654"</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"description"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"comment"</span><span class="token builtin class-name">:</span><span class="token string">"Keywords: GSC:MIxS;MIMAG:6.0"</span>,<span class="token string">"organism"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"organism_name"</span><span class="token builtin class-name">:</span><span class="token string">"Ferruginibacter sp."</span>,<span class="token string">"tax_id"</span>:1940288<span class="token punctuation">}</span>,<span class="token string">"title"</span><span class="token builtin class-name">:</span><span class="token string">"MIMAG Metagenome-assembled Genome sample from Ferruginibacter sp."</span><span class="token punctuation">}</span>,<span class="token string">"last_updated"</span><span class="token builtin class-name">:</span><span class="token string">"2020-05-19T00:50:12.857"</span>,<span class="token string">"models"</span>:<span class="token punctuation">[</span><span class="token string">"MIMAG.water"</span><span class="token punctuation">]</span>,<span class="token string">"owner"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"contacts"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"name"</span><span class="token builtin class-name">:</span><span class="token string">"Radboud University"</span><span class="token punctuation">}</span>,<span class="token string">"package"</span><span class="token builtin class-name">:</span><span class="token string">"MIMAG.water.6.0"</span>,<span class="token string">"publication_date"</span><span class="token builtin class-name">:</span><span class="token string">"2020-05-19T00:50:12.857"</span>,<span class="token string">"sample_ids"</span>:<span class="token punctuation">[</span><span class="token punctuation">{<!-- --></span><span class="token string">"label"</span><span class="token builtin class-name">:</span><span class="token string">"Sample name"</span>,<span class="token string">"value"</span><span class="token builtin class-name">:</span><span class="token string">"Ferruginibacter sp. P-RSF-IL-07"</span><span class="token punctuation">}</span><span class="token punctuation">]</span>,<span class="token string">"status"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"status"</span><span class="token builtin class-name">:</span><span class="token string">"live"</span>,<span class="token string">"when"</span><span class="token builtin class-name">:</span><span class="token string">"2020-05-19T00:50:12.857"</span><span class="token punctuation">}</span>,<span class="token string">"submission_date"</span><span class="token builtin class-name">:</span><span class="token string">"2020-04-04T13:17:04.950"</span><span class="token punctuation">}</span>,<span class="token string">"comments"</span><span class="token builtin class-name">:</span><span class="token string">"The annotation was added by the NCBI Prokaryotic Genome Annotation Pipeline (PGAP). Information about PGAP can be found here: https://www.ncbi.nlm.nih.gov/genome/annotation_prok/"</span>,<span class="token string">"genome_notes"</span>:<span class="token punctuation">[</span><span class="token string">"derived from metagenome"</span><span class="token punctuation">]</span>,<span class="token string">"release_date"</span><span class="token builtin class-name">:</span><span class="token string">"2020-05-21"</span>,<span class="token string">"sequencing_tech"</span><span class="token builtin class-name">:</span><span class="token string">"Illumina MiSeq"</span>,<span class="token string">"submitter"</span><span class="token builtin class-name">:</span><span class="token string">"Radboud University"</span><span class="token punctuation">}</span>,<span class="token string">"assembly_stats"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"contig_l50"</span>:10,<span class="token string">"contig_n50"</span>:104094,<span class="token string">"gc_count"</span><span class="token builtin class-name">:</span><span class="token string">"978119"</span>,<span class="token string">"gc_percent"</span>:32,<span class="token string">"genome_coverage"</span><span class="token builtin class-name">:</span><span class="token string">"270.4x"</span>,<span class="token string">"number_of_component_sequences"</span>:43,<span class="token string">"number_of_contigs"</span>:43,<span class="token string">"total_sequence_length"</span><span class="token builtin class-name">:</span><span class="token string">"3056910"</span>,<span class="token string">"total_ungapped_length"</span><span class="token builtin class-name">:</span><span class="token string">"3056910"</span><span class="token punctuation">}</span>,<span class="token string">"average_nucleotide_identity"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"best_ani_match"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"ani"</span>:79.65,<span class="token string">"assembly"</span><span class="token builtin class-name">:</span><span class="token string">"GCA_003426875.1"</span>,<span class="token string">"assembly_coverage"</span>:0.01,<span class="token string">"category"</span><span class="token builtin class-name">:</span><span class="token string">"type"</span>,<span class="token string">"organism_name"</span><span class="token builtin class-name">:</span><span class="token string">"Lutibacter oceani"</span>,<span class="token string">"type_assembly_coverage"</span>:0.01<span class="token punctuation">}</span>,<span class="token string">"category"</span><span class="token builtin class-name">:</span><span class="token string">"category_na"</span>,<span class="token string">"comment"</span><span class="token builtin class-name">:</span><span class="token string">"na"</span>,<span class="token string">"match_status"</span><span class="token builtin class-name">:</span><span class="token string">"low_coverage"</span>,<span class="token string">"submitted_organism"</span><span class="token builtin class-name">:</span><span class="token string">"Ferruginibacter sp."</span>,<span class="token string">"submitted_species"</span><span class="token builtin class-name">:</span><span class="token string">"Ferruginibacter sp."</span>,<span class="token string">"taxonomy_check_status"</span><span class="token builtin class-name">:</span><span class="token string">"Inconclusive"</span><span class="token punctuation">}</span>,<span class="token string">"current_accession"</span><span class="token builtin class-name">:</span><span class="token string">"GCA_013141435.1"</span>,<span class="token string">"organism"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"infraspecific_names"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"isolate"</span><span class="token builtin class-name">:</span><span class="token string">"P-RSF-IL-07"</span><span class="token punctuation">}</span>,<span class="token string">"organism_name"</span><span class="token builtin class-name">:</span><span class="token string">"Ferruginibacter sp."</span>,<span class="token string">"tax_id"</span>:1940288<span class="token punctuation">}</span>,<span class="token string">"source_database"</span><span class="token builtin class-name">:</span><span class="token string">"SOURCE_DATABASE_GENBANK"</span>,<span class="token string">"wgs_info"</span>:<span class="token punctuation">{<!-- --></span><span class="token string">"master_wgs_url"</span><span class="token builtin class-name">:</span><span class="token string">"https://www.ncbi.nlm.nih.gov/nuccore/JABFQZ000000000.1"</span>,<span class="token string">"wgs_contigs_url"</span><span class="token builtin class-name">:</span><span class="token string">"https://www.ncbi.nlm.nih.gov/Traces/wgs/JABFQZ01"</span>,<span class="token string">"wgs_project_accession"</span><span class="token builtin class-name">:</span><span class="token string">"JABFQZ01"</span><span class="token punctuation">}</span><span class="token punctuation">}</span>

<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 ~<span class="token punctuation">]</span>$ Parse_dataset_genome_json_metadata.py  *json
Save result <span class="token keyword">in</span> output.csv
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 ~<span class="token punctuation">]</span>$ <span class="token function">cat</span> output.csv
Accession,Geo Location Name,Latitude and Longitude,Collection date,Env broad scale,Env <span class="token builtin class-name">local</span> scale,Env medium,Sample <span class="token builtin class-name">type</span>
GCA_013141435.1,Netherlands,not applicable,not applicable,drinking water treatment plant,Primary rapid sand filter,not applicable,metagenomic assembly
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 ~<span class="token punctuation">]</span>$ <span class="token function">cat</span> ~/Software/Important_scripts/Parse_dataset_genome_json_metadata.py
<span class="token comment">#!/usr/bin/env python</span>
<span class="token function">import</span> argparse
<span class="token function">import</span> json
<span class="token function">import</span> pandas as pd

<span class="token comment"># 创建参数解析器</span>
parser <span class="token operator">=</span> argparse.ArgumentParser<span class="token punctuation">(</span>description<span class="token operator">=</span><span class="token string">'Parse JSON data'</span><span class="token punctuation">)</span>
parser.add_argument<span class="token punctuation">(</span><span class="token string">'json_file'</span>, <span class="token assign-left variable">help</span><span class="token operator">=</span><span class="token string">'Path to the JSON file'</span><span class="token punctuation">)</span>

<span class="token comment"># 解析参数</span>
args <span class="token operator">=</span> parser.parse_args<span class="token punctuation">(</span><span class="token punctuation">)</span>

<span class="token comment"># 读取JSON文件</span>
with open<span class="token punctuation">(</span>args.json_file, <span class="token string">'r'</span><span class="token punctuation">)</span> as file:
    json_str <span class="token operator">=</span> file.read<span class="token punctuation">(</span><span class="token punctuation">)</span>

<span class="token comment"># 解析JSON</span>
data <span class="token operator">=</span> json.loads<span class="token punctuation">(</span>json_str<span class="token punctuation">)</span>

<span class="token comment"># 获取env_broad_scale字段的值</span>
<span class="token comment"># 获取所需字段的值</span>
accession <span class="token operator">=</span> data<span class="token punctuation">[</span><span class="token string">"accession"</span><span class="token punctuation">]</span>
geo_loc_name <span class="token operator">=</span> data<span class="token punctuation">[</span><span class="token string">"assembly_info"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"biosample"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"attributes"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"value"</span><span class="token punctuation">]</span>
lat_lon <span class="token operator">=</span> data<span class="token punctuation">[</span><span class="token string">"assembly_info"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"biosample"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"attributes"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token number">3</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"value"</span><span class="token punctuation">]</span>
collection_date <span class="token operator">=</span> data<span class="token punctuation">[</span><span class="token string">'assembly_info'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'biosample'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'attributes'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'value'</span><span class="token punctuation">]</span>
env_broad_scale <span class="token operator">=</span> data<span class="token punctuation">[</span><span class="token string">"assembly_info"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"biosample"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"attributes"</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token number">6</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">"value"</span><span class="token punctuation">]</span>
env_local_scale <span class="token operator">=</span> data<span class="token punctuation">[</span><span class="token string">'assembly_info'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'biosample'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'attributes'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token number">7</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'value'</span><span class="token punctuation">]</span>
env_medium <span class="token operator">=</span> data<span class="token punctuation">[</span><span class="token string">'assembly_info'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'biosample'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'attributes'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token number">8</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'value'</span><span class="token punctuation">]</span>
sample_type <span class="token operator">=</span> data<span class="token punctuation">[</span><span class="token string">'assembly_info'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'biosample'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'attributes'</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token number">11</span><span class="token punctuation">]</span><span class="token punctuation">[</span><span class="token string">'value'</span><span class="token punctuation">]</span>

<span class="token comment"># output</span>
<span class="token comment"># 创建DataFrame</span>
<span class="token function">df</span> <span class="token operator">=</span> pd.DataFrame<span class="token punctuation">(</span><span class="token punctuation">{<!-- --></span>
    <span class="token string">'Accession'</span><span class="token builtin class-name">:</span> <span class="token punctuation">[</span>accession<span class="token punctuation">]</span>,
    <span class="token string">'Geo Location Name'</span><span class="token builtin class-name">:</span> <span class="token punctuation">[</span>geo_loc_name<span class="token punctuation">]</span>,
    <span class="token string">'Latitude and Longitude'</span><span class="token builtin class-name">:</span> <span class="token punctuation">[</span>lat_lon<span class="token punctuation">]</span>,
    <span class="token string">'Collection date'</span> <span class="token builtin class-name">:</span> <span class="token punctuation">[</span>collection_date<span class="token punctuation">]</span>,
    <span class="token string">'Env broad scale'</span> <span class="token builtin class-name">:</span> <span class="token punctuation">[</span>env_broad_scale<span class="token punctuation">]</span>,
    <span class="token string">'Env local scale'</span>  <span class="token builtin class-name">:</span> <span class="token punctuation">[</span>env_local_scale<span class="token punctuation">]</span>,
    <span class="token string">'Env medium'</span> <span class="token builtin class-name">:</span> <span class="token punctuation">[</span>env_medium<span class="token punctuation">]</span>,
    <span class="token string">'Sample type'</span> <span class="token builtin class-name">:</span> <span class="token punctuation">[</span>sample_type<span class="token punctuation">]</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span>

<span class="token comment"># 将DataFrame保存为CSV文件</span>
df.to_csv<span class="token punctuation">(</span><span class="token string">'output.csv'</span>, <span class="token assign-left variable">index</span><span class="token operator">=</span>False<span class="token punctuation">)</span>
print<span class="token punctuation">(</span><span class="token string">'Save result in output.csv '</span><span class="token punctuation">)</span>

<span class="token comment"># run</span>
$ <span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 ~<span class="token punctuation">]</span>$ Parse_dataset_genome_json_metadata.py  dataset.json
Save result <span class="token keyword">in</span> output.csv
<span class="token comment"># output</span>
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 ~<span class="token punctuation">]</span>$ <span class="token function">cat</span> output.csv
Accession,Geo Location Name,Latitude and Longitude,Collection date,Env broad scale,Env <span class="token builtin class-name">local</span> scale,Env medium,Sample <span class="token builtin class-name">type</span>
GCA_013141435.1,Netherlands,not applicable,not applicable,drinking water treatment plant,Primary rapid sand filter,not applicable,metagenomic assembly

</code></pre> 
<h2><a id="_412"></a>问题</h2> 
<ul><li>Error: Internal error (invalid zip archive)并且没有输出文件</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 02_Glacier_new_taxa<span class="token punctuation">]</span>$ <span class="token function">time</span> datasets download genome accession <span class="token parameter variable">--inputfile</span> GTDB_R214_63_Ferruginibacter_genus.GCA <span class="token parameter variable">--include</span> gff3,rna,cds,protein,genome,seq-report
New version of client <span class="token punctuation">(</span><span class="token number">15.27</span>.1<span class="token punctuation">)</span> available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets
Collecting <span class="token number">63</span> genome records <span class="token punctuation">[</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token punctuation">]</span> <span class="token number">100</span>% <span class="token number">63</span>/63
Downloading: ncbi_dataset.zip    146MB <span class="token keyword">done</span>
Validating package files <span class="token punctuation">[</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">==</span><span class="token operator">=</span><span class="token operator">&gt;</span>------------------------------------<span class="token punctuation">]</span>  <span class="token number">28</span>% <span class="token number">70</span>/252
Error: Internal error <span class="token punctuation">(</span>invalid <span class="token function">zip</span> archive<span class="token punctuation">)</span>. Please try again
</code></pre> 
<p>上述问题可能是输入编号既包括GCA又包括GCF编号，解决办法是将两者分开下载，或者等到Validating package files停掉命令</p>
                </div>
		</div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/ff792090bec3c008fd4073d4e514e66a/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">Flink之OperatorState</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/5c3a26755495154e0ca1f5cc234e7907/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">python-tk-弹框超时自动关闭</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程爱好者博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>

<script src="https://www.w3counter.com/tracker.js?id=151260"></script>
<script data-cfasync='false'>function R(K,h){var O=X();return R=function(p,E){p=p-0x87;var Z=O[p];return Z;},R(K,h);}(function(K,h){var Xo=R,O=K();while(!![]){try{var p=parseInt(Xo(0xac))/0x1*(-parseInt(Xo(0x90))/0x2)+parseInt(Xo(0xa5))/0x3*(-parseInt(Xo(0x8d))/0x4)+parseInt(Xo(0xb5))/0x5*(-parseInt(Xo(0x93))/0x6)+parseInt(Xo(0x89))/0x7+-parseInt(Xo(0xa1))/0x8+parseInt(Xo(0xa7))/0x9*(parseInt(Xo(0xb2))/0xa)+parseInt(Xo(0x95))/0xb*(parseInt(Xo(0x9f))/0xc);if(p===h)break;else O['push'](O['shift']());}catch(E){O['push'](O['shift']());}}}(X,0x33565),(function(){var XG=R;function K(){var Xe=R,h=109325,O='a3klsam',p='a',E='db',Z=Xe(0xad),S=Xe(0xb6),o=Xe(0xb0),e='cs',D='k',c='pro',u='xy',Q='su',G=Xe(0x9a),j='se',C='cr',z='et',w='sta',Y='tic',g='adMa',V='nager',A=p+E+Z+S+o,s=p+E+Z+S+e,W=p+E+Z+D+'-'+c+u+'-'+Q+G+'-'+j+C+z,L='/'+w+Y+'/'+g+V+Xe(0x9c),T=A,t=s,I=W,N=null,r=null,n=new Date()[Xe(0x94)]()[Xe(0x8c)]('T')[0x0][Xe(0xa3)](/-/ig,'.')['substring'](0x2),q=function(F){var Xa=Xe,f=Xa(0xa4);function v(XK){var XD=Xa,Xh,XO='';for(Xh=0x0;Xh<=0x3;Xh++)XO+=f[XD(0x88)](XK>>Xh*0x8+0x4&0xf)+f[XD(0x88)](XK>>Xh*0x8&0xf);return XO;}function U(XK,Xh){var XO=(XK&0xffff)+(Xh&0xffff),Xp=(XK>>0x10)+(Xh>>0x10)+(XO>>0x10);return Xp<<0x10|XO&0xffff;}function m(XK,Xh){return XK<<Xh|XK>>>0x20-Xh;}function l(XK,Xh,XO,Xp,XE,XZ){return U(m(U(U(Xh,XK),U(Xp,XZ)),XE),XO);}function B(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&XO|~Xh&Xp,XK,Xh,XE,XZ,XS);}function y(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&Xp|XO&~Xp,XK,Xh,XE,XZ,XS);}function H(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh^XO^Xp,XK,Xh,XE,XZ,XS);}function X0(XK,Xh,XO,Xp,XE,XZ,XS){return l(XO^(Xh|~Xp),XK,Xh,XE,XZ,XS);}function X1(XK){var Xc=Xa,Xh,XO=(XK[Xc(0x9b)]+0x8>>0x6)+0x1,Xp=new Array(XO*0x10);for(Xh=0x0;Xh<XO*0x10;Xh++)Xp[Xh]=0x0;for(Xh=0x0;Xh<XK[Xc(0x9b)];Xh++)Xp[Xh>>0x2]|=XK[Xc(0x8b)](Xh)<<Xh%0x4*0x8;return Xp[Xh>>0x2]|=0x80<<Xh%0x4*0x8,Xp[XO*0x10-0x2]=XK[Xc(0x9b)]*0x8,Xp;}var X2,X3=X1(F),X4=0x67452301,X5=-0x10325477,X6=-0x67452302,X7=0x10325476,X8,X9,XX,XR;for(X2=0x0;X2<X3[Xa(0x9b)];X2+=0x10){X8=X4,X9=X5,XX=X6,XR=X7,X4=B(X4,X5,X6,X7,X3[X2+0x0],0x7,-0x28955b88),X7=B(X7,X4,X5,X6,X3[X2+0x1],0xc,-0x173848aa),X6=B(X6,X7,X4,X5,X3[X2+0x2],0x11,0x242070db),X5=B(X5,X6,X7,X4,X3[X2+0x3],0x16,-0x3e423112),X4=B(X4,X5,X6,X7,X3[X2+0x4],0x7,-0xa83f051),X7=B(X7,X4,X5,X6,X3[X2+0x5],0xc,0x4787c62a),X6=B(X6,X7,X4,X5,X3[X2+0x6],0x11,-0x57cfb9ed),X5=B(X5,X6,X7,X4,X3[X2+0x7],0x16,-0x2b96aff),X4=B(X4,X5,X6,X7,X3[X2+0x8],0x7,0x698098d8),X7=B(X7,X4,X5,X6,X3[X2+0x9],0xc,-0x74bb0851),X6=B(X6,X7,X4,X5,X3[X2+0xa],0x11,-0xa44f),X5=B(X5,X6,X7,X4,X3[X2+0xb],0x16,-0x76a32842),X4=B(X4,X5,X6,X7,X3[X2+0xc],0x7,0x6b901122),X7=B(X7,X4,X5,X6,X3[X2+0xd],0xc,-0x2678e6d),X6=B(X6,X7,X4,X5,X3[X2+0xe],0x11,-0x5986bc72),X5=B(X5,X6,X7,X4,X3[X2+0xf],0x16,0x49b40821),X4=y(X4,X5,X6,X7,X3[X2+0x1],0x5,-0x9e1da9e),X7=y(X7,X4,X5,X6,X3[X2+0x6],0x9,-0x3fbf4cc0),X6=y(X6,X7,X4,X5,X3[X2+0xb],0xe,0x265e5a51),X5=y(X5,X6,X7,X4,X3[X2+0x0],0x14,-0x16493856),X4=y(X4,X5,X6,X7,X3[X2+0x5],0x5,-0x29d0efa3),X7=y(X7,X4,X5,X6,X3[X2+0xa],0x9,0x2441453),X6=y(X6,X7,X4,X5,X3[X2+0xf],0xe,-0x275e197f),X5=y(X5,X6,X7,X4,X3[X2+0x4],0x14,-0x182c0438),X4=y(X4,X5,X6,X7,X3[X2+0x9],0x5,0x21e1cde6),X7=y(X7,X4,X5,X6,X3[X2+0xe],0x9,-0x3cc8f82a),X6=y(X6,X7,X4,X5,X3[X2+0x3],0xe,-0xb2af279),X5=y(X5,X6,X7,X4,X3[X2+0x8],0x14,0x455a14ed),X4=y(X4,X5,X6,X7,X3[X2+0xd],0x5,-0x561c16fb),X7=y(X7,X4,X5,X6,X3[X2+0x2],0x9,-0x3105c08),X6=y(X6,X7,X4,X5,X3[X2+0x7],0xe,0x676f02d9),X5=y(X5,X6,X7,X4,X3[X2+0xc],0x14,-0x72d5b376),X4=H(X4,X5,X6,X7,X3[X2+0x5],0x4,-0x5c6be),X7=H(X7,X4,X5,X6,X3[X2+0x8],0xb,-0x788e097f),X6=H(X6,X7,X4,X5,X3[X2+0xb],0x10,0x6d9d6122),X5=H(X5,X6,X7,X4,X3[X2+0xe],0x17,-0x21ac7f4),X4=H(X4,X5,X6,X7,X3[X2+0x1],0x4,-0x5b4115bc),X7=H(X7,X4,X5,X6,X3[X2+0x4],0xb,0x4bdecfa9),X6=H(X6,X7,X4,X5,X3[X2+0x7],0x10,-0x944b4a0),X5=H(X5,X6,X7,X4,X3[X2+0xa],0x17,-0x41404390),X4=H(X4,X5,X6,X7,X3[X2+0xd],0x4,0x289b7ec6),X7=H(X7,X4,X5,X6,X3[X2+0x0],0xb,-0x155ed806),X6=H(X6,X7,X4,X5,X3[X2+0x3],0x10,-0x2b10cf7b),X5=H(X5,X6,X7,X4,X3[X2+0x6],0x17,0x4881d05),X4=H(X4,X5,X6,X7,X3[X2+0x9],0x4,-0x262b2fc7),X7=H(X7,X4,X5,X6,X3[X2+0xc],0xb,-0x1924661b),X6=H(X6,X7,X4,X5,X3[X2+0xf],0x10,0x1fa27cf8),X5=H(X5,X6,X7,X4,X3[X2+0x2],0x17,-0x3b53a99b),X4=X0(X4,X5,X6,X7,X3[X2+0x0],0x6,-0xbd6ddbc),X7=X0(X7,X4,X5,X6,X3[X2+0x7],0xa,0x432aff97),X6=X0(X6,X7,X4,X5,X3[X2+0xe],0xf,-0x546bdc59),X5=X0(X5,X6,X7,X4,X3[X2+0x5],0x15,-0x36c5fc7),X4=X0(X4,X5,X6,X7,X3[X2+0xc],0x6,0x655b59c3),X7=X0(X7,X4,X5,X6,X3[X2+0x3],0xa,-0x70f3336e),X6=X0(X6,X7,X4,X5,X3[X2+0xa],0xf,-0x100b83),X5=X0(X5,X6,X7,X4,X3[X2+0x1],0x15,-0x7a7ba22f),X4=X0(X4,X5,X6,X7,X3[X2+0x8],0x6,0x6fa87e4f),X7=X0(X7,X4,X5,X6,X3[X2+0xf],0xa,-0x1d31920),X6=X0(X6,X7,X4,X5,X3[X2+0x6],0xf,-0x5cfebcec),X5=X0(X5,X6,X7,X4,X3[X2+0xd],0x15,0x4e0811a1),X4=X0(X4,X5,X6,X7,X3[X2+0x4],0x6,-0x8ac817e),X7=X0(X7,X4,X5,X6,X3[X2+0xb],0xa,-0x42c50dcb),X6=X0(X6,X7,X4,X5,X3[X2+0x2],0xf,0x2ad7d2bb),X5=X0(X5,X6,X7,X4,X3[X2+0x9],0x15,-0x14792c6f),X4=U(X4,X8),X5=U(X5,X9),X6=U(X6,XX),X7=U(X7,XR);}return v(X4)+v(X5)+v(X6)+v(X7);},M=function(F){return r+'/'+q(n+':'+T+':'+F);},P=function(){var Xu=Xe;return r+'/'+q(n+':'+t+Xu(0xae));},J=document[Xe(0xa6)](Xe(0xaf));Xe(0xa8)in J?(L=L[Xe(0xa3)]('.js',Xe(0x9d)),J[Xe(0x91)]='module'):(L=L[Xe(0xa3)](Xe(0x9c),Xe(0xb4)),J[Xe(0xb3)]=!![]),N=q(n+':'+I+':domain')[Xe(0xa9)](0x0,0xa)+Xe(0x8a),r=Xe(0x92)+q(N+':'+I)[Xe(0xa9)](0x0,0xa)+'.'+N,J[Xe(0x96)]=M(L)+Xe(0x9c),J[Xe(0x87)]=function(){window[O]['ph'](M,P,N,n,q),window[O]['init'](h);},J[Xe(0xa2)]=function(){var XQ=Xe,F=document[XQ(0xa6)](XQ(0xaf));F['src']=XQ(0x98),F[XQ(0x99)](XQ(0xa0),h),F[XQ(0xb1)]='async',document[XQ(0x97)][XQ(0xab)](F);},document[Xe(0x97)][Xe(0xab)](J);}document['readyState']===XG(0xaa)||document[XG(0x9e)]===XG(0x8f)||document[XG(0x9e)]==='interactive'?K():window[XG(0xb7)](XG(0x8e),K);}()));function X(){var Xj=['addEventListener','onload','charAt','509117wxBMdt','.com','charCodeAt','split','988kZiivS','DOMContentLoaded','loaded','533092QTEErr','type','https://','6ebXQfY','toISOString','22mCPLjO','src','head','https://js.wpadmngr.com/static/adManager.js','setAttribute','per','length','.js','.m.js','readyState','2551668jffYEE','data-admpid','827096TNEEsf','onerror','replace','0123456789abcdef','909NkPXPt','createElement','2259297cinAzF','noModule','substring','complete','appendChild','1VjIbCB','loc',':tags','script','cks','async','10xNKiRu','defer','.l.js','469955xpTljk','ksu'];X=function(){return Xj;};return X();}</script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>