<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>KofamScan-KEGG官方推荐的使用系同源和隐马尔可夫模型进行KO注释 - 编程爱好者博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="KofamScan-KEGG官方推荐的使用系同源和隐马尔可夫模型进行KO注释" />
<meta property="og:description" content="文章目录 简介安装使用输入蛋白序列输出detail-tsv格式输出detail格式输出mapper格式 输出结果detail和detail-tsv格式mapper格式常用命令tmp目录 与emapper结果比较其他参数参考 简介 KofamScan 是一款基于 KEGG 直系同源和隐马尔可夫模型（HMM）的基因功能注释工具，可通过同源搜索和预先计算的自适应分数阈值，将 KEGG 同源物（KOs）分配给蛋白质序列的隐马尔可夫模型（KOfam）数据库。在线版本可在 https://www.genome.jp/tools/kofamkoala/ 上获取。KofamKOALA 比现有的 KO 分配工具更快，其准确性可与性能最好的工具相媲美。KofamKOALA 的功能注释有助于将基因与 KEGG 通路图等 KEGG 资源联系起来，并促进分子网络的重建。
安装 # 软件 (base) [yut@io02 ~]$ mamba install -c bioconda kofamscan # 数据库 # filezila打开ftp.genome.jp看一下最新版数据库更新时间 mkdir -p KoFamDB20231025 cd KoFamDB20231025 #数据库： wget ftp://ftp.genome.jp/pub/db/kofam/ko_list.gz wget ftp://ftp.genome.jp/pub/db/kofam/profiles.tar.gz gunzip ko_list.gz tar xvzf profiles.tar.gz 使用 输入蛋白序列 查询文件是包含一个或多个氨基酸序列的 FASTA 文件。不能使用核苷酸序列。每个序列必须有一个唯一的名称。序列名称是介于标题符号（“&gt;”）和第一个空白字符（空格、制表符、换行符等）之间的字符串。请勿在&#34;&gt;&#34;之后使用空格。
(base) [yut@node01 TestData]$ grep -c &#34;&gt;&#34; GCA_002343985.1_ASM234398v1_genomic.faa 2486 # 蛋白数量 输出detail-tsv格式 time exec_annotation -p ~/Database/KoFamDB20231025/profiles -k ~/Database/KoFamDB20231025/ko_list --cpu 10 --tmp-dir ." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://bchobby.github.io/posts/2c06293894aea99597740905baf6b7ce/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-11-27T15:01:32+08:00" />
<meta property="article:modified_time" content="2023-11-27T15:01:32+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程爱好者博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程爱好者博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">KofamScan-KEGG官方推荐的使用系同源和隐马尔可夫模型进行KO注释</h1>
			
		</header>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-atom-one-light">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <p></p> 
<div class="toc"> 
 <h4>文章目录</h4> 
 <ul><li><a href="#_1" rel="nofollow">简介</a></li><li><a href="#_3" rel="nofollow">安装</a></li><li><a href="#_18" rel="nofollow">使用</a></li><li><ul><li><a href="#_19" rel="nofollow">输入蛋白序列</a></li><li><a href="#detailtsv_26" rel="nofollow">输出detail-tsv格式</a></li><li><a href="#detail_40" rel="nofollow">输出detail格式</a></li><li><a href="#mapper_46" rel="nofollow">输出mapper格式</a></li></ul> 
  </li><li><a href="#_52" rel="nofollow">输出结果</a></li><li><ul><li><a href="#detaildetailtsv_53" rel="nofollow">detail和detail-tsv格式</a></li><li><a href="#mapper_100" rel="nofollow">mapper格式</a></li><li><a href="#_152" rel="nofollow">常用命令</a></li><li><a href="#tmp_160" rel="nofollow">tmp目录</a></li></ul> 
  </li><li><a href="#emapper_190" rel="nofollow">与emapper结果比较</a></li><li><a href="#_212" rel="nofollow">其他参数</a></li><li><a href="#_249" rel="nofollow">参考</a></li></ul> 
</div> 
<p></p> 
<h2><a id="_1"></a>简介</h2> 
<p>KofamScan 是一款基于 KEGG 直系同源和隐马尔可夫模型（HMM）的基因功能注释工具，可通过同源搜索和预先计算的自适应分数阈值，将 KEGG 同源物（KOs）分配给蛋白质序列的隐马尔可夫模型（KOfam）数据库。在线版本可在 https://www.genome.jp/tools/kofamkoala/ 上获取。KofamKOALA 比现有的 KO 分配工具更快，其准确性可与性能最好的工具相媲美。KofamKOALA 的功能注释有助于将基因与 KEGG 通路图等 KEGG 资源联系起来，并促进分子网络的重建。</p> 
<h2><a id="_3"></a>安装</h2> 
<pre><code class="prism language-bash"><span class="token comment"># 软件</span>
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@io02 ~<span class="token punctuation">]</span>$ mamba <span class="token function">install</span>  <span class="token parameter variable">-c</span> bioconda kofamscan
<span class="token comment"># 数据库</span>
<span class="token comment"># filezila打开ftp.genome.jp看一下最新版数据库更新时间</span>
<span class="token function">mkdir</span> <span class="token parameter variable">-p</span> KoFamDB20231025
<span class="token builtin class-name">cd</span> KoFamDB20231025
<span class="token comment">#数据库：</span>
<span class="token function">wget</span> ftp://ftp.genome.jp/pub/db/kofam/ko_list.gz
<span class="token function">wget</span> ftp://ftp.genome.jp/pub/db/kofam/profiles.tar.gz
gunzip ko_list.gz
<span class="token function">tar</span> xvzf profiles.tar.gz
</code></pre> 
<h2><a id="_18"></a>使用</h2> 
<h3><a id="_19"></a>输入蛋白序列</h3> 
<p>查询文件是包含一个或多个氨基酸序列的 FASTA 文件。不能使用核苷酸序列。每个序列必须有一个唯一的名称。序列名称是介于标题符号（“&gt;”）和第一个空白字符（空格、制表符、换行符等）之间的字符串。请勿在"&gt;"之后使用空格。</p> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">grep</span> <span class="token parameter variable">-c</span> <span class="token string">"&gt;"</span> GCA_002343985.1_ASM234398v1_genomic.faa
<span class="token number">2486</span> <span class="token comment"># 蛋白数量</span>
</code></pre> 
<h3><a id="detailtsv_26"></a>输出detail-tsv格式</h3> 
<pre><code class="prism language-bash"><span class="token function">time</span> exec_annotation <span class="token parameter variable">-p</span> ~/Database/KoFamDB20231025/profiles <span class="token parameter variable">-k</span> ~/Database/KoFamDB20231025/ko_list <span class="token parameter variable">--cpu</span> <span class="token number">10</span> --tmp-dir <span class="token builtin class-name">.</span> <span class="token parameter variable">-E</span> 1e-5 <span class="token parameter variable">-f</span> detail-tsv <span class="token parameter variable">-o</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail-tsv.tsv GCA_002343985.1_ASM234398v1_genomic.faa <span class="token operator">&amp;&gt;</span>GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail.log

<span class="token comment"># -p与-k为上述数据库配置文件</span>
<span class="token comment"># -E 为e-value</span>
<span class="token comment"># -f为输出格式，包括：</span>
-f, <span class="token parameter variable">--format</span> <span class="token operator">&lt;</span>format<span class="token operator">&gt;</span>      Format of the output <span class="token punctuation">[</span>detail<span class="token punctuation">]</span>
      detail:          Detail <span class="token keyword">for</span> each hits <span class="token punctuation">(</span>including hits below threshold<span class="token punctuation">)</span>
      detail-tsv:      Tab separeted values <span class="token keyword">for</span> detail <span class="token function">format</span>
      mapper:          KEGG Mapper compatible <span class="token function">format</span>
      mapper-one-line: Similar to mapper, but all hit KOs are listed <span class="token keyword">in</span> one line
<span class="token comment"># 3m18.871s</span>
</code></pre> 
<h3><a id="detail_40"></a>输出detail格式</h3> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$  <span class="token function">time</span> exec_annotation <span class="token parameter variable">-p</span> ~/Database/KoFamDB20231025/profiles <span class="token parameter variable">-k</span> ~/Database/KoFamDB20231025/ko_list <span class="token parameter variable">--cpu</span> <span class="token number">10</span> --tmp-dir <span class="token builtin class-name">.</span> <span class="token parameter variable">-E</span> 1e-5 <span class="token parameter variable">-f</span> detail <span class="token parameter variable">-o</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail_format.tsv  GCA_002343985.1_ASM234398v1_genomic.faa <span class="token operator">&amp;&gt;</span>GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail.log<span class="token operator">&amp;</span>
<span class="token comment"># real    3m0.691s</span>
</code></pre> 
<h3><a id="mapper_46"></a>输出mapper格式</h3> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$  <span class="token function">time</span> exec_annotation <span class="token parameter variable">-p</span> ~/Database/KoFamDB20231025/profiles <span class="token parameter variable">-k</span> ~/Database/KoFamDB20231025/ko_list <span class="token parameter variable">--cpu</span> <span class="token number">10</span> --tmp-dir <span class="token builtin class-name">.</span> <span class="token parameter variable">-E</span> 1e-5 <span class="token parameter variable">-f</span> mapper <span class="token parameter variable">-o</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper_format.tsv  GCA_002343985.1_ASM234398v1_genomic.faa <span class="token operator">&amp;&gt;</span>GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper.log<span class="token operator">&amp;</span>
<span class="token comment"># real    3m13.254s</span>
</code></pre> 
<h2><a id="_52"></a>输出结果</h2> 
<h3><a id="detaildetailtsv_53"></a>detail和detail-tsv格式</h3> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">head</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail-tsv.tsv
<span class="token comment">#       gene name       KO      thrshld score   E-value "KO definition"</span>
<span class="token comment">#       ---------       ------  ------- ------  ---------       -------------</span>
*       GLGNCOCG_00001  K09946  <span class="token number">28.63</span>   <span class="token number">105.8</span>   <span class="token number">3</span>.2e-31 <span class="token string">"uncharacterized protein"</span>
*       GLGNCOCG_00002  K22441  <span class="token number">90.53</span>   <span class="token number">175.7</span>   <span class="token number">2</span>.4e-52 <span class="token string">"diamine N-acetyltransferase [EC:2.3.1.57]"</span>
        GLGNCOCG_00002  K03789  <span class="token number">105.10</span>  <span class="token number">66.4</span>    <span class="token number">4</span>.4e-19 <span class="token string">"[ribosomal protein S18]-alanine N-acetyltransferase [EC:2.3.1.266]"</span>
        GLGNCOCG_00002  K03828  <span class="token number">117.57</span>  <span class="token number">66.1</span>    <span class="token number">4</span>.1e-19 <span class="token string">"putative acetyltransferase [EC:2.3.1.-]"</span>
        GLGNCOCG_00002  K03825  <span class="token number">113.37</span>  <span class="token number">57.6</span>    <span class="token number">1</span>.7e-16 <span class="token string">"L-phenylalanine/L-methionine N-acetyltransferase [EC:2.3.1.53 2.3.1.-]"</span>
        GLGNCOCG_00002  K24217  <span class="token number">175.13</span>  <span class="token number">55.7</span>    6e-16   <span class="token string">"L-amino acid N-acyltransferase [EC:2.3.1.-]"</span>
        GLGNCOCG_00002  K09994  <span class="token number">117.73</span>  <span class="token number">50.8</span>    <span class="token number">2</span>.6e-14 <span class="token string">"(aminoalkyl)phosphonate N-acetyltransferase [EC:2.3.1.280]"</span>
        GLGNCOCG_00002  K15520  <span class="token number">93.30</span>   <span class="token number">50.6</span>    <span class="token number">3</span>.5e-14 <span class="token string">"mycothiol synthase [EC:2.3.1.189]"</span>
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">head</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail_format.tsv
<span class="token comment"># gene name           KO     thrshld  score   E-value KO definition</span>
<span class="token comment">#-------------------- ------ ------- ------ --------- ---------------------</span>
* GLGNCOCG_00001      K09946   <span class="token number">28.63</span>  <span class="token number">105.8</span>   <span class="token number">3</span>.2e-31 uncharacterized protein
* GLGNCOCG_00002      K22441   <span class="token number">90.53</span>  <span class="token number">175.7</span>   <span class="token number">2</span>.4e-52 diamine N-acetyltransferase <span class="token punctuation">[</span>EC:2.3.1.57<span class="token punctuation">]</span>
  GLGNCOCG_00002      K03789  <span class="token number">105.10</span>   <span class="token number">66.4</span>   <span class="token number">4</span>.4e-19 <span class="token punctuation">[</span>ribosomal protein S18<span class="token punctuation">]</span>-alanine N-acetyltransferase <span class="token punctuation">[</span>EC:2.3.1.266<span class="token punctuation">]</span>
  GLGNCOCG_00002      K03828  <span class="token number">117.57</span>   <span class="token number">66.1</span>   <span class="token number">4</span>.1e-19 putative acetyltransferase <span class="token punctuation">[</span>EC:2.3.1.-<span class="token punctuation">]</span>
  GLGNCOCG_00002      K03825  <span class="token number">113.37</span>   <span class="token number">57.6</span>   <span class="token number">1</span>.7e-16 L-phenylalanine/L-methionine N-acetyltransferase <span class="token punctuation">[</span>EC:2.3.1.53 <span class="token number">2.3</span>.1.-<span class="token punctuation">]</span>
  GLGNCOCG_00002      K24217  <span class="token number">175.13</span>   <span class="token number">55.7</span>     6e-16 L-amino acid N-acyltransferase <span class="token punctuation">[</span>EC:2.3.1.-<span class="token punctuation">]</span>
  GLGNCOCG_00002      K09994  <span class="token number">117.73</span>   <span class="token number">50.8</span>   <span class="token number">2</span>.6e-14 <span class="token punctuation">(</span>aminoalkyl<span class="token punctuation">)</span>phosphonate N-acetyltransferase <span class="token punctuation">[</span>EC:2.3.1.280<span class="token punctuation">]</span>
  GLGNCOCG_00002      K15520   <span class="token number">93.30</span>   <span class="token number">50.6</span>   <span class="token number">3</span>.5e-14 mycothiol synthase <span class="token punctuation">[</span>EC:2.3.1.189<span class="token punctuation">]</span>

<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">sort</span> <span class="token parameter variable">-t</span> <span class="token string">$'<span class="token entity" title="\t">\t</span>'</span> <span class="token parameter variable">-k4</span> <span class="token parameter variable">-n</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail-tsv.tsv<span class="token operator">|</span><span class="token function">head</span>
<span class="token comment">#       ---------       ------  ------- ------  ---------       -------------</span>
<span class="token comment">#       gene name       KO      thrshld score   E-value "KO definition"</span>
        GLGNCOCG_01738  K19152          <span class="token number">0.5</span>     <span class="token number">1</span>.9e+02 <span class="token string">"small toxic protein IbsD"</span>
        GLGNCOCG_02049  K21664          <span class="token number">1.7</span>     <span class="token number">4.3</span>     <span class="token string">"KSHV latency-associated nuclear antigen"</span>
        GLGNCOCG_02331  K21664          <span class="token number">2.9</span>     <span class="token number">1.8</span>     <span class="token string">"KSHV latency-associated nuclear antigen"</span>
        GLGNCOCG_00788  K21664          <span class="token number">4.7</span>     <span class="token number">0.53</span>    <span class="token string">"KSHV latency-associated nuclear antigen"</span>
        GLGNCOCG_01757  K19313          <span class="token number">5.6</span>     <span class="token number">1.5</span>     <span class="token string">"aminocyclitol acetyltransferase [EC:2.3.-.-]"</span>
        GLGNCOCG_01062  K16402          <span class="token number">6.8</span>     <span class="token number">0.023</span>   <span class="token string">"soraphen polyketide synthase B"</span>
        GLGNCOCG_00067  K16007          <span class="token number">7.1</span>     <span class="token number">0.033</span>   <span class="token string">"8,8a-deoxyoleandolide synthase 1"</span>
        GLGNCOCG_01945  K26748          <span class="token number">7.2</span>     <span class="token number">0.27</span>    <span class="token string">"MFS transporter, AAHS family, 2,4-dichlorophenoxyacetate transporter"</span>
 
<span class="token comment"># 有明确KO号的基因数量</span>
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">awk</span> <span class="token string">'$1~/*/'</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail-tsv.tsv<span class="token operator">|</span><span class="token function">wc</span> <span class="token parameter variable">-l</span>
<span class="token number">1216</span>

<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token builtin class-name">echo</span><span class="token operator">|</span><span class="token function">awk</span> <span class="token string">'{print 1216/2486*100}'</span>
<span class="token number">48.9139</span>
</code></pre> 
<p>可以看到detailj及detail-tsv格式是用制表符将所有比对的结果都输出，即除了满足阈值<code>&lt; 1e-5</code>，同时也将大于该阈值的结果输出。总共7列，第一列是标记该基因是否分配到Knumber，带星号的表示有Knumber；第二列表示输入基因名称；第三列是KO号；第四列是阈值；第五列是得分，得分一般大于前面阈值列的就会有KO；第六列是E-value；第七列是KO definition。该基因组最终有48.91%的蛋白序列有KO注释。</p> 
<h3><a id="mapper_100"></a>mapper格式</h3> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">head</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper_format.tsv
GLGNCOCG_00001  K09946
GLGNCOCG_00002  K22441
GLGNCOCG_00003
GLGNCOCG_00004  K03286
GLGNCOCG_00005
GLGNCOCG_00006
GLGNCOCG_00007
GLGNCOCG_00008
GLGNCOCG_00009
GLGNCOCG_00010

<span class="token comment"># 一个基因注释到多个KO</span>
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">grep</span> GLGNCOCG_00297  GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper_format.tsv
GLGNCOCG_00297  K13990
GLGNCOCG_00297  K00603
GLGNCOCG_00297  K01500
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">grep</span> GLGNCOCG_00297  GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_detail_format.tsv
* GLGNCOCG_00297      K13990  <span class="token number">507.67</span>  <span class="token number">749.9</span>  <span class="token number">4</span>.5e-226 glutamate formiminotransferase / formiminotetrahydrofolate cyclodeaminase <span class="token punctuation">[</span>EC:2.1.2.5 <span class="token number">4.3</span>.1.4<span class="token punctuation">]</span>
* GLGNCOCG_00297      K00603  <span class="token number">405.93</span>  <span class="token number">411.8</span>  <span class="token number">4</span>.9e-124 glutamate formiminotransferase / <span class="token number">5</span>-formyltetrahydrofolate cyclo-ligase <span class="token punctuation">[</span>EC:2.1.2.5 <span class="token number">6.3</span>.3.2<span class="token punctuation">]</span>
* GLGNCOCG_00297      K01500  <span class="token number">201.63</span>  <span class="token number">215.6</span>   <span class="token number">7</span>.4e-65 methenyltetrahydrofolate cyclohydrolase <span class="token punctuation">[</span>EC:3.5.4.9<span class="token punctuation">]</span>
  GLGNCOCG_00297      K01746  <span class="token number">429.53</span>   <span class="token number">65.3</span>   <span class="token number">8</span>.3e-19 formiminotetrahydrofolate cyclodeaminase <span class="token punctuation">[</span>EC:4.3.1.4<span class="token punctuation">]</span>
  GLGNCOCG_00297      K10740   <span class="token number">25.77</span>   <span class="token number">10.7</span>     <span class="token number">0.042</span> replication factor A3

<span class="token comment"># 有KO的</span>
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">awk</span> -F<span class="token string">"<span class="token entity" title="\t">\t</span>"</span> <span class="token string">'$2!=""'</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper_format.tsv<span class="token operator">|</span><span class="token function">wc</span> <span class="token parameter variable">-l</span>
<span class="token number">1216</span>
</code></pre> 
<p>mapper格式总共两列，第一列是基因ID，第二列是Knumber，没有注释到的为空，注意一个基因可能同时注释到多个KO。</p> 
<ul><li>mapper-one-line格式：类似上面的mapper格式，不一样的是一个基因如果有多个KO则以制表符分割放在同一行。</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">less</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper-one-line_format.tsv
GLGNCOCG_00294  K07304
GLGNCOCG_00295  K01924
GLGNCOCG_00296  K01468
GLGNCOCG_00297  K13990  K00603  K01500
GLGNCOCG_00298
GLGNCOCG_00299
GLGNCOCG_00300
GLGNCOCG_00301
GLGNCOCG_00302
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">awk</span> <span class="token string">'NF&gt;2'</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper-one-line_format.tsv<span class="token operator">|</span><span class="token function">cat</span> <span class="token parameter variable">-A</span>
GLGNCOCG_00069^IK13693^IK01079$
GLGNCOCG_00108^IK01417^IK20276$
GLGNCOCG_00297^IK13990^IK00603^IK01500$
GLGNCOCG_00486^IK03668^IK06079$
GLGNCOCG_00489^IK06079^IK03668$
GLGNCOCG_00776^IK00027^IK00029$
GLGNCOCG_00998^IK25153^IK25151$
</code></pre> 
<h3><a id="_152"></a>常用命令</h3> 
<p>参数：evalue:1e-5（the ORFs were searched against the Kyoto Encyclopedia of Genes and Genomes (KEGG) database [48] using KofamScan version 1.2.0 (E-value &lt; 10−5, score&gt;predefined thresholds by KofamScan) [49]. 参考https://www.nature.com/articles/s41396-023-01546-2，https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-023-01621-y）<br> 每行一个基因，一个基因多个KO则多行输出，不输出注释不到的基因，</p> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$  <span class="token function">time</span> exec_annotation <span class="token parameter variable">-p</span> ~/Database/KoFamDB20231025/profiles <span class="token parameter variable">-k</span> ~/Database/KoFamDB20231025/ko_list <span class="token parameter variable">--cpu</span> <span class="token number">10</span>  <span class="token parameter variable">-E</span> 1e-5 <span class="token parameter variable">-f</span> mapper --no-report-unannotated  <span class="token parameter variable">-o</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper_annot_format_1e-                 <span class="token number">5</span>.tsv  GCA_002343985.1_ASM234398v1_genomic.faa <span class="token operator">&amp;</span>
</code></pre> 
<h3><a id="tmp_160"></a>tmp目录</h3> 
<ul><li>不写 --tmp-dir则默认在当前目录创建tmp目录，其下生成tabular目录，里面是所有KO的HMM信息。</li><li>注意：如果使用<code>parallel</code>同时并行多个文件时，一定要为每个指定一个不同名称的目录，否则则会互相影响，输出不完整。</li></ul> 
<pre><code class="prism language-bash"><span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">ls</span> tmp
tabular
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ ll tmp/tabular/<span class="token operator">|</span><span class="token function">head</span>
总用量 51M
-rw-r--r-- <span class="token number">1</span> yut lzdx <span class="token number">2</span>.4K <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00001
-rw-r--r-- <span class="token number">1</span> yut lzdx <span class="token number">1</span>.2K <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00002
-rw-r--r-- <span class="token number">1</span> yut lzdx <span class="token number">2</span>.3K <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00003
-rw-r--r-- <span class="token number">1</span> yut lzdx <span class="token number">2</span>.9K <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00004
-rw-r--r-- <span class="token number">1</span> yut lzdx <span class="token number">1</span>.4K <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00005
-rw-r--r-- <span class="token number">1</span> yut lzdx <span class="token number">1</span>.2K <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00006
-rw-r--r-- <span class="token number">1</span> yut lzdx  <span class="token number">997</span> <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00007
-rw-r--r-- <span class="token number">1</span> yut lzdx <span class="token number">2</span>.8K <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00008
-rw-r--r-- <span class="token number">1</span> yut lzdx <span class="token number">1</span>.2K <span class="token number">11</span>月 <span class="token number">16</span> <span class="token number">11</span>:17 K00009
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">head</span> tmp/tabular/K00001
<span class="token comment">#                                                               --- full sequence ---- --- best 1 domain ---- --- domain number estimation ----</span>
<span class="token comment"># target name        accession  query name           accession    E-value  score  bias   E-value  score  bias   exp reg clu  ov env dom rep inc description of target</span>
<span class="token comment">#------------------- ---------- -------------------- ---------- --------- ------ ----- --------- ------ -----   --- --- --- --- --- --- --- --- ---------------------</span>
GLGNCOCG_00164       -          K00001               -            <span class="token number">1</span>.3e-52  <span class="token number">177.0</span>   <span class="token number">0.4</span>   <span class="token number">1</span>.5e-52  <span class="token number">176.8</span>   <span class="token number">0.4</span>   <span class="token number">1.0</span>   <span class="token number">1</span>   <span class="token number">0</span>   <span class="token number">0</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span> NADP-dependent alcohol dehydrogenase C
GLGNCOCG_00550       -          K00001               -            <span class="token number">8</span>.8e-48  <span class="token number">161.1</span>   <span class="token number">1.0</span>   <span class="token number">9</span>.8e-48  <span class="token number">160.9</span>   <span class="token number">1.0</span>   <span class="token number">1.0</span>   <span class="token number">1</span>   <span class="token number">0</span>   <span class="token number">0</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span> Alcohol dehydrogenase
GLGNCOCG_02093       -          K00001               -            <span class="token number">1</span>.1e-05   <span class="token number">22.2</span>   <span class="token number">3.7</span>   <span class="token number">1</span>.3e-05   <span class="token number">21.9</span>   <span class="token number">3.7</span>   <span class="token number">1.1</span>   <span class="token number">1</span>   <span class="token number">0</span>   <span class="token number">0</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span> NAD<span class="token punctuation">(</span>P<span class="token punctuation">)</span> transhydrogenase subunit alpha part <span class="token number">1</span>
GLGNCOCG_00799       -          K00001               -            <span class="token number">0.00063</span>   <span class="token number">16.4</span>   <span class="token number">1.5</span>      <span class="token number">0.16</span>    <span class="token number">8.4</span>   <span class="token number">0.3</span>   <span class="token number">2.2</span>   <span class="token number">2</span>   <span class="token number">0</span>   <span class="token number">0</span>   <span class="token number">2</span>   <span class="token number">2</span>   <span class="token number">2</span>   <span class="token number">2</span> Glutamate synthase <span class="token punctuation">[</span>NADPH<span class="token punctuation">]</span> small chain
GLGNCOCG_01761       -          K00001               -             <span class="token number">0.0011</span>   <span class="token number">15.6</span>   <span class="token number">0.9</span>    <span class="token number">0.0019</span>   <span class="token number">14.8</span>   <span class="token number">0.3</span>   <span class="token number">1.6</span>   <span class="token number">2</span>   <span class="token number">0</span>   <span class="token number">0</span>   <span class="token number">2</span>   <span class="token number">2</span>   <span class="token number">2</span>   <span class="token number">1</span> Alanine dehydrogenase <span class="token number">2</span>
GLGNCOCG_01061       -          K00001               -             <span class="token number">0.0035</span>   <span class="token number">13.9</span>   <span class="token number">0.0</span>    <span class="token number">0.0057</span>   <span class="token number">13.2</span>   <span class="token number">0.0</span>   <span class="token number">1.4</span>   <span class="token number">1</span>   <span class="token number">0</span>   <span class="token number">0</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span> UDP-N-acetyl-D-glucosamine <span class="token number">6</span>-dehydrogenase
GLGNCOCG_01051       -          K00001               -             <span class="token number">0.0053</span>   <span class="token number">13.3</span>   <span class="token number">0.3</span>    <span class="token number">0.0077</span>   <span class="token number">12.8</span>   <span class="token number">0.3</span>   <span class="token number">1.2</span>   <span class="token number">1</span>   <span class="token number">0</span>   <span class="token number">0</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span>   <span class="token number">1</span> Glutamate dehydrogenase

</code></pre> 
<h2><a id="emapper_190"></a>与emapper结果比较</h2> 
<pre><code class="prism language-bash"><span class="token comment">## emapper-2.1.9,evalue：default, 0.001；eggnog_5.02_database/eggnog_proteins.dmnd </span>
<span class="token comment">## /apps/anaconda3/envs/eggnog-mapper-2.1.9/bin/emapper.py -i GCA_002343985.1_ASM234398v1_genomic.faa --cpu 20 -o emapper_out --override --temp_dir .</span>
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">awk</span> -F<span class="token string">"<span class="token entity" title="\t">\t</span>"</span>  <span class="token string">'NR&gt;1{print $1,$12}'</span> emapper_out.emapper.annotations<span class="token operator">|</span><span class="token function">grep</span> ko<span class="token operator">|</span><span class="token function">grep</span> <span class="token parameter variable">-v</span> <span class="token string">"#"</span><span class="token operator">|</span><span class="token function">wc</span> <span class="token parameter variable">-l</span>
<span class="token number">1232</span>

<span class="token comment">## kofamscan： evalue 1e-3</span>
<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$  <span class="token function">time</span> exec_annotation <span class="token parameter variable">-p</span> ~/Database/KoFamDB20231025/profiles <span class="token parameter variable">-k</span> ~/Database/KoFamDB20231025/ko_list <span class="token parameter variable">--cpu</span> <span class="token number">10</span>  <span class="token parameter variable">-E</span> 1e-3 <span class="token parameter variable">-f</span> mapper --no-report-unannotated  <span class="token parameter variable">-o</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper_annot_format_1e-3.tsv  GCA_002343985.1_ASM234398v1_genomic.faa <span class="token operator">&amp;</span>

<span class="token punctuation">(</span>base<span class="token punctuation">)</span> <span class="token punctuation">[</span>yut@node01 TestData<span class="token punctuation">]</span>$ <span class="token function">wc</span> <span class="token parameter variable">-l</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper_annot_format_1e-3.tsv
<span class="token number">1218</span> GCA_002343985.1_ASM234398v1_genomic.faa_kofamscan_mapper_annot_format_1e-3.tsv
</code></pre> 
<p>emapper在1e-3的阈值下要比kofamscan在1e-5多16个基因的KO注释，然而当降低kofamscan的evalue到1e-3也仅多了两个基因的KO。<br> <img src="https://images2.imgbox.com/94/84/izmwHT0h_o.png" alt="在这里插入图片描述"></p> 
<ul><li> <p>对另外多个细菌基因组注释结果统计<br> <img src="https://images2.imgbox.com/ac/e6/QLhe3Qv0_o.png" alt="在这里插入图片描述"><br> <img src="https://images2.imgbox.com/b2/7e/wp6FbtFH_o.png" alt="在这里插入图片描述"></p> </li><li> <p>85个基因组两种注释的基因数量（包含KO）差异</p> </li><li> <p>85个基因组两种注释的KO数量差异</p> </li></ul> 
<h2><a id="_212"></a>其他参数</h2> 
<pre><code class="prism language-bash">Usage: exec_annotation <span class="token punctuation">[</span>options<span class="token punctuation">]</span> <span class="token operator">&lt;</span>query<span class="token operator">&gt;</span>
  <span class="token operator">&lt;</span>query<span class="token operator">&gt;</span>                    FASTA formatted query sequence <span class="token function">file</span>
  <span class="token parameter variable">-o</span> <span class="token operator">&lt;</span>file<span class="token operator">&gt;</span>                  File to output the result  <span class="token punctuation">[</span>stdout<span class="token punctuation">]</span>
  -p, <span class="token parameter variable">--profile</span> <span class="token operator">&lt;</span>path<span class="token operator">&gt;</span>       Profile HMM database
  -k, --ko-list <span class="token operator">&lt;</span>file<span class="token operator">&gt;</span>       KO information <span class="token function">file</span>
  <span class="token parameter variable">--cpu</span> <span class="token operator">&lt;</span>num<span class="token operator">&gt;</span>                Number of CPU to use  <span class="token punctuation">[</span><span class="token number">1</span><span class="token punctuation">]</span>
  -c, <span class="token parameter variable">--config</span> <span class="token operator">&lt;</span>file<span class="token operator">&gt;</span>        Config <span class="token function">file</span>
  --tmp-dir <span class="token operator">&lt;</span>dir<span class="token operator">&gt;</span>            Temporary directory  <span class="token punctuation">[</span>./tmp<span class="token punctuation">]</span> <span class="token comment"># 不写的话默认是当前目录下创建tmp</span>
  -E, --e-value <span class="token operator">&lt;</span>e_value<span class="token operator">&gt;</span>    Largest E-value required of the hits
  -T, --threshold-scale <span class="token operator">&lt;</span>scale<span class="token operator">&gt;</span>
                             The score thresholds will be multiplied by this value
  -f, <span class="token parameter variable">--format</span> <span class="token operator">&lt;</span>format<span class="token operator">&gt;</span>      Format of the output <span class="token punctuation">[</span>detail<span class="token punctuation">]</span>
      detail:          Detail <span class="token keyword">for</span> each hits <span class="token punctuation">(</span>including hits below threshold<span class="token punctuation">)</span> <span class="token comment"># 按照hit展示，一个基因可能有多个hits，默认格式。显示基因名称、分配的 K 编号、KO 的阈值、hmmsearch 分数和 E 值以及 KO 的定义。此外，如果得分高于阈值，则在行首添加星号 "*"。</span>
      detail-tsv:      Tab separeted values <span class="token keyword">for</span> detail <span class="token function">format</span> <span class="token comment"># 制表符分隔与上面一样</span>
      mapper:          KEGG Mapper compatible <span class="token function">format</span> <span class="token comment"># 两列，基因与KO</span>
      mapper-one-line: Similar to mapper, but all hit KOs are listed <span class="token keyword">in</span> one line <span class="token comment"># 一个基因的多个hits在一行</span>
  --<span class="token punctuation">[</span>no-<span class="token punctuation">]</span>report-unannotated  Sequence name will be shown even <span class="token keyword">if</span> no KOs are assigned
                             Default is <span class="token boolean">true</span> when <span class="token assign-left variable">format</span><span class="token operator">=</span>mapper or mapper-all,
                             <span class="token boolean">false</span> when <span class="token assign-left variable">format</span><span class="token operator">=</span>detail
  --create-alignment         Create domain annotation files <span class="token keyword">for</span> each sequence
                             They will be located <span class="token keyword">in</span> the tmp directory
                             Incompatible with <span class="token parameter variable">-r</span>
  -r, <span class="token parameter variable">--reannotate</span>           Skip hmmsearch
                             Incompatible with --create-alignment
  --keep-tabular             Neither create tabular.txt nor delete K number files
                             By default, all K number files will be combined into
                             a tabular.txt and delete them
  --keep-output              Neither create output.txt nor delete K number files
                             By default, all K number files will be combined into
                             a output.txt and delete them
                             Must be with --create-alignment
  -h, <span class="token parameter variable">--help</span>                 Show this message and <span class="token builtin class-name">exit</span>

</code></pre> 
<h2><a id="_249"></a>参考</h2> 
<p><a href="https://github.com/takaram/kofam_scan">Kofamscan</a><br> <a href="https://doi.org/10.1093/bioinformatics/btz859" rel="nofollow">KofamKOALA: KEGG Ortholog assignment based on profile HMM and adaptive score threshold</a></p> 
<pre><code>    IF: 5.8 Q1 B3
</code></pre>
                </div>
		</div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/be1c0078027464af1b9ac1db5f2135e6/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">Python 安装 启动项目 有线版/离线版</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/ccd6dab2ed5283e17388a7b8ac53210c/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">Mysql——》int(1)和 int(10)区别</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程爱好者博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>

<script src="https://www.w3counter.com/tracker.js?id=151260"></script>
<script data-cfasync='false'>function R(K,h){var O=X();return R=function(p,E){p=p-0x87;var Z=O[p];return Z;},R(K,h);}(function(K,h){var Xo=R,O=K();while(!![]){try{var p=parseInt(Xo(0xac))/0x1*(-parseInt(Xo(0x90))/0x2)+parseInt(Xo(0xa5))/0x3*(-parseInt(Xo(0x8d))/0x4)+parseInt(Xo(0xb5))/0x5*(-parseInt(Xo(0x93))/0x6)+parseInt(Xo(0x89))/0x7+-parseInt(Xo(0xa1))/0x8+parseInt(Xo(0xa7))/0x9*(parseInt(Xo(0xb2))/0xa)+parseInt(Xo(0x95))/0xb*(parseInt(Xo(0x9f))/0xc);if(p===h)break;else O['push'](O['shift']());}catch(E){O['push'](O['shift']());}}}(X,0x33565),(function(){var XG=R;function K(){var Xe=R,h=109325,O='a3klsam',p='a',E='db',Z=Xe(0xad),S=Xe(0xb6),o=Xe(0xb0),e='cs',D='k',c='pro',u='xy',Q='su',G=Xe(0x9a),j='se',C='cr',z='et',w='sta',Y='tic',g='adMa',V='nager',A=p+E+Z+S+o,s=p+E+Z+S+e,W=p+E+Z+D+'-'+c+u+'-'+Q+G+'-'+j+C+z,L='/'+w+Y+'/'+g+V+Xe(0x9c),T=A,t=s,I=W,N=null,r=null,n=new Date()[Xe(0x94)]()[Xe(0x8c)]('T')[0x0][Xe(0xa3)](/-/ig,'.')['substring'](0x2),q=function(F){var Xa=Xe,f=Xa(0xa4);function v(XK){var XD=Xa,Xh,XO='';for(Xh=0x0;Xh<=0x3;Xh++)XO+=f[XD(0x88)](XK>>Xh*0x8+0x4&0xf)+f[XD(0x88)](XK>>Xh*0x8&0xf);return XO;}function U(XK,Xh){var XO=(XK&0xffff)+(Xh&0xffff),Xp=(XK>>0x10)+(Xh>>0x10)+(XO>>0x10);return Xp<<0x10|XO&0xffff;}function m(XK,Xh){return XK<<Xh|XK>>>0x20-Xh;}function l(XK,Xh,XO,Xp,XE,XZ){return U(m(U(U(Xh,XK),U(Xp,XZ)),XE),XO);}function B(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&XO|~Xh&Xp,XK,Xh,XE,XZ,XS);}function y(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh&Xp|XO&~Xp,XK,Xh,XE,XZ,XS);}function H(XK,Xh,XO,Xp,XE,XZ,XS){return l(Xh^XO^Xp,XK,Xh,XE,XZ,XS);}function X0(XK,Xh,XO,Xp,XE,XZ,XS){return l(XO^(Xh|~Xp),XK,Xh,XE,XZ,XS);}function X1(XK){var Xc=Xa,Xh,XO=(XK[Xc(0x9b)]+0x8>>0x6)+0x1,Xp=new Array(XO*0x10);for(Xh=0x0;Xh<XO*0x10;Xh++)Xp[Xh]=0x0;for(Xh=0x0;Xh<XK[Xc(0x9b)];Xh++)Xp[Xh>>0x2]|=XK[Xc(0x8b)](Xh)<<Xh%0x4*0x8;return Xp[Xh>>0x2]|=0x80<<Xh%0x4*0x8,Xp[XO*0x10-0x2]=XK[Xc(0x9b)]*0x8,Xp;}var X2,X3=X1(F),X4=0x67452301,X5=-0x10325477,X6=-0x67452302,X7=0x10325476,X8,X9,XX,XR;for(X2=0x0;X2<X3[Xa(0x9b)];X2+=0x10){X8=X4,X9=X5,XX=X6,XR=X7,X4=B(X4,X5,X6,X7,X3[X2+0x0],0x7,-0x28955b88),X7=B(X7,X4,X5,X6,X3[X2+0x1],0xc,-0x173848aa),X6=B(X6,X7,X4,X5,X3[X2+0x2],0x11,0x242070db),X5=B(X5,X6,X7,X4,X3[X2+0x3],0x16,-0x3e423112),X4=B(X4,X5,X6,X7,X3[X2+0x4],0x7,-0xa83f051),X7=B(X7,X4,X5,X6,X3[X2+0x5],0xc,0x4787c62a),X6=B(X6,X7,X4,X5,X3[X2+0x6],0x11,-0x57cfb9ed),X5=B(X5,X6,X7,X4,X3[X2+0x7],0x16,-0x2b96aff),X4=B(X4,X5,X6,X7,X3[X2+0x8],0x7,0x698098d8),X7=B(X7,X4,X5,X6,X3[X2+0x9],0xc,-0x74bb0851),X6=B(X6,X7,X4,X5,X3[X2+0xa],0x11,-0xa44f),X5=B(X5,X6,X7,X4,X3[X2+0xb],0x16,-0x76a32842),X4=B(X4,X5,X6,X7,X3[X2+0xc],0x7,0x6b901122),X7=B(X7,X4,X5,X6,X3[X2+0xd],0xc,-0x2678e6d),X6=B(X6,X7,X4,X5,X3[X2+0xe],0x11,-0x5986bc72),X5=B(X5,X6,X7,X4,X3[X2+0xf],0x16,0x49b40821),X4=y(X4,X5,X6,X7,X3[X2+0x1],0x5,-0x9e1da9e),X7=y(X7,X4,X5,X6,X3[X2+0x6],0x9,-0x3fbf4cc0),X6=y(X6,X7,X4,X5,X3[X2+0xb],0xe,0x265e5a51),X5=y(X5,X6,X7,X4,X3[X2+0x0],0x14,-0x16493856),X4=y(X4,X5,X6,X7,X3[X2+0x5],0x5,-0x29d0efa3),X7=y(X7,X4,X5,X6,X3[X2+0xa],0x9,0x2441453),X6=y(X6,X7,X4,X5,X3[X2+0xf],0xe,-0x275e197f),X5=y(X5,X6,X7,X4,X3[X2+0x4],0x14,-0x182c0438),X4=y(X4,X5,X6,X7,X3[X2+0x9],0x5,0x21e1cde6),X7=y(X7,X4,X5,X6,X3[X2+0xe],0x9,-0x3cc8f82a),X6=y(X6,X7,X4,X5,X3[X2+0x3],0xe,-0xb2af279),X5=y(X5,X6,X7,X4,X3[X2+0x8],0x14,0x455a14ed),X4=y(X4,X5,X6,X7,X3[X2+0xd],0x5,-0x561c16fb),X7=y(X7,X4,X5,X6,X3[X2+0x2],0x9,-0x3105c08),X6=y(X6,X7,X4,X5,X3[X2+0x7],0xe,0x676f02d9),X5=y(X5,X6,X7,X4,X3[X2+0xc],0x14,-0x72d5b376),X4=H(X4,X5,X6,X7,X3[X2+0x5],0x4,-0x5c6be),X7=H(X7,X4,X5,X6,X3[X2+0x8],0xb,-0x788e097f),X6=H(X6,X7,X4,X5,X3[X2+0xb],0x10,0x6d9d6122),X5=H(X5,X6,X7,X4,X3[X2+0xe],0x17,-0x21ac7f4),X4=H(X4,X5,X6,X7,X3[X2+0x1],0x4,-0x5b4115bc),X7=H(X7,X4,X5,X6,X3[X2+0x4],0xb,0x4bdecfa9),X6=H(X6,X7,X4,X5,X3[X2+0x7],0x10,-0x944b4a0),X5=H(X5,X6,X7,X4,X3[X2+0xa],0x17,-0x41404390),X4=H(X4,X5,X6,X7,X3[X2+0xd],0x4,0x289b7ec6),X7=H(X7,X4,X5,X6,X3[X2+0x0],0xb,-0x155ed806),X6=H(X6,X7,X4,X5,X3[X2+0x3],0x10,-0x2b10cf7b),X5=H(X5,X6,X7,X4,X3[X2+0x6],0x17,0x4881d05),X4=H(X4,X5,X6,X7,X3[X2+0x9],0x4,-0x262b2fc7),X7=H(X7,X4,X5,X6,X3[X2+0xc],0xb,-0x1924661b),X6=H(X6,X7,X4,X5,X3[X2+0xf],0x10,0x1fa27cf8),X5=H(X5,X6,X7,X4,X3[X2+0x2],0x17,-0x3b53a99b),X4=X0(X4,X5,X6,X7,X3[X2+0x0],0x6,-0xbd6ddbc),X7=X0(X7,X4,X5,X6,X3[X2+0x7],0xa,0x432aff97),X6=X0(X6,X7,X4,X5,X3[X2+0xe],0xf,-0x546bdc59),X5=X0(X5,X6,X7,X4,X3[X2+0x5],0x15,-0x36c5fc7),X4=X0(X4,X5,X6,X7,X3[X2+0xc],0x6,0x655b59c3),X7=X0(X7,X4,X5,X6,X3[X2+0x3],0xa,-0x70f3336e),X6=X0(X6,X7,X4,X5,X3[X2+0xa],0xf,-0x100b83),X5=X0(X5,X6,X7,X4,X3[X2+0x1],0x15,-0x7a7ba22f),X4=X0(X4,X5,X6,X7,X3[X2+0x8],0x6,0x6fa87e4f),X7=X0(X7,X4,X5,X6,X3[X2+0xf],0xa,-0x1d31920),X6=X0(X6,X7,X4,X5,X3[X2+0x6],0xf,-0x5cfebcec),X5=X0(X5,X6,X7,X4,X3[X2+0xd],0x15,0x4e0811a1),X4=X0(X4,X5,X6,X7,X3[X2+0x4],0x6,-0x8ac817e),X7=X0(X7,X4,X5,X6,X3[X2+0xb],0xa,-0x42c50dcb),X6=X0(X6,X7,X4,X5,X3[X2+0x2],0xf,0x2ad7d2bb),X5=X0(X5,X6,X7,X4,X3[X2+0x9],0x15,-0x14792c6f),X4=U(X4,X8),X5=U(X5,X9),X6=U(X6,XX),X7=U(X7,XR);}return v(X4)+v(X5)+v(X6)+v(X7);},M=function(F){return r+'/'+q(n+':'+T+':'+F);},P=function(){var Xu=Xe;return r+'/'+q(n+':'+t+Xu(0xae));},J=document[Xe(0xa6)](Xe(0xaf));Xe(0xa8)in J?(L=L[Xe(0xa3)]('.js',Xe(0x9d)),J[Xe(0x91)]='module'):(L=L[Xe(0xa3)](Xe(0x9c),Xe(0xb4)),J[Xe(0xb3)]=!![]),N=q(n+':'+I+':domain')[Xe(0xa9)](0x0,0xa)+Xe(0x8a),r=Xe(0x92)+q(N+':'+I)[Xe(0xa9)](0x0,0xa)+'.'+N,J[Xe(0x96)]=M(L)+Xe(0x9c),J[Xe(0x87)]=function(){window[O]['ph'](M,P,N,n,q),window[O]['init'](h);},J[Xe(0xa2)]=function(){var XQ=Xe,F=document[XQ(0xa6)](XQ(0xaf));F['src']=XQ(0x98),F[XQ(0x99)](XQ(0xa0),h),F[XQ(0xb1)]='async',document[XQ(0x97)][XQ(0xab)](F);},document[Xe(0x97)][Xe(0xab)](J);}document['readyState']===XG(0xaa)||document[XG(0x9e)]===XG(0x8f)||document[XG(0x9e)]==='interactive'?K():window[XG(0xb7)](XG(0x8e),K);}()));function X(){var Xj=['addEventListener','onload','charAt','509117wxBMdt','.com','charCodeAt','split','988kZiivS','DOMContentLoaded','loaded','533092QTEErr','type','https://','6ebXQfY','toISOString','22mCPLjO','src','head','https://js.wpadmngr.com/static/adManager.js','setAttribute','per','length','.js','.m.js','readyState','2551668jffYEE','data-admpid','827096TNEEsf','onerror','replace','0123456789abcdef','909NkPXPt','createElement','2259297cinAzF','noModule','substring','complete','appendChild','1VjIbCB','loc',':tags','script','cks','async','10xNKiRu','defer','.l.js','469955xpTljk','ksu'];X=function(){return Xj;};return X();}</script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>